<!-- HTML header for doxygen 1.8.8-->
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.13"/>
<title>trimAl: newAlignment Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtreedata.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<script type="text/javascript">
  $(document).ready(initResizable);
</script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0" width=100%>
 <tbody>
 <tr style="height: 12px;">
  <td style="padding-left: 0.5em;">
   <div id="projectname">trimAl
   &#20;<span id="projectnumber" style="font-size:12px">1.x</span>
   </div>
   <div id="projectbrief">A Multiple Alignment Sequence framework for Cleaning and Reformating.</div>
  </td>
<!--  <td style="float:right;margin-top:18px;vertical-align:bottom">
     <a href="http://assimilationsystems.com/download/">
       <img src="http://assimilationsystems.com/wp-content/uploads/2015/03/Download-Doxygen.png" style="float:right;" height="20" width="73">
     </a>
  </td>-->
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.13 -->
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
$(function() {
  initMenu('',false,false,'search.php','Search');
});
</script>
<div id="main-nav"></div>
</div><!-- top -->
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
      <div id="nav-sync" class="sync"></div>
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
$(document).ready(function(){initNavTree('classnew_alignment.html','');});
</script>
<div id="doc-content">
<div class="header">
  <div class="summary">
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="#pub-attribs">Public Attributes</a> &#124;
<a href="#pri-attribs">Private Attributes</a> &#124;
<a href="classnew_alignment-members.html">List of all members</a>  </div>
  <div class="headertitle">
<div class="title">newAlignment Class Reference</div>  </div>
</div><!--header-->
<div class="contents">

<p>Class containing an alignment<br />
 This class stores the alignment sequences with it's names, residues and extra information.<br />
 It contains multiple methods regarding the sequences.<br />
 It also contains submodules that provide methods for<b> Calculating statistics</b>, <b> Trim the alignment</b> and<b> Printing sequences information</b>.  
 <a href="classnew_alignment.html#details">More...</a></p>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:acbeb320f0f70222f81a906ccfc33a016"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#acbeb320f0f70222f81a906ccfc33a016">fillMatrices</a> (bool aligned)</td></tr>
<tr class="memdesc:acbeb320f0f70222f81a906ccfc33a016"><td class="mdescLeft">&#160;</td><td class="mdescRight">Method to initialize matrices and check if the sequences have been correctly loaded and are free of errors<br />
It checks if sequences contain unknown characters, sets the isAligned flag and initializes 'saveResidues' and 'saveSequences', depending on the sizes of the sequences (whether or not they have the same length).  <a href="#acbeb320f0f70222f81a906ccfc33a016">More...</a><br /></td></tr>
<tr class="separator:acbeb320f0f70222f81a906ccfc33a016"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a07b1cb8e0045677f6684ff37352af1b1"><td class="memItemLeft" align="right" valign="top"><a id="a07b1cb8e0045677f6684ff37352af1b1"></a>
&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a07b1cb8e0045677f6684ff37352af1b1">newAlignment</a> ()</td></tr>
<tr class="memdesc:a07b1cb8e0045677f6684ff37352af1b1"><td class="mdescLeft">&#160;</td><td class="mdescRight">Constructor. <br /></td></tr>
<tr class="separator:a07b1cb8e0045677f6684ff37352af1b1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8e7fbe8f58d588cb10cb10981efb9915"><td class="memItemLeft" align="right" valign="top"><a id="a8e7fbe8f58d588cb10cb10981efb9915"></a>
&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a8e7fbe8f58d588cb10cb10981efb9915">newAlignment</a> (<a class="el" href="classnew_alignment.html">newAlignment</a> &amp;)</td></tr>
<tr class="memdesc:a8e7fbe8f58d588cb10cb10981efb9915"><td class="mdescLeft">&#160;</td><td class="mdescRight">Copy Constructor. <br /></td></tr>
<tr class="separator:a8e7fbe8f58d588cb10cb10981efb9915"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a02ca374daf15b6af5fdf88c75454323a"><td class="memItemLeft" align="right" valign="top"><a id="a02ca374daf15b6af5fdf88c75454323a"></a>
&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a02ca374daf15b6af5fdf88c75454323a">~newAlignment</a> ()</td></tr>
<tr class="memdesc:a02ca374daf15b6af5fdf88c75454323a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Destructor. <br /></td></tr>
<tr class="separator:a02ca374daf15b6af5fdf88c75454323a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:acfd820148c267a3d698a39758ecf190d"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#acfd820148c267a3d698a39758ecf190d">getNumSpecies</a> ()</td></tr>
<tr class="memdesc:acfd820148c267a3d698a39758ecf190d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Number of sequences getter.  <a href="#acfd820148c267a3d698a39758ecf190d">More...</a><br /></td></tr>
<tr class="separator:acfd820148c267a3d698a39758ecf190d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0db3e76f82a71c7a2fe5b9279f8fa310"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a0db3e76f82a71c7a2fe5b9279f8fa310">getSequences</a> (string *names)</td></tr>
<tr class="memdesc:a0db3e76f82a71c7a2fe5b9279f8fa310"><td class="mdescLeft">&#160;</td><td class="mdescRight">Getter for the sequences names.  <a href="#a0db3e76f82a71c7a2fe5b9279f8fa310">More...</a><br /></td></tr>
<tr class="separator:a0db3e76f82a71c7a2fe5b9279f8fa310"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af32ab6b3b73c5ad37c9f1c796a4fbc44"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#af32ab6b3b73c5ad37c9f1c796a4fbc44">getSequences</a> (string *names, int *lengths)</td></tr>
<tr class="memdesc:af32ab6b3b73c5ad37c9f1c796a4fbc44"><td class="mdescLeft">&#160;</td><td class="mdescRight">Getter for the sequences names and its lengths.  <a href="#af32ab6b3b73c5ad37c9f1c796a4fbc44">More...</a><br /></td></tr>
<tr class="separator:af32ab6b3b73c5ad37c9f1c796a4fbc44"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2fc0d691c0cce400f007446cc162704e"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a2fc0d691c0cce400f007446cc162704e">getSequences</a> (string *names, string *<a class="el" href="classnew_alignment.html#a23f06ad268f0e720bf3d5699a4cb7225">sequences</a>, int *lenghts)</td></tr>
<tr class="memdesc:a2fc0d691c0cce400f007446cc162704e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Getter for the sequences, its names and lenghts.  <a href="#a2fc0d691c0cce400f007446cc162704e">More...</a><br /></td></tr>
<tr class="separator:a2fc0d691c0cce400f007446cc162704e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac3d5c9670904bba45707746640aac382"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#ac3d5c9670904bba45707746640aac382">getSequenceNameOrder</a> (string *names, int *orderVector)</td></tr>
<tr class="memdesc:ac3d5c9670904bba45707746640aac382"><td class="mdescLeft">&#160;</td><td class="mdescRight">Method to map two sets of sequences, own and external.<br />
The method accepts a vector of names, and test them against the alignment own sequences' names.<br />
If the sequence is present, its index will be stored in orderVector.<br />
.  <a href="#ac3d5c9670904bba45707746640aac382">More...</a><br /></td></tr>
<tr class="separator:ac3d5c9670904bba45707746640aac382"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad08d05c08eb66665defed4a16ab08e91"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#ad08d05c08eb66665defed4a16ab08e91">getNumAminos</a> ()</td></tr>
<tr class="memdesc:ad08d05c08eb66665defed4a16ab08e91"><td class="mdescLeft">&#160;</td><td class="mdescRight">Residues number getter. It counts gaps as residue.  <a href="#ad08d05c08eb66665defed4a16ab08e91">More...</a><br /></td></tr>
<tr class="separator:ad08d05c08eb66665defed4a16ab08e91"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8dd7e7438161fcaca8c8f35e63c5079e"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a8dd7e7438161fcaca8c8f35e63c5079e">setWindowsSize</a> (int ghWindow, int shWindow)</td></tr>
<tr class="memdesc:a8dd7e7438161fcaca8c8f35e63c5079e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Windows setter.  <a href="#a8dd7e7438161fcaca8c8f35e63c5079e">More...</a><br /></td></tr>
<tr class="separator:a8dd7e7438161fcaca8c8f35e63c5079e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a42894635dca6466308d9e21cec14d365"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a42894635dca6466308d9e21cec14d365">setBlockSize</a> (int blockSize)</td></tr>
<tr class="memdesc:a42894635dca6466308d9e21cec14d365"><td class="mdescLeft">&#160;</td><td class="mdescRight">blockSize BlockSize Setter.  <a href="#a42894635dca6466308d9e21cec14d365">More...</a><br /></td></tr>
<tr class="separator:a42894635dca6466308d9e21cec14d365"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4761e21288fc15f3b86cf80dd4c6c91b"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a4761e21288fc15f3b86cf80dd4c6c91b">getBlockSize</a> ()</td></tr>
<tr class="memdesc:a4761e21288fc15f3b86cf80dd4c6c91b"><td class="mdescLeft">&#160;</td><td class="mdescRight">BlockSize getter.  <a href="#a4761e21288fc15f3b86cf80dd4c6c91b">More...</a><br /></td></tr>
<tr class="separator:a4761e21288fc15f3b86cf80dd4c6c91b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa9178bd574db2b6171c0bb09133bdb94"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#aa9178bd574db2b6171c0bb09133bdb94">calculateSeqIdentity</a> ()</td></tr>
<tr class="memdesc:aa9178bd574db2b6171c0bb09133bdb94"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculates identity between sequences.  <a href="#aa9178bd574db2b6171c0bb09133bdb94">More...</a><br /></td></tr>
<tr class="separator:aa9178bd574db2b6171c0bb09133bdb94"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:adaeafc911344fc10069f40d9fc41303f"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#adaeafc911344fc10069f40d9fc41303f">calculateRelaxedSeqIdentity</a> ()</td></tr>
<tr class="memdesc:adaeafc911344fc10069f40d9fc41303f"><td class="mdescLeft">&#160;</td><td class="mdescRight">Lightweight version of <a class="el" href="classnew_alignment.html#aa9178bd574db2b6171c0bb09133bdb94" title="Calculates identity between sequences. ">newAlignment::calculateSeqIdentity</a> for use in huge alignments.  <a href="#adaeafc911344fc10069f40d9fc41303f">More...</a><br /></td></tr>
<tr class="separator:adaeafc911344fc10069f40d9fc41303f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae157ddd55e9f41d809fb664116619802"><td class="memItemLeft" align="right" valign="top"><a id="ae157ddd55e9f41d809fb664116619802"></a>
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#ae157ddd55e9f41d809fb664116619802">calculateSeqOverlap</a> ()</td></tr>
<tr class="memdesc:ae157ddd55e9f41d809fb664116619802"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculates overlap between sequences. <br /></td></tr>
<tr class="separator:ae157ddd55e9f41d809fb664116619802"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1180033903a517103612ae59fd024b2f"><td class="memItemLeft" align="right" valign="top"><a id="a1180033903a517103612ae59fd024b2f"></a>
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a1180033903a517103612ae59fd024b2f">printSeqIdentity</a> ()</td></tr>
<tr class="memdesc:a1180033903a517103612ae59fd024b2f"><td class="mdescLeft">&#160;</td><td class="mdescRight">Method to print different identity values computed from the alignment.<br />
In this method we assess the identity values matrix, as well as different average values. <br />
Moreover, the method computes which one is the most similar sequence. <br /></td></tr>
<tr class="separator:a1180033903a517103612ae59fd024b2f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8fa808afb7d7813adbd5acdd2b0effef"><td class="memItemLeft" align="right" valign="top"><a id="a8fa808afb7d7813adbd5acdd2b0effef"></a>
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a8fa808afb7d7813adbd5acdd2b0effef">printSeqOverlap</a> ()</td></tr>
<tr class="memdesc:a8fa808afb7d7813adbd5acdd2b0effef"><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints the overlap between sequences. <br /></td></tr>
<tr class="separator:a8fa808afb7d7813adbd5acdd2b0effef"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a670d5ef1ced9ed6041c3f6afdbd2cc02"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a670d5ef1ced9ed6041c3f6afdbd2cc02">getAlignmentType</a> ()</td></tr>
<tr class="memdesc:a670d5ef1ced9ed6041c3f6afdbd2cc02"><td class="mdescLeft">&#160;</td><td class="mdescRight">Alignment type getter. <br />
 See <a class="el" href="">SequenceTypes.</a> <a href="#a670d5ef1ced9ed6041c3f6afdbd2cc02">More...</a><br /></td></tr>
<tr class="separator:a670d5ef1ced9ed6041c3f6afdbd2cc02"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ace3be92e3229fc241557597090977ba5"><td class="memItemLeft" align="right" valign="top">int *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#ace3be92e3229fc241557597090977ba5">getCorrespResidues</a> ()</td></tr>
<tr class="memdesc:ace3be92e3229fc241557597090977ba5"><td class="mdescLeft">&#160;</td><td class="mdescRight">Method that returns the correspondence between old and new alignment residues. <br />
 See <a class="el" href="classnew_alignment.html#a5239bd4e742631b58efc7f09dd67f165" title="Vector containing which residues will be kept/removed. -1 Indicates a &#39;removed/toRemove&#39;. Otherwise residue/column will be kept. ">newAlignment::saveResidues</a>.  <a href="#ace3be92e3229fc241557597090977ba5">More...</a><br /></td></tr>
<tr class="separator:ace3be92e3229fc241557597090977ba5"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0c3cd77c8a6b118efb84b577bb4ab1ce"><td class="memItemLeft" align="right" valign="top">int *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a0c3cd77c8a6b118efb84b577bb4ab1ce">getCorrespSequences</a> ()</td></tr>
<tr class="memdesc:a0c3cd77c8a6b118efb84b577bb4ab1ce"><td class="mdescLeft">&#160;</td><td class="mdescRight">Method that returns the correspondence between old and new alignment sequences. <br />
 See <a class="el" href="classnew_alignment.html#a76d057ddcebc7c88dd7f9e3ff0ce95bf" title="Vector containing which sequences will be kept/removed. -1 Indicates a &#39;removed/toRemove&#39;. Otherwise sequence will be kept. ">newAlignment::saveSequences</a>.  <a href="#a0c3cd77c8a6b118efb84b577bb4ab1ce">More...</a><br /></td></tr>
<tr class="separator:a0c3cd77c8a6b118efb84b577bb4ab1ce"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab865e218a05742c00d604f4a71ce9e10"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#ab865e218a05742c00d604f4a71ce9e10">isFileAligned</a> ()</td></tr>
<tr class="memdesc:ab865e218a05742c00d604f4a71ce9e10"><td class="mdescLeft">&#160;</td><td class="mdescRight">isAligned getter.  <a href="#ab865e218a05742c00d604f4a71ce9e10">More...</a><br /></td></tr>
<tr class="separator:ab865e218a05742c00d604f4a71ce9e10"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a781fd0f2ae6a9bda2802ffd28372bfaa"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classnew_alignment.html">newAlignment</a> *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a781fd0f2ae6a9bda2802ffd28372bfaa">getTranslationCDS</a> (<a class="el" href="classnew_alignment.html">newAlignment</a> *proteinAlignment)</td></tr>
<tr class="memdesc:a781fd0f2ae6a9bda2802ffd28372bfaa"><td class="mdescLeft">&#160;</td><td class="mdescRight">Method to back translate a protein alignment using the sequences present on the current alignment. <br />
 Current alignment should contain the original sequences which were translated to protein sequences, <b> without any gap.</b>  <a href="#a781fd0f2ae6a9bda2802ffd28372bfaa">More...</a><br /></td></tr>
<tr class="separator:a781fd0f2ae6a9bda2802ffd28372bfaa"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae8501233fe4f4ef6bfe30c99c744c7c2"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#ae8501233fe4f4ef6bfe30c99c744c7c2">checkCorrespondence</a> (string *names, int *lenghts, int totalInputSequences, int multiple)</td></tr>
<tr class="memdesc:ae8501233fe4f4ef6bfe30c99c744c7c2"><td class="mdescLeft">&#160;</td><td class="mdescRight">Function to check CDS file.<br />
.  <a href="#ae8501233fe4f4ef6bfe30c99c744c7c2">More...</a><br /></td></tr>
<tr class="separator:ae8501233fe4f4ef6bfe30c99c744c7c2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3de39c62c08a265178f1fa7df76c3ab7"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a3de39c62c08a265178f1fa7df76c3ab7">fillNewDataStructure</a> (string *newMatrix, string *newNames)</td></tr>
<tr class="memdesc:a3de39c62c08a265178f1fa7df76c3ab7"><td class="mdescLeft">&#160;</td><td class="mdescRight">Method that copies information of the alignment to the pointers given.  <a href="#a3de39c62c08a265178f1fa7df76c3ab7">More...</a><br /></td></tr>
<tr class="separator:a3de39c62c08a265178f1fa7df76c3ab7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad916722e4ee892ee9f5cf85f0bdb80aa"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#ad916722e4ee892ee9f5cf85f0bdb80aa">calculateColIdentity</a> (float *columnIdentity)</td></tr>
<tr class="memdesc:ad916722e4ee892ee9f5cf85f0bdb80aa"><td class="mdescLeft">&#160;</td><td class="mdescRight">Method that calculates the columns identity value.<br />
This is, the frequency of the most present element in the column, being residue, indetermination and gap allowed.  <a href="#ad916722e4ee892ee9f5cf85f0bdb80aa">More...</a><br /></td></tr>
<tr class="separator:ad916722e4ee892ee9f5cf85f0bdb80aa"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a98c9b5bd9a859600ae8a317d368c85b3"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a98c9b5bd9a859600ae8a317d368c85b3">printColumnsIdentity_DescriptiveStats</a> ()</td></tr>
<tr class="memdesc:a98c9b5bd9a859600ae8a317d368c85b3"><td class="mdescLeft">&#160;</td><td class="mdescRight">Method to print Identity stats: Min, Max, Avg and Std.  <a href="#a98c9b5bd9a859600ae8a317d368c85b3">More...</a><br /></td></tr>
<tr class="separator:a98c9b5bd9a859600ae8a317d368c85b3"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a29f7f52115fd4ecf13120e54c71b1a1a"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a29f7f52115fd4ecf13120e54c71b1a1a">setKeepSequencesFlag</a> (bool newFlagValue)</td></tr>
<tr class="memdesc:a29f7f52115fd4ecf13120e54c71b1a1a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Keep Sequences setter.  <a href="#a29f7f52115fd4ecf13120e54c71b1a1a">More...</a><br /></td></tr>
<tr class="separator:a29f7f52115fd4ecf13120e54c71b1a1a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a66015efb0c5b9cd596d02d3c736cffe3"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a66015efb0c5b9cd596d02d3c736cffe3">setKeepSeqsHeaderFlag</a> (bool newFlagValue)</td></tr>
<tr class="memdesc:a66015efb0c5b9cd596d02d3c736cffe3"><td class="mdescLeft">&#160;</td><td class="mdescRight">Keep Header setter.  <a href="#a66015efb0c5b9cd596d02d3c736cffe3">More...</a><br /></td></tr>
<tr class="separator:a66015efb0c5b9cd596d02d3c736cffe3"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3b75055480c60265ea57ca557213fe06"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a3b75055480c60265ea57ca557213fe06">printAlignmentInfo</a> (ostream &amp;output)</td></tr>
<tr class="memdesc:a3b75055480c60265ea57ca557213fe06"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print information about sequences number, average sequence length, maximum and minimum sequences length, etc.  <a href="#a3b75055480c60265ea57ca557213fe06">More...</a><br /></td></tr>
<tr class="separator:a3b75055480c60265ea57ca557213fe06"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad50ba126247276cd0bf89119e439c483"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#ad50ba126247276cd0bf89119e439c483">prepareCodingSequence</a> (bool splitByStopCodon, bool ignoreStopCodon, <a class="el" href="classnew_alignment.html">newAlignment</a> *proteinAlignment)</td></tr>
<tr class="memdesc:ad50ba126247276cd0bf89119e439c483"><td class="mdescLeft">&#160;</td><td class="mdescRight">Method to check if the CDS file is correct.<br />
Based on nature of residues: DNA/RNA (Most of the residues)<br />
There is no gaps on the whole dataset.<br />
Each sequence is a multiple of 3.<br />
<br />
It will also remove or split stop codons depending on the flags passed.  <a href="#ad50ba126247276cd0bf89119e439c483">More...</a><br /></td></tr>
<tr class="separator:ad50ba126247276cd0bf89119e439c483"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2607115c839209324dca54d70fc8708d"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a2607115c839209324dca54d70fc8708d">alignmentSummaryHTML</a> (<a class="el" href="classnew_alignment.html">newAlignment</a> &amp;_trimmedAlignment, char *destFile)</td></tr>
<tr class="memdesc:a2607115c839209324dca54d70fc8708d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Method to report the trimming results in HTML.<br />
Outputs an HTML file that shows visually what has been done to the alignment (removed sequences or residues) and the stats used to do so.  <a href="#a2607115c839209324dca54d70fc8708d">More...</a><br /></td></tr>
<tr class="separator:a2607115c839209324dca54d70fc8708d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4c739b8951d20fbe59251041c1131bd3"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a4c739b8951d20fbe59251041c1131bd3">alignmentSummarySVG</a> (<a class="el" href="classnew_alignment.html">newAlignment</a> &amp;_trimmedAlignment, char *destFile, int blocks)</td></tr>
<tr class="memdesc:a4c739b8951d20fbe59251041c1131bd3"><td class="mdescLeft">&#160;</td><td class="mdescRight">Method to report the trimming results in SVG.<br />
Outputs a SVG file that shows visually what has been done to the alignment (removed sequences or residues) and the stats used to do so.  <a href="#a4c739b8951d20fbe59251041c1131bd3">More...</a><br /></td></tr>
<tr class="separator:a4c739b8951d20fbe59251041c1131bd3"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a26d10a23f8f71f95657a05e88a9a8fff"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a26d10a23f8f71f95657a05e88a9a8fff">alignmentColourHTML</a> (ostream &amp;file)</td></tr>
<tr class="memdesc:a26d10a23f8f71f95657a05e88a9a8fff"><td class="mdescLeft">&#160;</td><td class="mdescRight">Method that saves a clustal-color based visualization of the alignment.<br />
It doesn't report any trimming result, meaning this only shows information of an alignment.  <a href="#a26d10a23f8f71f95657a05e88a9a8fff">More...</a><br /></td></tr>
<tr class="separator:a26d10a23f8f71f95657a05e88a9a8fff"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a52710ff330519de8a4e41b27e7e55958"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a52710ff330519de8a4e41b27e7e55958">updateSequencesAndResiduesNums</a> (bool countSequences=true, bool countResidues=true)</td></tr>
<tr class="memdesc:a52710ff330519de8a4e41b27e7e55958"><td class="mdescLeft">&#160;</td><td class="mdescRight">Updates the sequence number and residue number based on saveResidues and saveSequences.  <a href="#a52710ff330519de8a4e41b27e7e55958">More...</a><br /></td></tr>
<tr class="separator:a52710ff330519de8a4e41b27e7e55958"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-attribs"></a>
Public Attributes</h2></td></tr>
<tr class="memitem:a2417feb815493866d0137ae5f5cd1028"><td class="memItemLeft" align="right" valign="top"><a id="a2417feb815493866d0137ae5f5cd1028"></a>
<a class="el" href="class_cleaner.html">Cleaner</a> *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a2417feb815493866d0137ae5f5cd1028">Cleaning</a> = nullptr</td></tr>
<tr class="memdesc:a2417feb815493866d0137ae5f5cd1028"><td class="mdescLeft">&#160;</td><td class="mdescRight">Trimming submodule.<br />
It contains methods and variables related to trimming. <br /></td></tr>
<tr class="separator:a2417feb815493866d0137ae5f5cd1028"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6324267e64b68bb85d993cf602d008eb"><td class="memItemLeft" align="right" valign="top"><a id="a6324267e64b68bb85d993cf602d008eb"></a>
<a class="el" href="class_statistics_manager.html">StatisticsManager</a> *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a6324267e64b68bb85d993cf602d008eb">Statistics</a> = nullptr</td></tr>
<tr class="memdesc:a6324267e64b68bb85d993cf602d008eb"><td class="mdescLeft">&#160;</td><td class="mdescRight">Statistics submodule.<br />
It contains methods and variables related to Statistics calculation and reporting. <br /></td></tr>
<tr class="separator:a6324267e64b68bb85d993cf602d008eb"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8b328dfd8e7d3304cb144f2c2114ab5c"><td class="memItemLeft" align="right" valign="top"><a id="a8b328dfd8e7d3304cb144f2c2114ab5c"></a>
<a class="el" href="classsequences_matrix.html">sequencesMatrix</a> *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a8b328dfd8e7d3304cb144f2c2114ab5c">SequencesMatrix</a> = nullptr</td></tr>
<tr class="memdesc:a8b328dfd8e7d3304cb144f2c2114ab5c"><td class="mdescLeft">&#160;</td><td class="mdescRight">SequencesMatrix submodule<br />
 This helper module stores the sequences without gaps. Useful when comparing MSA. <br /></td></tr>
<tr class="separator:a8b328dfd8e7d3304cb144f2c2114ab5c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a94afba152da4b05735f429781cd3da57"><td class="memItemLeft" align="right" valign="top"><a id="a94afba152da4b05735f429781cd3da57"></a>
int *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a94afba152da4b05735f429781cd3da57">SeqRef</a> = nullptr</td></tr>
<tr class="memdesc:a94afba152da4b05735f429781cd3da57"><td class="mdescLeft">&#160;</td><td class="mdescRight">Pointer to keep a count of how many alignments depend on the same shared-sequences.<br />
 Increased on each trim step, and automatically reduced on each alignment destroy.<br />
 If it arrives to 0, the last alignment being destroyed will also take care of releasing memory for shared information. <br /></td></tr>
<tr class="separator:a94afba152da4b05735f429781cd3da57"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6cdace1a82bc124173c80c978e42ac79"><td class="memItemLeft" align="right" valign="top"><a id="a6cdace1a82bc124173c80c978e42ac79"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a6cdace1a82bc124173c80c978e42ac79">originalSequenNumber</a> = 0</td></tr>
<tr class="memdesc:a6cdace1a82bc124173c80c978e42ac79"><td class="mdescLeft">&#160;</td><td class="mdescRight">Number of sequences the alignment had when it was loaded. <br /></td></tr>
<tr class="separator:a6cdace1a82bc124173c80c978e42ac79"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4a6b47de8bef3e783f30548dfa48d84c"><td class="memItemLeft" align="right" valign="top"><a id="a4a6b47de8bef3e783f30548dfa48d84c"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a4a6b47de8bef3e783f30548dfa48d84c">sequenNumber</a> = 0</td></tr>
<tr class="memdesc:a4a6b47de8bef3e783f30548dfa48d84c"><td class="mdescLeft">&#160;</td><td class="mdescRight">Number of sequences present on the alignment. <br /></td></tr>
<tr class="separator:a4a6b47de8bef3e783f30548dfa48d84c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4250245146147c153df58a3334d7fb9d"><td class="memItemLeft" align="right" valign="top"><a id="a4250245146147c153df58a3334d7fb9d"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a4250245146147c153df58a3334d7fb9d">originalResidNumber</a> = 0</td></tr>
<tr class="memdesc:a4250245146147c153df58a3334d7fb9d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Number of residues the alignment had when it was loaded. <br /></td></tr>
<tr class="separator:a4250245146147c153df58a3334d7fb9d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aff4e7761ec80bea74cdf28c0269b27db"><td class="memItemLeft" align="right" valign="top"><a id="aff4e7761ec80bea74cdf28c0269b27db"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#aff4e7761ec80bea74cdf28c0269b27db">residNumber</a> = 0</td></tr>
<tr class="memdesc:aff4e7761ec80bea74cdf28c0269b27db"><td class="mdescLeft">&#160;</td><td class="mdescRight">Number of residues present on the alignment if it is aligned. <br /></td></tr>
<tr class="separator:aff4e7761ec80bea74cdf28c0269b27db"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab6b16a670e278cd5e283573665fa450f"><td class="memItemLeft" align="right" valign="top"><a id="ab6b16a670e278cd5e283573665fa450f"></a>
bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#ab6b16a670e278cd5e283573665fa450f">isAligned</a> = false</td></tr>
<tr class="memdesc:ab6b16a670e278cd5e283573665fa450f"><td class="mdescLeft">&#160;</td><td class="mdescRight">Flag that indicates if all sequences on the alignment have the same length (Including gaps). <br /></td></tr>
<tr class="separator:ab6b16a670e278cd5e283573665fa450f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a23f06ad268f0e720bf3d5699a4cb7225"><td class="memItemLeft" align="right" valign="top"><a id="a23f06ad268f0e720bf3d5699a4cb7225"></a>
string *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a23f06ad268f0e720bf3d5699a4cb7225">sequences</a> = nullptr</td></tr>
<tr class="memdesc:a23f06ad268f0e720bf3d5699a4cb7225"><td class="mdescLeft">&#160;</td><td class="mdescRight">Vector containing the sequences. <br /></td></tr>
<tr class="separator:a23f06ad268f0e720bf3d5699a4cb7225"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa4bf60277ddf15e935a15b22becd6ad9"><td class="memItemLeft" align="right" valign="top"><a id="aa4bf60277ddf15e935a15b22becd6ad9"></a>
string *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#aa4bf60277ddf15e935a15b22becd6ad9">seqsName</a> = nullptr</td></tr>
<tr class="memdesc:aa4bf60277ddf15e935a15b22becd6ad9"><td class="mdescLeft">&#160;</td><td class="mdescRight">String vector containing the sequences names. <br /></td></tr>
<tr class="separator:aa4bf60277ddf15e935a15b22becd6ad9"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad4e91d14819ffa8e6e57ae7c23d35327"><td class="memItemLeft" align="right" valign="top"><a id="ad4e91d14819ffa8e6e57ae7c23d35327"></a>
string *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#ad4e91d14819ffa8e6e57ae7c23d35327">seqsInfo</a> = nullptr</td></tr>
<tr class="memdesc:ad4e91d14819ffa8e6e57ae7c23d35327"><td class="mdescLeft">&#160;</td><td class="mdescRight">Vector containing extra information about each sequence. <br /></td></tr>
<tr class="separator:ad4e91d14819ffa8e6e57ae7c23d35327"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6c9904f7a01e200bd167f3cf736af4d6"><td class="memItemLeft" align="right" valign="top"><a id="a6c9904f7a01e200bd167f3cf736af4d6"></a>
string&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a6c9904f7a01e200bd167f3cf736af4d6">filename</a></td></tr>
<tr class="memdesc:a6c9904f7a01e200bd167f3cf736af4d6"><td class="mdescLeft">&#160;</td><td class="mdescRight">Filename where this alignment was loaded from. <br /></td></tr>
<tr class="separator:a6c9904f7a01e200bd167f3cf736af4d6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a34b1ee01d1030ce737034b35f3631e8a"><td class="memItemLeft" align="right" valign="top"><a id="a34b1ee01d1030ce737034b35f3631e8a"></a>
string&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a34b1ee01d1030ce737034b35f3631e8a">aligInfo</a></td></tr>
<tr class="memdesc:a34b1ee01d1030ce737034b35f3631e8a"><td class="mdescLeft">&#160;</td><td class="mdescRight">String containing information of the alignment as a whole. <br /></td></tr>
<tr class="separator:a34b1ee01d1030ce737034b35f3631e8a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a792877eafe085fe6aa5b267901fe5be9"><td class="memItemLeft" align="right" valign="top">float **&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a792877eafe085fe6aa5b267901fe5be9">identities</a> = nullptr</td></tr>
<tr class="memdesc:a792877eafe085fe6aa5b267901fe5be9"><td class="mdescLeft">&#160;</td><td class="mdescRight">2D Matrix that represents how much each sequence resembles other sequences in the same MSA  <a href="#a792877eafe085fe6aa5b267901fe5be9">More...</a><br /></td></tr>
<tr class="separator:a792877eafe085fe6aa5b267901fe5be9"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aee3815b0ff8076b8929e4128414ef320"><td class="memItemLeft" align="right" valign="top"><a id="aee3815b0ff8076b8929e4128414ef320"></a>
float **&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#aee3815b0ff8076b8929e4128414ef320">overlaps</a> = nullptr</td></tr>
<tr class="memdesc:aee3815b0ff8076b8929e4128414ef320"><td class="mdescLeft">&#160;</td><td class="mdescRight">Matrix that stores the overlap between sequences of an alignment. See <a class="el" href="classnew_alignment.html#ae157ddd55e9f41d809fb664116619802" title="Calculates overlap between sequences. ">newAlignment::calculateSeqOverlap</a>. <br /></td></tr>
<tr class="separator:aee3815b0ff8076b8929e4128414ef320"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5239bd4e742631b58efc7f09dd67f165"><td class="memItemLeft" align="right" valign="top"><a id="a5239bd4e742631b58efc7f09dd67f165"></a>
int *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a5239bd4e742631b58efc7f09dd67f165">saveResidues</a> = nullptr</td></tr>
<tr class="memdesc:a5239bd4e742631b58efc7f09dd67f165"><td class="mdescLeft">&#160;</td><td class="mdescRight">Vector containing which residues will be kept/removed.<br />
-1 Indicates a 'removed/toRemove'. Otherwise residue/column will be kept. <br /></td></tr>
<tr class="separator:a5239bd4e742631b58efc7f09dd67f165"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a76d057ddcebc7c88dd7f9e3ff0ce95bf"><td class="memItemLeft" align="right" valign="top"><a id="a76d057ddcebc7c88dd7f9e3ff0ce95bf"></a>
int *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a76d057ddcebc7c88dd7f9e3ff0ce95bf">saveSequences</a> = nullptr</td></tr>
<tr class="memdesc:a76d057ddcebc7c88dd7f9e3ff0ce95bf"><td class="mdescLeft">&#160;</td><td class="mdescRight">Vector containing which sequences will be kept/removed.<br />
-1 Indicates a 'removed/toRemove'. Otherwise sequence will be kept. <br /></td></tr>
<tr class="separator:a76d057ddcebc7c88dd7f9e3ff0ce95bf"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pri-attribs"></a>
Private Attributes</h2></td></tr>
<tr class="memitem:ada2fb94cccdf0760c22dab39d278680c"><td class="memItemLeft" align="right" valign="top"><a id="ada2fb94cccdf0760c22dab39d278680c"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#ada2fb94cccdf0760c22dab39d278680c">dataType</a> = 0</td></tr>
<tr class="memdesc:ada2fb94cccdf0760c22dab39d278680c"><td class="mdescLeft">&#160;</td><td class="mdescRight">Flag variable that represents the type of the alignment. <br /></td></tr>
<tr class="separator:ada2fb94cccdf0760c22dab39d278680c"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p>Class containing an alignment<br />
 This class stores the alignment sequences with it's names, residues and extra information.<br />
 It contains multiple methods regarding the sequences.<br />
 It also contains submodules that provide methods for<b> Calculating statistics</b>, <b> Trim the alignment</b> and<b> Printing sequences information</b>. </p>

<p class="definition">Definition at line <a class="el" href="new_alignment_8h_source.html#l00052">52</a> of file <a class="el" href="new_alignment_8h_source.html">newAlignment.h</a>.</p>
</div><h2 class="groupheader">Member Function Documentation</h2>
<a id="a26d10a23f8f71f95657a05e88a9a8fff"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a26d10a23f8f71f95657a05e88a9a8fff">&#9670;&nbsp;</a></span>alignmentColourHTML()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool newAlignment::alignmentColourHTML </td>
          <td>(</td>
          <td class="paramtype">ostream &amp;&#160;</td>
          <td class="paramname"><em>file</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Method that saves a clustal-color based visualization of the alignment.<br />
It doesn't report any trimming result, meaning this only shows information of an alignment. </p>
<dl class="section warning"><dt>Warning</dt><dd>Not In Use </dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">file</td><td>Where to save the visualization. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd><b>True</b> if everything went ok. <br />
 <b>False</b> if alignment is not aligned. </dd></dl>

<p class="definition">Definition at line <a class="el" href="new_alignment_8cpp_source.html#l02044">2044</a> of file <a class="el" href="new_alignment_8cpp_source.html">newAlignment.cpp</a>.</p>

<p class="reference">References <a class="el" href="utils_8cpp_source.html#l00083">utils::max()</a>, and <a class="el" href="reportsystem_8cpp_source.html#l00375">__internalReport::__reportSystem::report()</a>.</p>
<div id="dynsection-0" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-0-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-0-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-0-content" class="dyncontent" style="display:none;">
<div class="center"><iframe scrolling="no" frameborder="0" src="classnew_alignment_a26d10a23f8f71f95657a05e88a9a8fff_cgraph.svg" width="456" height="104"><p><b>This browser is not able to show SVG: try Firefox, Chrome, Safari, or Opera instead.</b></p></iframe></div>
</div>

</div>
</div>
<a id="a2607115c839209324dca54d70fc8708d"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a2607115c839209324dca54d70fc8708d">&#9670;&nbsp;</a></span>alignmentSummaryHTML()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool newAlignment::alignmentSummaryHTML </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="classnew_alignment.html">newAlignment</a> &amp;&#160;</td>
          <td class="paramname"><em>_trimmedAlignment</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char *&#160;</td>
          <td class="paramname"><em>destFile</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Method to report the trimming results in HTML.<br />
Outputs an HTML file that shows visually what has been done to the alignment (removed sequences or residues) and the stats used to do so. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">_trimmedAlignment</td><td>Trimmed alignment to compare with. </td></tr>
    <tr><td class="paramname">destFile</td><td>Filename where to save the results. </td></tr>
    <tr><td class="paramname">consValues</td><td>Vector containing the consistency values of columns.<br />
The consistency may be a null pointer, meaning we don't want to report consistency values.<br />
This may happen if we didn't calculated the consistency values or used them to choose the best alignment among a set, but not used to trim the alignment itself. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd><b>True</b> if correct. <br />
 <b>False</b> if file couldn't be open or alignment is not aligned. </dd></dl>

<p class="definition">Definition at line <a class="el" href="new_alignment_8cpp_source.html#l01145">1145</a> of file <a class="el" href="new_alignment_8cpp_source.html">newAlignment.cpp</a>.</p>

<p class="reference">References <a class="el" href="statistics_conservation_8cpp_source.html#l00314">statisticsConservation::getMdkWindowedVector()</a>, <a class="el" href="utils_8cpp_source.html#l00083">utils::max()</a>, <a class="el" href="reportsystem_8cpp_source.html#l00375">__internalReport::__reportSystem::report()</a>, <a class="el" href="new_alignment_8h_source.html#l00075">residNumber</a>, <a class="el" href="new_alignment_8h_source.html#l00071">sequenNumber</a>, and <a class="el" href="new_alignment_8h_source.html#l00061">Statistics</a>.</p>
<div id="dynsection-1" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-1-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-1-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-1-content" class="dyncontent" style="display:none;">
<div class="center"><iframe scrolling="no" frameborder="0" src="classnew_alignment_a2607115c839209324dca54d70fc8708d_cgraph.svg" width="1592" height="220"><p><b>This browser is not able to show SVG: try Firefox, Chrome, Safari, or Opera instead.</b></p></iframe></div>
</div>

</div>
</div>
<a id="a4c739b8951d20fbe59251041c1131bd3"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a4c739b8951d20fbe59251041c1131bd3">&#9670;&nbsp;</a></span>alignmentSummarySVG()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool newAlignment::alignmentSummarySVG </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="classnew_alignment.html">newAlignment</a> &amp;&#160;</td>
          <td class="paramname"><em>_trimmedAlignment</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char *&#160;</td>
          <td class="paramname"><em>destFile</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>blocks</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Method to report the trimming results in SVG.<br />
Outputs a SVG file that shows visually what has been done to the alignment (removed sequences or residues) and the stats used to do so. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">_trimmedAlignment</td><td>Trimmed alignment obtained from the object calling this function. </td></tr>
    <tr><td class="paramname">destFile</td><td>Filename where to save the results. </td></tr>
    <tr><td class="paramname">consValues</td><td>Vector containing the consistency values of columns. <br />
 The consistency may be a null pointer, meaning we don't want to report consistency values.<br />
This may happen if we didn't calculated the consistency values or used them to choose the best alignment among a set, but not used to trim the alignment itself. </td></tr>
    <tr><td class="paramname">blocks</td><td>Size in residues to report. 120 by default. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd><b>True</b> if everything went ok. <br />
 <b>False</b> if file couldn't be open or alignment is not aligned. </dd></dl>

<p class="definition">Definition at line <a class="el" href="new_alignment_8cpp_source.html#l01409">1409</a> of file <a class="el" href="new_alignment_8cpp_source.html">newAlignment.cpp</a>.</p>

<p class="reference">References <a class="el" href="utils_8cpp_source.html#l00623">utils::determineColor()</a>, <a class="el" href="utils_8cpp_source.html#l00083">utils::max()</a>, <a class="el" href="reportsystem_8cpp_source.html#l00375">__internalReport::__reportSystem::report()</a>, <a class="el" href="new_alignment_8h_source.html#l00075">residNumber</a>, <a class="el" href="new_alignment_8h_source.html#l00096">saveResidues</a>, <a class="el" href="new_alignment_8h_source.html#l00097">saveSequences</a>, and <a class="el" href="new_alignment_8h_source.html#l00071">sequenNumber</a>.</p>
<div id="dynsection-2" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-2-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-2-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-2-content" class="dyncontent" style="display:none;">
<div class="center"><iframe scrolling="no" frameborder="0" src="classnew_alignment_a4c739b8951d20fbe59251041c1131bd3_cgraph.svg" width="652" height="155"><p><b>This browser is not able to show SVG: try Firefox, Chrome, Safari, or Opera instead.</b></p></iframe></div>
</div>

</div>
</div>
<a id="ad916722e4ee892ee9f5cf85f0bdb80aa"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ad916722e4ee892ee9f5cf85f0bdb80aa">&#9670;&nbsp;</a></span>calculateColIdentity()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void newAlignment::calculateColIdentity </td>
          <td>(</td>
          <td class="paramtype">float *&#160;</td>
          <td class="paramname"><em>columnIdentity</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Method that calculates the columns identity value.<br />
This is, the frequency of the most present element in the column, being residue, indetermination and gap allowed. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir">[out]</td><td class="paramname">columnIdentity</td><td>Vector to fill with identities for each column. </td></tr>
  </table>
  </dd>
</dl>

<p class="definition">Definition at line <a class="el" href="new_alignment_8cpp_source.html#l01054">1054</a> of file <a class="el" href="new_alignment_8cpp_source.html">newAlignment.cpp</a>.</p>

<p class="reference">References <a class="el" href="defines_8h_source.html#l00086">AA</a>.</p>

</div>
</div>
<a id="adaeafc911344fc10069f40d9fc41303f"></a>
<h2 class="memtitle"><span class="permalink"><a href="#adaeafc911344fc10069f40d9fc41303f">&#9670;&nbsp;</a></span>calculateRelaxedSeqIdentity()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void newAlignment::calculateRelaxedSeqIdentity </td>
          <td>(</td>
          <td class="paramtype">void&#160;</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Lightweight version of <a class="el" href="classnew_alignment.html#aa9178bd574db2b6171c0bb09133bdb94" title="Calculates identity between sequences. ">newAlignment::calculateSeqIdentity</a> for use in huge alignments. </p>
<dl class="section warning"><dt>Warning</dt><dd>Not In Use </dd></dl>

<p class="definition">Definition at line <a class="el" href="new_alignment_8cpp_source.html#l00346">346</a> of file <a class="el" href="new_alignment_8cpp_source.html">newAlignment.cpp</a>.</p>

</div>
</div>
<a id="aa9178bd574db2b6171c0bb09133bdb94"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aa9178bd574db2b6171c0bb09133bdb94">&#9670;&nbsp;</a></span>calculateSeqIdentity()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void newAlignment::calculateSeqIdentity </td>
          <td>(</td>
          <td class="paramtype">void&#160;</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Calculates identity between sequences. </p>
<dl class="section warning"><dt>Warning</dt><dd>Not In Use </dd></dl>

<p class="definition">Definition at line <a class="el" href="new_alignment_8cpp_source.html#l00302">302</a> of file <a class="el" href="new_alignment_8cpp_source.html">newAlignment.cpp</a>.</p>

<p class="reference">References <a class="el" href="defines_8h_source.html#l00086">AA</a>.</p>

</div>
</div>
<a id="ae8501233fe4f4ef6bfe30c99c744c7c2"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ae8501233fe4f4ef6bfe30c99c744c7c2">&#9670;&nbsp;</a></span>checkCorrespondence()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool newAlignment::checkCorrespondence </td>
          <td>(</td>
          <td class="paramtype">string *&#160;</td>
          <td class="paramname"><em>names</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int *&#160;</td>
          <td class="paramname"><em>lenghts</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>totalInputSequences</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>multiple</em> = <code>1</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Function to check CDS file.<br />
. </p>
<ul>
<li>It checks if sequences of input alignment are all present on the CDS file.</li>
<li>If nucleotide sequence is larger than protein sequence length * 3, warns about it and cuts the nucleotide sequence</li>
<li>If sequence has indetermination symbols, it warns about it.</li>
<li>If nucleotide sequence is smaller than protein sequence * 3, it adds some 'N' at the end of the nucleotide sequence. <dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">names</td><td>Vector containing the names to check. </td></tr>
    <tr><td class="paramname">lenghts</td><td>Vector containing the length of each sequence. </td></tr>
    <tr><td class="paramname">totalInputSequences</td><td>Number of sequences present. </td></tr>
    <tr><td class="paramname">multiple</td><td>Multiplies the length of each sequence by this number. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd><b>True</b> if all went right.<br />
 <b>False</b> if there is a sequence in the alignment but not in the names vector. </dd></dl>
</li>
</ul>

<p class="definition">Definition at line <a class="el" href="new_alignment_8cpp_source.html#l00701">701</a> of file <a class="el" href="new_alignment_8cpp_source.html">newAlignment.cpp</a>.</p>

<p class="reference">References <a class="el" href="utils_8cpp_source.html#l00101">utils::min()</a>, <a class="el" href="utils_8cpp_source.html#l00493">utils::removeCharacter()</a>, and <a class="el" href="reportsystem_8cpp_source.html#l00375">__internalReport::__reportSystem::report()</a>.</p>

<p class="reference">Referenced by <a class="el" href="trimal_manager_8cpp_source.html#l01274">trimAlManager::check_backtranslation_infile_names_corresponde()</a>.</p>
<div id="dynsection-3" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-3-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-3-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-3-content" class="dyncontent" style="display:none;">
<div class="center"><iframe scrolling="no" frameborder="0" src="classnew_alignment_ae8501233fe4f4ef6bfe30c99c744c7c2_cgraph.svg" width="456" height="155"><p><b>This browser is not able to show SVG: try Firefox, Chrome, Safari, or Opera instead.</b></p></iframe></div>
</div>
<div id="dynsection-4" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-4-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-4-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-4-content" class="dyncontent" style="display:none;">
<div class="center"><iframe scrolling="no" frameborder="0" src="classnew_alignment_ae8501233fe4f4ef6bfe30c99c744c7c2_icgraph.svg" width="920" height="67"><p><b>This browser is not able to show SVG: try Firefox, Chrome, Safari, or Opera instead.</b></p></iframe></div>
</div>

</div>
</div>
<a id="acbeb320f0f70222f81a906ccfc33a016"></a>
<h2 class="memtitle"><span class="permalink"><a href="#acbeb320f0f70222f81a906ccfc33a016">&#9670;&nbsp;</a></span>fillMatrices()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool newAlignment::fillMatrices </td>
          <td>(</td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>aligned</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Method to initialize matrices and check if the sequences have been correctly loaded and are free of errors<br />
It checks if sequences contain unknown characters, sets the isAligned flag and initializes 'saveResidues' and 'saveSequences', depending on the sizes of the sequences (whether or not they have the same length). </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">aligned</td><td>Flag to make the method check if the alignment is aligned or not. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section note"><dt>Note</dt><dd>Even with the aligned flag set to false, if the alignment is aligned it will initialize the variables 'saveResidues' and 'saveSequences', allowing the alignment to be trimmed. </dd></dl>
<dl class="section return"><dt>Returns</dt><dd><b>True</b> if the alignment information was ok. <br />
 <b>False</b> if there was a problem.<br />
This could happen if the sequences contain any unknown character or if aligned flag is set up to true and the sequences have variable sizes. </dd></dl>

<p class="definition">Definition at line <a class="el" href="new_alignment_8cpp_source.html#l00779">779</a> of file <a class="el" href="new_alignment_8cpp_source.html">newAlignment.cpp</a>.</p>

<p class="reference">References <a class="el" href="reportsystem_8cpp_source.html#l00375">__internalReport::__reportSystem::report()</a>.</p>
<div id="dynsection-5" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-5-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-5-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-5-content" class="dyncontent" style="display:none;">
<div class="center"><iframe scrolling="no" frameborder="0" src="classnew_alignment_acbeb320f0f70222f81a906ccfc33a016_cgraph.svg" width="388" height="52"><p><b>This browser is not able to show SVG: try Firefox, Chrome, Safari, or Opera instead.</b></p></iframe></div>
</div>

</div>
</div>
<a id="a3de39c62c08a265178f1fa7df76c3ab7"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a3de39c62c08a265178f1fa7df76c3ab7">&#9670;&nbsp;</a></span>fillNewDataStructure()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void newAlignment::fillNewDataStructure </td>
          <td>(</td>
          <td class="paramtype">string *&#160;</td>
          <td class="paramname"><em>newMatrix</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">string *&#160;</td>
          <td class="paramname"><em>newNames</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Method that copies information of the alignment to the pointers given. </p>
<dl class="section warning"><dt>Warning</dt><dd>Not In Use </dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir">[out]</td><td class="paramname">newMatrix</td><td>Sequences vector to fill. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">newNames</td><td>Names vector to fill. </td></tr>
  </table>
  </dd>
</dl>

<p class="definition">Definition at line <a class="el" href="new_alignment_8cpp_source.html#l00503">503</a> of file <a class="el" href="new_alignment_8cpp_source.html">newAlignment.cpp</a>.</p>

</div>
</div>
<a id="a670d5ef1ced9ed6041c3f6afdbd2cc02"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a670d5ef1ced9ed6041c3f6afdbd2cc02">&#9670;&nbsp;</a></span>getAlignmentType()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int newAlignment::getAlignmentType </td>
          <td>(</td>
          <td class="paramtype">void&#160;</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Alignment type getter. <br />
 See <a class="el" href="">SequenceTypes.</a></p>
<dl class="section return"><dt>Returns</dt><dd>Int representing the alignment type.<br />
</dd></dl>

<p class="definition">Definition at line <a class="el" href="new_alignment_8cpp_source.html#l00473">473</a> of file <a class="el" href="new_alignment_8cpp_source.html">newAlignment.cpp</a>.</p>

<p class="reference">References <a class="el" href="utils_8cpp_source.html#l00506">utils::checkAlignmentType()</a>.</p>

<p class="reference">Referenced by <a class="el" href="statistics_conservation_8cpp_source.html#l00092">statisticsConservation::calculateMatrixIdentity()</a>, <a class="el" href="_cleaner_8cpp_source.html#l01415">Cleaner::calculateSeqIdentity()</a>, <a class="el" href="_cleaner_8cpp_source.html#l00842">Cleaner::calculateSpuriousVector()</a>, <a class="el" href="statistics_conservation_8cpp_source.html#l00155">statisticsConservation::calculateVectors()</a>, <a class="el" href="trimal_manager_8cpp_source.html#l01249">trimAlManager::check_coding_sequences_type()</a>, and <a class="el" href="trimal_manager_8cpp_source.html#l01538">trimAlManager::create_or_use_similarity_matrix()</a>.</p>
<div id="dynsection-6" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-6-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-6-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-6-content" class="dyncontent" style="display:none;">
<div class="center"><iframe scrolling="no" frameborder="0" src="classnew_alignment_a670d5ef1ced9ed6041c3f6afdbd2cc02_cgraph.svg" width="452" height="38"><p><b>This browser is not able to show SVG: try Firefox, Chrome, Safari, or Opera instead.</b></p></iframe></div>
</div>
<div id="dynsection-7" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-7-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-7-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-7-content" class="dyncontent" style="display:none;">
<div class="center"><iframe scrolling="no" frameborder="0" src="classnew_alignment_a670d5ef1ced9ed6041c3f6afdbd2cc02_icgraph.svg" width="2495" height="435"><p><b>This browser is not able to show SVG: try Firefox, Chrome, Safari, or Opera instead.</b></p></iframe></div>
</div>

</div>
</div>
<a id="a4761e21288fc15f3b86cf80dd4c6c91b"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a4761e21288fc15f3b86cf80dd4c6c91b">&#9670;&nbsp;</a></span>getBlockSize()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int newAlignment::getBlockSize </td>
          <td>(</td>
          <td class="paramtype">void&#160;</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>BlockSize getter. </p>
<dl class="section warning"><dt>Warning</dt><dd>Not In Use </dd></dl>
<dl class="section return"><dt>Returns</dt><dd>New BlockSize value. </dd></dl>

<p class="definition">Definition at line <a class="el" href="new_alignment_8cpp_source.html#l00295">295</a> of file <a class="el" href="new_alignment_8cpp_source.html">newAlignment.cpp</a>.</p>

</div>
</div>
<a id="ace3be92e3229fc241557597090977ba5"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ace3be92e3229fc241557597090977ba5">&#9670;&nbsp;</a></span>getCorrespResidues()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int * newAlignment::getCorrespResidues </td>
          <td>(</td>
          <td class="paramtype">void&#160;</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Method that returns the correspondence between old and new alignment residues. <br />
 See <a class="el" href="classnew_alignment.html#a5239bd4e742631b58efc7f09dd67f165" title="Vector containing which residues will be kept/removed. -1 Indicates a &#39;removed/toRemove&#39;. Otherwise residue/column will be kept. ">newAlignment::saveResidues</a>. </p>
<dl class="section warning"><dt>Warning</dt><dd>Not In Use </dd></dl>
<dl class="section return"><dt>Returns</dt><dd>Vector that represents which residues are kept (!= -1) and which are rejected (== -1) </dd></dl>

<p class="definition">Definition at line <a class="el" href="new_alignment_8cpp_source.html#l00482">482</a> of file <a class="el" href="new_alignment_8cpp_source.html">newAlignment.cpp</a>.</p>

</div>
</div>
<a id="a0c3cd77c8a6b118efb84b577bb4ab1ce"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a0c3cd77c8a6b118efb84b577bb4ab1ce">&#9670;&nbsp;</a></span>getCorrespSequences()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int * newAlignment::getCorrespSequences </td>
          <td>(</td>
          <td class="paramtype">void&#160;</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Method that returns the correspondence between old and new alignment sequences. <br />
 See <a class="el" href="classnew_alignment.html#a76d057ddcebc7c88dd7f9e3ff0ce95bf" title="Vector containing which sequences will be kept/removed. -1 Indicates a &#39;removed/toRemove&#39;. Otherwise sequence will be kept. ">newAlignment::saveSequences</a>. </p>
<dl class="section warning"><dt>Warning</dt><dd>Not In Use </dd></dl>
<dl class="section return"><dt>Returns</dt><dd>Vector that represents which sequences are kept (!= -1) and which are rejected (== -1) </dd></dl>

<p class="definition">Definition at line <a class="el" href="new_alignment_8cpp_source.html#l00489">489</a> of file <a class="el" href="new_alignment_8cpp_source.html">newAlignment.cpp</a>.</p>

</div>
</div>
<a id="ad08d05c08eb66665defed4a16ab08e91"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ad08d05c08eb66665defed4a16ab08e91">&#9670;&nbsp;</a></span>getNumAminos()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int newAlignment::getNumAminos </td>
          <td>(</td>
          <td class="paramtype">void&#160;</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Residues number getter. It counts gaps as residue. </p>
<dl class="section return"><dt>Returns</dt><dd>Number of residues in the alignment. </dd></dl>

<p class="definition">Definition at line <a class="el" href="new_alignment_8cpp_source.html#l00258">258</a> of file <a class="el" href="new_alignment_8cpp_source.html">newAlignment.cpp</a>.</p>

<p class="reference">Referenced by <a class="el" href="trimal_manager_8cpp_source.html#l01218">trimAlManager::check_block_size()</a>, <a class="el" href="trimal_manager_8cpp_source.html#l01705">trimAlManager::CleanResiduesNonAuto()</a>, and <a class="el" href="statistics_consistency_8cpp_source.html#l00368">statisticsConsistency::forceComparison()</a>.</p>
<div id="dynsection-8" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-8-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-8-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-8-content" class="dyncontent" style="display:none;">
<div class="center"><iframe scrolling="no" frameborder="0" src="classnew_alignment_ad08d05c08eb66665defed4a16ab08e91_icgraph.svg" width="932" height="183"><p><b>This browser is not able to show SVG: try Firefox, Chrome, Safari, or Opera instead.</b></p></iframe></div>
</div>

</div>
</div>
<a id="acfd820148c267a3d698a39758ecf190d"></a>
<h2 class="memtitle"><span class="permalink"><a href="#acfd820148c267a3d698a39758ecf190d">&#9670;&nbsp;</a></span>getNumSpecies()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int newAlignment::getNumSpecies </td>
          <td>(</td>
          <td class="paramtype">void&#160;</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Number of sequences getter. </p>
<dl class="section return"><dt>Returns</dt><dd>Number of sequences in the alignment. </dd></dl>

<p class="definition">Definition at line <a class="el" href="new_alignment_8cpp_source.html#l00251">251</a> of file <a class="el" href="new_alignment_8cpp_source.html">newAlignment.cpp</a>.</p>

<p class="reference">Referenced by <a class="el" href="trimal_manager_8cpp_source.html#l01274">trimAlManager::check_backtranslation_infile_names_corresponde()</a>, <a class="el" href="trimal_manager_8cpp_source.html#l00952">trimAlManager::check_clusters_incompatibilities()</a>, <a class="el" href="trimal_manager_8cpp_source.html#l00874">trimAlManager::check_select_cols_and_seqs_incompatibilities()</a>, <a class="el" href="statistics_consistency_8cpp_source.html#l00194">statisticsConsistency::compareAndChoose()</a>, <a class="el" href="statistics_consistency_8cpp_source.html#l00368">statisticsConsistency::forceComparison()</a>, and <a class="el" href="new_alignment_8cpp_source.html#l00525">prepareCodingSequence()</a>.</p>
<div id="dynsection-9" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-9-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-9-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-9-content" class="dyncontent" style="display:none;">
<div class="center"><iframe scrolling="no" frameborder="0" src="classnew_alignment_acfd820148c267a3d698a39758ecf190d_icgraph.svg" width="1184" height="391"><p><b>This browser is not able to show SVG: try Firefox, Chrome, Safari, or Opera instead.</b></p></iframe></div>
</div>

</div>
</div>
<a id="ac3d5c9670904bba45707746640aac382"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ac3d5c9670904bba45707746640aac382">&#9670;&nbsp;</a></span>getSequenceNameOrder()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool newAlignment::getSequenceNameOrder </td>
          <td>(</td>
          <td class="paramtype">string *&#160;</td>
          <td class="paramname"><em>names</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int *&#160;</td>
          <td class="paramname"><em>orderVector</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Method to map two sets of sequences, own and external.<br />
The method accepts a vector of names, and test them against the alignment own sequences' names.<br />
If the sequence is present, its index will be stored in orderVector.<br />
. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir"></td><td class="paramname">names</td><td>Vector of sequences names to map. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">orderVector</td><td>Vector of orders to fill. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd><b>True</b> if both sets have the same sequences.<br />
 <b>False</b> otherwise. </dd></dl>

<p class="definition">Definition at line <a class="el" href="new_alignment_8cpp_source.html#l00449">449</a> of file <a class="el" href="new_alignment_8cpp_source.html">newAlignment.cpp</a>.</p>

<p class="reference">Referenced by <a class="el" href="statistics_consistency_8cpp_source.html#l00194">statisticsConsistency::compareAndChoose()</a>, and <a class="el" href="statistics_consistency_8cpp_source.html#l00368">statisticsConsistency::forceComparison()</a>.</p>
<div id="dynsection-10" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-10-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-10-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-10-content" class="dyncontent" style="display:none;">
<div class="center"><iframe scrolling="no" frameborder="0" src="classnew_alignment_ac3d5c9670904bba45707746640aac382_icgraph.svg" width="395" height="118"><p><b>This browser is not able to show SVG: try Firefox, Chrome, Safari, or Opera instead.</b></p></iframe></div>
</div>

</div>
</div>
<a id="a0db3e76f82a71c7a2fe5b9279f8fa310"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a0db3e76f82a71c7a2fe5b9279f8fa310">&#9670;&nbsp;</a></span>getSequences() <span class="overload">[1/3]</span></h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void newAlignment::getSequences </td>
          <td>(</td>
          <td class="paramtype">string *&#160;</td>
          <td class="paramname"><em>names</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Getter for the sequences names. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir">[out]</td><td class="paramname">names</td><td>Vector of sequences names to fill. </td></tr>
  </table>
  </dd>
</dl>

<p class="definition">Definition at line <a class="el" href="new_alignment_8cpp_source.html#l00420">420</a> of file <a class="el" href="new_alignment_8cpp_source.html">newAlignment.cpp</a>.</p>

<p class="reference">Referenced by <a class="el" href="trimal_manager_8cpp_source.html#l01274">trimAlManager::check_backtranslation_infile_names_corresponde()</a>, <a class="el" href="statistics_consistency_8cpp_source.html#l00194">statisticsConsistency::compareAndChoose()</a>, <a class="el" href="statistics_consistency_8cpp_source.html#l00368">statisticsConsistency::forceComparison()</a>, and <a class="el" href="new_alignment_8cpp_source.html#l00525">prepareCodingSequence()</a>.</p>
<div id="dynsection-11" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-11-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-11-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-11-content" class="dyncontent" style="display:none;">
<div class="center"><iframe scrolling="no" frameborder="0" src="classnew_alignment_a0db3e76f82a71c7a2fe5b9279f8fa310_icgraph.svg" width="1174" height="220"><p><b>This browser is not able to show SVG: try Firefox, Chrome, Safari, or Opera instead.</b></p></iframe></div>
</div>

</div>
</div>
<a id="af32ab6b3b73c5ad37c9f1c796a4fbc44"></a>
<h2 class="memtitle"><span class="permalink"><a href="#af32ab6b3b73c5ad37c9f1c796a4fbc44">&#9670;&nbsp;</a></span>getSequences() <span class="overload">[2/3]</span></h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void newAlignment::getSequences </td>
          <td>(</td>
          <td class="paramtype">string *&#160;</td>
          <td class="paramname"><em>names</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int *&#160;</td>
          <td class="paramname"><em>lengths</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Getter for the sequences names and its lengths. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir">[out]</td><td class="paramname">names</td><td>Vector of sequences names to fill. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">lengths</td><td>Vector of lengths to fill. </td></tr>
  </table>
  </dd>
</dl>

<p class="definition">Definition at line <a class="el" href="new_alignment_8cpp_source.html#l00428">428</a> of file <a class="el" href="new_alignment_8cpp_source.html">newAlignment.cpp</a>.</p>

<p class="reference">References <a class="el" href="utils_8cpp_source.html#l00493">utils::removeCharacter()</a>.</p>
<div id="dynsection-12" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-12-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-12-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-12-content" class="dyncontent" style="display:none;">
<div class="center"><iframe scrolling="no" frameborder="0" src="classnew_alignment_af32ab6b3b73c5ad37c9f1c796a4fbc44_cgraph.svg" width="408" height="38"><p><b>This browser is not able to show SVG: try Firefox, Chrome, Safari, or Opera instead.</b></p></iframe></div>
</div>

</div>
</div>
<a id="a2fc0d691c0cce400f007446cc162704e"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a2fc0d691c0cce400f007446cc162704e">&#9670;&nbsp;</a></span>getSequences() <span class="overload">[3/3]</span></h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void newAlignment::getSequences </td>
          <td>(</td>
          <td class="paramtype">string *&#160;</td>
          <td class="paramname"><em>names</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">string *&#160;</td>
          <td class="paramname"><em>sequences</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int *&#160;</td>
          <td class="paramname"><em>lenghts</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Getter for the sequences, its names and lenghts. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir">[out]</td><td class="paramname">names</td><td>Vector of sequences names to fill. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">sequences</td><td>Vector of sequences to fill. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">lengths</td><td>Vector of lengths to fill. </td></tr>
  </table>
  </dd>
</dl>

<p class="definition">Definition at line <a class="el" href="new_alignment_8cpp_source.html#l00438">438</a> of file <a class="el" href="new_alignment_8cpp_source.html">newAlignment.cpp</a>.</p>

<p class="reference">References <a class="el" href="utils_8cpp_source.html#l00493">utils::removeCharacter()</a>.</p>
<div id="dynsection-13" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-13-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-13-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-13-content" class="dyncontent" style="display:none;">
<div class="center"><iframe scrolling="no" frameborder="0" src="classnew_alignment_a2fc0d691c0cce400f007446cc162704e_cgraph.svg" width="408" height="38"><p><b>This browser is not able to show SVG: try Firefox, Chrome, Safari, or Opera instead.</b></p></iframe></div>
</div>

</div>
</div>
<a id="a781fd0f2ae6a9bda2802ffd28372bfaa"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a781fd0f2ae6a9bda2802ffd28372bfaa">&#9670;&nbsp;</a></span>getTranslationCDS()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classnew_alignment.html">newAlignment</a> * newAlignment::getTranslationCDS </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="classnew_alignment.html">newAlignment</a> *&#160;</td>
          <td class="paramname"><em>proteinAlignment</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Method to back translate a protein alignment using the sequences present on the current alignment. <br />
 Current alignment should contain the original sequences which were translated to protein sequences, <b> without any gap.</b> </p>
<dl class="section note"><dt>Note</dt><dd>It's recommended to translate from dna to protein and then, align the protein sequences. </dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">proteinAlignment</td><td>Protein Alignment to use as origin. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>Pointer to new alignment. </dd></dl>

<p class="definition">Definition at line <a class="el" href="new_alignment_8cpp_source.html#l00190">190</a> of file <a class="el" href="new_alignment_8cpp_source.html">newAlignment.cpp</a>.</p>

<p class="reference">References <a class="el" href="new_alignment_8h_source.html#l00069">originalSequenNumber</a>, <a class="el" href="new_alignment_8h_source.html#l00096">saveResidues</a>, <a class="el" href="new_alignment_8h_source.html#l00083">seqsInfo</a>, <a class="el" href="new_alignment_8h_source.html#l00081">seqsName</a>, <a class="el" href="new_alignment_8h_source.html#l00079">sequences</a>, and <a class="el" href="new_alignment_8h_source.html#l00071">sequenNumber</a>.</p>

<p class="reference">Referenced by <a class="el" href="trimal_manager_8cpp_source.html#l01574">trimAlManager::clean_alignment()</a>.</p>
<div id="dynsection-14" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-14-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-14-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-14-content" class="dyncontent" style="display:none;">
<div class="center"><iframe scrolling="no" frameborder="0" src="classnew_alignment_a781fd0f2ae6a9bda2802ffd28372bfaa_icgraph.svg" width="634" height="52"><p><b>This browser is not able to show SVG: try Firefox, Chrome, Safari, or Opera instead.</b></p></iframe></div>
</div>

</div>
</div>
<a id="ab865e218a05742c00d604f4a71ce9e10"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ab865e218a05742c00d604f4a71ce9e10">&#9670;&nbsp;</a></span>isFileAligned()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool newAlignment::isFileAligned </td>
          <td>(</td>
          <td class="paramtype">void&#160;</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>isAligned getter. </p>
<dl class="section return"><dt>Returns</dt><dd>Whether the alignment is aligned or not. </dd></dl>

<p class="definition">Definition at line <a class="el" href="new_alignment_8cpp_source.html#l00496">496</a> of file <a class="el" href="new_alignment_8cpp_source.html">newAlignment.cpp</a>.</p>

<p class="reference">Referenced by <a class="el" href="trimal_manager_8cpp_source.html#l01227">trimAlManager::check_backtranslations()</a>, <a class="el" href="trimal_manager_8cpp_source.html#l01071">trimAlManager::check_file_aligned()</a>, and <a class="el" href="trimal_manager_8cpp_source.html#l01574">trimAlManager::clean_alignment()</a>.</p>
<div id="dynsection-15" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-15-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-15-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-15-content" class="dyncontent" style="display:none;">
<div class="center"><iframe scrolling="no" frameborder="0" src="classnew_alignment_ab865e218a05742c00d604f4a71ce9e10_icgraph.svg" width="872" height="183"><p><b>This browser is not able to show SVG: try Firefox, Chrome, Safari, or Opera instead.</b></p></iframe></div>
</div>

</div>
</div>
<a id="ad50ba126247276cd0bf89119e439c483"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ad50ba126247276cd0bf89119e439c483">&#9670;&nbsp;</a></span>prepareCodingSequence()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool newAlignment::prepareCodingSequence </td>
          <td>(</td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>splitByStopCodon</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>ignoreStopCodon</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="classnew_alignment.html">newAlignment</a> *&#160;</td>
          <td class="paramname"><em>proteinAlignment</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Method to check if the CDS file is correct.<br />
Based on nature of residues: DNA/RNA (Most of the residues)<br />
There is no gaps on the whole dataset.<br />
Each sequence is a multiple of 3.<br />
<br />
It will also remove or split stop codons depending on the flags passed. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">splitByStopCodon</td><td>Flag that informs the method to split sequences if it finds any stop codon. </td></tr>
    <tr><td class="paramname">ignoreStopCodon</td><td>Flag that informs the method to stop reading sequence if it finds any stop codon. </td></tr>
    <tr><td class="paramname">proteinAlignment</td><td>Alignment containing protein sequences which names contains all names in the current alignment. </td></tr>
  </table>
  </dd>
</dl>

<p class="definition">Definition at line <a class="el" href="new_alignment_8cpp_source.html#l00525">525</a> of file <a class="el" href="new_alignment_8cpp_source.html">newAlignment.cpp</a>.</p>

<p class="reference">References <a class="el" href="defines_8h_source.html#l00086">AA</a>, <a class="el" href="new_alignment_8cpp_source.html#l00251">getNumSpecies()</a>, <a class="el" href="new_alignment_8cpp_source.html#l00420">getSequences()</a>, and <a class="el" href="reportsystem_8cpp_source.html#l00375">__internalReport::__reportSystem::report()</a>.</p>

<p class="reference">Referenced by <a class="el" href="trimal_manager_8cpp_source.html#l01260">trimAlManager::check_and_prepare_coding_sequence()</a>.</p>
<div id="dynsection-16" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-16-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-16-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-16-content" class="dyncontent" style="display:none;">
<div class="center"><iframe scrolling="no" frameborder="0" src="classnew_alignment_ad50ba126247276cd0bf89119e439c483_cgraph.svg" width="460" height="155"><p><b>This browser is not able to show SVG: try Firefox, Chrome, Safari, or Opera instead.</b></p></iframe></div>
</div>
<div id="dynsection-17" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-17-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-17-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-17-content" class="dyncontent" style="display:none;">
<div class="center"><iframe scrolling="no" frameborder="0" src="classnew_alignment_ad50ba126247276cd0bf89119e439c483_icgraph.svg" width="931" height="52"><p><b>This browser is not able to show SVG: try Firefox, Chrome, Safari, or Opera instead.</b></p></iframe></div>
</div>

</div>
</div>
<a id="a3b75055480c60265ea57ca557213fe06"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a3b75055480c60265ea57ca557213fe06">&#9670;&nbsp;</a></span>printAlignmentInfo()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void newAlignment::printAlignmentInfo </td>
          <td>(</td>
          <td class="paramtype">ostream &amp;&#160;</td>
          <td class="paramname"><em>output</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Print information about sequences number, average sequence length, maximum and minimum sequences length, etc. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">output</td><td>Output stream. </td></tr>
  </table>
  </dd>
</dl>

<p class="definition">Definition at line <a class="el" href="new_alignment_8cpp_source.html#l00848">848</a> of file <a class="el" href="new_alignment_8cpp_source.html">newAlignment.cpp</a>.</p>

</div>
</div>
<a id="a98c9b5bd9a859600ae8a317d368c85b3"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a98c9b5bd9a859600ae8a317d368c85b3">&#9670;&nbsp;</a></span>printColumnsIdentity_DescriptiveStats()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void newAlignment::printColumnsIdentity_DescriptiveStats </td>
          <td>(</td>
          <td class="paramtype">void&#160;</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Method to print Identity stats: Min, Max, Avg and Std. </p>
<dl class="section warning"><dt>Warning</dt><dd>Not In Use </dd></dl>

<p class="definition">Definition at line <a class="el" href="new_alignment_8cpp_source.html#l01105">1105</a> of file <a class="el" href="new_alignment_8cpp_source.html">newAlignment.cpp</a>.</p>

<p class="reference">References <a class="el" href="utils_8cpp_source.html#l00043">utils::initlVect()</a>.</p>
<div id="dynsection-18" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-18-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-18-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-18-content" class="dyncontent" style="display:none;">
<div class="center"><iframe scrolling="no" frameborder="0" src="classnew_alignment_a98c9b5bd9a859600ae8a317d368c85b3_cgraph.svg" width="350" height="52"><p><b>This browser is not able to show SVG: try Firefox, Chrome, Safari, or Opera instead.</b></p></iframe></div>
</div>

</div>
</div>
<a id="a42894635dca6466308d9e21cec14d365"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a42894635dca6466308d9e21cec14d365">&#9670;&nbsp;</a></span>setBlockSize()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void newAlignment::setBlockSize </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>blockSize</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>blockSize BlockSize Setter. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">blockSize</td><td>New value. </td></tr>
  </table>
  </dd>
</dl>

<p class="definition">Definition at line <a class="el" href="new_alignment_8cpp_source.html#l00273">273</a> of file <a class="el" href="new_alignment_8cpp_source.html">newAlignment.cpp</a>.</p>

<p class="reference">Referenced by <a class="el" href="trimal_manager_8cpp_source.html#l01307">trimAlManager::perform()</a>.</p>
<div id="dynsection-19" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-19-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-19-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-19-content" class="dyncontent" style="display:none;">
<div class="center"><iframe scrolling="no" frameborder="0" src="classnew_alignment_a42894635dca6466308d9e21cec14d365_icgraph.svg" width="407" height="38"><p><b>This browser is not able to show SVG: try Firefox, Chrome, Safari, or Opera instead.</b></p></iframe></div>
</div>

</div>
</div>
<a id="a66015efb0c5b9cd596d02d3c736cffe3"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a66015efb0c5b9cd596d02d3c736cffe3">&#9670;&nbsp;</a></span>setKeepSeqsHeaderFlag()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void newAlignment::setKeepSeqsHeaderFlag </td>
          <td>(</td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>newFlagValue</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Keep Header setter. </p>
<dl class="section warning"><dt>Warning</dt><dd>Not In Use </dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">newFlagValue</td><td>New flag value </td></tr>
  </table>
  </dd>
</dl>

<p class="definition">Definition at line <a class="el" href="new_alignment_8cpp_source.html#l00287">287</a> of file <a class="el" href="new_alignment_8cpp_source.html">newAlignment.cpp</a>.</p>

</div>
</div>
<a id="a29f7f52115fd4ecf13120e54c71b1a1a"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a29f7f52115fd4ecf13120e54c71b1a1a">&#9670;&nbsp;</a></span>setKeepSequencesFlag()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void newAlignment::setKeepSequencesFlag </td>
          <td>(</td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>newFlagValue</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Keep Sequences setter. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">newFlagValue</td><td>New flag value </td></tr>
  </table>
  </dd>
</dl>

<p class="definition">Definition at line <a class="el" href="new_alignment_8cpp_source.html#l00280">280</a> of file <a class="el" href="new_alignment_8cpp_source.html">newAlignment.cpp</a>.</p>

<p class="reference">Referenced by <a class="el" href="trimal_manager_8cpp_source.html#l01307">trimAlManager::perform()</a>.</p>
<div id="dynsection-20" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-20-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-20-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-20-content" class="dyncontent" style="display:none;">
<div class="center"><iframe scrolling="no" frameborder="0" src="classnew_alignment_a29f7f52115fd4ecf13120e54c71b1a1a_icgraph.svg" width="468" height="38"><p><b>This browser is not able to show SVG: try Firefox, Chrome, Safari, or Opera instead.</b></p></iframe></div>
</div>

</div>
</div>
<a id="a8dd7e7438161fcaca8c8f35e63c5079e"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a8dd7e7438161fcaca8c8f35e63c5079e">&#9670;&nbsp;</a></span>setWindowsSize()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void newAlignment::setWindowsSize </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>ghWindow</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>shWindow</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Windows setter. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">ghWindow</td><td>Half the Gap Window. </td></tr>
    <tr><td class="paramname">shWindow</td><td>Half the Similarity Window. </td></tr>
  </table>
  </dd>
</dl>

<p class="definition">Definition at line <a class="el" href="new_alignment_8cpp_source.html#l00265">265</a> of file <a class="el" href="new_alignment_8cpp_source.html">newAlignment.cpp</a>.</p>

<p class="reference">Referenced by <a class="el" href="trimal_manager_8cpp_source.html#l01785">trimAlManager::set_window_size()</a>.</p>
<div id="dynsection-21" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-21-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-21-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-21-content" class="dyncontent" style="display:none;">
<div class="center"><iframe scrolling="no" frameborder="0" src="classnew_alignment_a8dd7e7438161fcaca8c8f35e63c5079e_icgraph.svg" width="608" height="52"><p><b>This browser is not able to show SVG: try Firefox, Chrome, Safari, or Opera instead.</b></p></iframe></div>
</div>

</div>
</div>
<a id="a52710ff330519de8a4e41b27e7e55958"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a52710ff330519de8a4e41b27e7e55958">&#9670;&nbsp;</a></span>updateSequencesAndResiduesNums()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void newAlignment::updateSequencesAndResiduesNums </td>
          <td>(</td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>countSequences</em> = <code>true</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>countResidues</em> = <code>true</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Updates the sequence number and residue number based on saveResidues and saveSequences. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">countSequences</td><td>Whether to count sequences </td></tr>
    <tr><td class="paramname">countResidues</td><td>Whether to count residues </td></tr>
  </table>
  </dd>
</dl>

<p class="definition">Definition at line <a class="el" href="new_alignment_8cpp_source.html#l02127">2127</a> of file <a class="el" href="new_alignment_8cpp_source.html">newAlignment.cpp</a>.</p>

<p class="reference">Referenced by <a class="el" href="_cleaner_8cpp_source.html#l01227">Cleaner::removeColumns()</a>, <a class="el" href="_cleaner_8cpp_source.html#l01271">Cleaner::removeOnlyTerminal()</a>, and <a class="el" href="_cleaner_8cpp_source.html#l01250">Cleaner::removeSequences()</a>.</p>
<div id="dynsection-22" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-22-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-22-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-22-content" class="dyncontent" style="display:none;">
<div class="center"><iframe scrolling="no" frameborder="0" src="classnew_alignment_a52710ff330519de8a4e41b27e7e55958_icgraph.svg" width="1119" height="155"><p><b>This browser is not able to show SVG: try Firefox, Chrome, Safari, or Opera instead.</b></p></iframe></div>
</div>

</div>
</div>
<h2 class="groupheader">Member Data Documentation</h2>
<a id="a792877eafe085fe6aa5b267901fe5be9"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a792877eafe085fe6aa5b267901fe5be9">&#9670;&nbsp;</a></span>identities</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">newAlignment::identities = nullptr</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>2D Matrix that represents how much each sequence resembles other sequences in the same MSA </p>
<p>Identity score between two sequences is the ratio of identical residues by the total length (common and no-common residues) among them. </p>

<p class="definition">Definition at line <a class="el" href="new_alignment_8h_source.html#l00090">90</a> of file <a class="el" href="new_alignment_8h_source.html">newAlignment.h</a>.</p>

<p class="reference">Referenced by <a class="el" href="_cleaner_8cpp_source.html#l01464">Cleaner::calculateRelaxedSeqIdentity()</a>, <a class="el" href="_cleaner_8cpp_source.html#l01506">Cleaner::calculateRepresentativeSeq()</a>, <a class="el" href="_cleaner_8cpp_source.html#l01415">Cleaner::calculateSeqIdentity()</a>, <a class="el" href="_cleaner_8cpp_source.html#l01055">Cleaner::getCutPointClusters()</a>, and <a class="el" href="_cleaner_8cpp_source.html#l00022">Cleaner::selectMethod()</a>.</p>

</div>
</div>
<hr/>The documentation for this class was generated from the following files:<ul>
<li>include/<a class="el" href="new_alignment_8h_source.html">newAlignment.h</a></li>
<li>include/documentation/externalHeadersConfiguration/newAlignment.dox</li>
<li>source/<a class="el" href="new_alignment_8cpp_source.html">newAlignment.cpp</a></li>
</ul>
</div><!-- contents -->
</div><!-- doc-content -->
<!-- start footer part -->
<div id="nav-path" class="navpath"><!-- id is needed for treeview function! -->
  <ul>
    <li class="navelem"><a class="el" href="classnew_alignment.html">newAlignment</a></li>
    <li class="footer">Generated by
    <a href="http://www.doxygen.org/index.html">
    <img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.8.13 </li>
  </ul>
</div>
</body>
</html>
