<!-- HTML header for doxygen 1.8.8-->
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.13"/>
<title>Trimal: newAlignment Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtreedata.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<script type="text/javascript">
  $(document).ready(initResizable);
</script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0" width=100%>
 <tbody>
 <tr style="height: 12px;">
  <td id="projectlogo"><img alt="Logo" src="canvas.png"/></td>
  <td style="padding-left: 0.5em;">
   <div id="projectname">Trimal
   </div>
   <div id="projectbrief">A Multiple Alignment Sequence framework for Cleaning and Reformating.</div>
  </td>
<!--  <td style="float:right;margin-top:18px;vertical-align:bottom">
     <a href="http://assimilationsystems.com/download/">
       <img src="http://assimilationsystems.com/wp-content/uploads/2015/03/Download-Doxygen.png" style="float:right;" height="20" width="73">
     </a>
  </td>-->
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.13 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
</script>
<div id="main-nav"></div>
</div><!-- top -->
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
      <div id="nav-sync" class="sync"></div>
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
$(document).ready(function(){initNavTree('classnew_alignment.html','');});
</script>
<div id="doc-content">
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div class="header">
  <div class="summary">
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="#pub-attribs">Public Attributes</a> &#124;
<a href="classnew_alignment-members.html">List of all members</a>  </div>
  <div class="headertitle">
<div class="title">newAlignment Class Reference</div>  </div>
</div><!--header-->
<div class="contents">

<p>Class containing an alignment<br />
 This class stores the alignment sequences with it's names, residues and extra information.<br />
 It contains multiple methods regarding the sequences.<br />
 It also contains submodules that provide methods for<b> Calculating statistics</b>, <b> Cleaning the alignment</b> and<b> Printing sequences information</b>.  
 <a href="classnew_alignment.html#details">More...</a></p>
<div class="dynheader">
Collaboration diagram for newAlignment:</div>
<div class="dyncontent">
<div class="center"><img src="classnew_alignment__coll__graph.png" border="0" usemap="#new_alignment_coll__map" alt="Collaboration graph"/></div>
<map name="new_alignment_coll__map" id="new_alignment_coll__map">
<area shape="rect" id="node5" href="classsequences_matrix.html" title="sequencesMatrix" alt="" coords="1033,512,1155,539"/>
<area shape="rect" id="node12" href="class_cleaner.html" title="Subclass contained by the alignment object to make cleaning operations on it. " alt="" coords="1061,124,1127,151"/>
<area shape="rect" id="node14" href="class_statistics_manager.html" title="StatisticsManager" alt="" coords="1031,328,1157,355"/>
<area shape="rect" id="node3" href="classstatistics_gaps.html" title="statisticsGaps" alt="" coords="375,323,481,349"/>
<area shape="rect" id="node8" href="classstatistics_conservation.html" title="statisticsConservation" alt="" coords="505,275,657,301"/>
<area shape="rect" id="node11" href="classsimilarity_matrix.html" title="similarityMatrix" alt="" coords="240,191,351,217"/>
</map>
<center><span class="legend">[<a target="top" href="graph_legend.html">legend</a>]</span></center></div>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:acbeb320f0f70222f81a906ccfc33a016"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#acbeb320f0f70222f81a906ccfc33a016">fillMatrices</a> (bool aligned)</td></tr>
<tr class="memdesc:acbeb320f0f70222f81a906ccfc33a016"><td class="mdescLeft">&#160;</td><td class="mdescRight">Method to intialize data that has been modified on the alignment.<br />
The method checks if the sequences have been correctly loaded and are free of errors.<br />
It checks if sequences contain unknown characters, sets the isAligned flag and initializes 'saveResidues' and 'saveSequences', depending on the sizes of the sequences (wether or not they have the same length).  <a href="#acbeb320f0f70222f81a906ccfc33a016">More...</a><br /></td></tr>
<tr class="separator:acbeb320f0f70222f81a906ccfc33a016"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a384a40982a61649e699d6733fe2818d3"><td class="memItemLeft" align="right" valign="top"><a id="a384a40982a61649e699d6733fe2818d3"></a>
&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a384a40982a61649e699d6733fe2818d3">newAlignment</a> (void)</td></tr>
<tr class="memdesc:a384a40982a61649e699d6733fe2818d3"><td class="mdescLeft">&#160;</td><td class="mdescRight">Constructor. <br /></td></tr>
<tr class="separator:a384a40982a61649e699d6733fe2818d3"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8e7fbe8f58d588cb10cb10981efb9915"><td class="memItemLeft" align="right" valign="top"><a id="a8e7fbe8f58d588cb10cb10981efb9915"></a>
&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a8e7fbe8f58d588cb10cb10981efb9915">newAlignment</a> (<a class="el" href="classnew_alignment.html">newAlignment</a> &amp;)</td></tr>
<tr class="memdesc:a8e7fbe8f58d588cb10cb10981efb9915"><td class="mdescLeft">&#160;</td><td class="mdescRight">Copy Constructor. <br /></td></tr>
<tr class="separator:a8e7fbe8f58d588cb10cb10981efb9915"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:afb16c10cb9a7b4b9868783ff61e67c8c"><td class="memItemLeft" align="right" valign="top"><a id="afb16c10cb9a7b4b9868783ff61e67c8c"></a>
&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#afb16c10cb9a7b4b9868783ff61e67c8c">~newAlignment</a> (void)</td></tr>
<tr class="memdesc:afb16c10cb9a7b4b9868783ff61e67c8c"><td class="mdescLeft">&#160;</td><td class="mdescRight">Destructor. <br /></td></tr>
<tr class="separator:afb16c10cb9a7b4b9868783ff61e67c8c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a71e1be2c77493d5094a4d7741e5ea26a"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a71e1be2c77493d5094a4d7741e5ea26a">getNumSpecies</a> (void)</td></tr>
<tr class="memdesc:a71e1be2c77493d5094a4d7741e5ea26a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Getter for the number of species.  <a href="#a71e1be2c77493d5094a4d7741e5ea26a">More...</a><br /></td></tr>
<tr class="separator:a71e1be2c77493d5094a4d7741e5ea26a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0db3e76f82a71c7a2fe5b9279f8fa310"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a0db3e76f82a71c7a2fe5b9279f8fa310">getSequences</a> (string *names)</td></tr>
<tr class="memdesc:a0db3e76f82a71c7a2fe5b9279f8fa310"><td class="mdescLeft">&#160;</td><td class="mdescRight">Getter for the sequences names.  <a href="#a0db3e76f82a71c7a2fe5b9279f8fa310">More...</a><br /></td></tr>
<tr class="separator:a0db3e76f82a71c7a2fe5b9279f8fa310"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af32ab6b3b73c5ad37c9f1c796a4fbc44"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#af32ab6b3b73c5ad37c9f1c796a4fbc44">getSequences</a> (string *names, int *lengths)</td></tr>
<tr class="memdesc:af32ab6b3b73c5ad37c9f1c796a4fbc44"><td class="mdescLeft">&#160;</td><td class="mdescRight">Getter for the sequences names and its lenghts.  <a href="#af32ab6b3b73c5ad37c9f1c796a4fbc44">More...</a><br /></td></tr>
<tr class="separator:af32ab6b3b73c5ad37c9f1c796a4fbc44"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2fc0d691c0cce400f007446cc162704e"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a2fc0d691c0cce400f007446cc162704e">getSequences</a> (string *names, string *<a class="el" href="classnew_alignment.html#a23f06ad268f0e720bf3d5699a4cb7225">sequences</a>, int *lenghts)</td></tr>
<tr class="memdesc:a2fc0d691c0cce400f007446cc162704e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Getter for the sequences, its names and lenghts.  <a href="#a2fc0d691c0cce400f007446cc162704e">More...</a><br /></td></tr>
<tr class="separator:a2fc0d691c0cce400f007446cc162704e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac3d5c9670904bba45707746640aac382"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#ac3d5c9670904bba45707746640aac382">getSequenceNameOrder</a> (string *names, int *orderVector)</td></tr>
<tr class="memdesc:ac3d5c9670904bba45707746640aac382"><td class="mdescLeft">&#160;</td><td class="mdescRight">Method to map two sets of sequences, own and external.<br />
The method accepts a vector of names, and test them against the alignment owned sequences.<br />
If the sequence is present, its order on the alignment sequences vector will be stored in orderVector, in the same position it's name was in the names vector.<br />
.  <a href="#ac3d5c9670904bba45707746640aac382">More...</a><br /></td></tr>
<tr class="separator:ac3d5c9670904bba45707746640aac382"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a906c64b2a7b38a2b45308d67429ff1f2"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a906c64b2a7b38a2b45308d67429ff1f2">getNumAminos</a> (void)</td></tr>
<tr class="memdesc:a906c64b2a7b38a2b45308d67429ff1f2"><td class="mdescLeft">&#160;</td><td class="mdescRight">Residues getter.  <a href="#a906c64b2a7b38a2b45308d67429ff1f2">More...</a><br /></td></tr>
<tr class="separator:a906c64b2a7b38a2b45308d67429ff1f2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8dd7e7438161fcaca8c8f35e63c5079e"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a8dd7e7438161fcaca8c8f35e63c5079e">setWindowsSize</a> (int ghWindow, int shWindow)</td></tr>
<tr class="memdesc:a8dd7e7438161fcaca8c8f35e63c5079e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Windows setter.  <a href="#a8dd7e7438161fcaca8c8f35e63c5079e">More...</a><br /></td></tr>
<tr class="separator:a8dd7e7438161fcaca8c8f35e63c5079e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a42894635dca6466308d9e21cec14d365"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a42894635dca6466308d9e21cec14d365">setBlockSize</a> (int blockSize)</td></tr>
<tr class="memdesc:a42894635dca6466308d9e21cec14d365"><td class="mdescLeft">&#160;</td><td class="mdescRight">blockSize Setter.  <a href="#a42894635dca6466308d9e21cec14d365">More...</a><br /></td></tr>
<tr class="separator:a42894635dca6466308d9e21cec14d365"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7a471f9393fa1da913bd9a65cc96a9f1"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a7a471f9393fa1da913bd9a65cc96a9f1">getBlockSize</a> (void)</td></tr>
<tr class="memdesc:a7a471f9393fa1da913bd9a65cc96a9f1"><td class="mdescLeft">&#160;</td><td class="mdescRight">blockSize getter.  <a href="#a7a471f9393fa1da913bd9a65cc96a9f1">More...</a><br /></td></tr>
<tr class="separator:a7a471f9393fa1da913bd9a65cc96a9f1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af730d0021082c5d52bc0b971a587c8e1"><td class="memItemLeft" align="right" valign="top"><a id="af730d0021082c5d52bc0b971a587c8e1"></a>
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#af730d0021082c5d52bc0b971a587c8e1">printSeqIdentity</a> (void)</td></tr>
<tr class="memdesc:af730d0021082c5d52bc0b971a587c8e1"><td class="mdescLeft">&#160;</td><td class="mdescRight">Method to print different identity values computed from the alignment.<br />
In this method we assess the identity values matrix, as well as diferent average values. <br />
Moreover, the method computes which one is the most similar sequence, in term of identity values, for each one on this alignment. <br /></td></tr>
<tr class="separator:af730d0021082c5d52bc0b971a587c8e1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae1ee3ea76e1cb268097f1df0c465d6f5"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#ae1ee3ea76e1cb268097f1df0c465d6f5">getAlignmentType</a> (void)</td></tr>
<tr class="memdesc:ae1ee3ea76e1cb268097f1df0c465d6f5"><td class="mdescLeft">&#160;</td><td class="mdescRight">Alignment type getter.  <a href="#ae1ee3ea76e1cb268097f1df0c465d6f5">More...</a><br /></td></tr>
<tr class="separator:ae1ee3ea76e1cb268097f1df0c465d6f5"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a882c53e08d4435c50328d5bf7b10e2d1"><td class="memItemLeft" align="right" valign="top">int *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a882c53e08d4435c50328d5bf7b10e2d1">getCorrespResidues</a> (void)</td></tr>
<tr class="memdesc:a882c53e08d4435c50328d5bf7b10e2d1"><td class="mdescLeft">&#160;</td><td class="mdescRight">Method that returns the correspondence between old and new alignment residues.  <a href="#a882c53e08d4435c50328d5bf7b10e2d1">More...</a><br /></td></tr>
<tr class="separator:a882c53e08d4435c50328d5bf7b10e2d1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:adf85c31144b199b65e24b98b6c755299"><td class="memItemLeft" align="right" valign="top">int *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#adf85c31144b199b65e24b98b6c755299">getCorrespSequences</a> (void)</td></tr>
<tr class="memdesc:adf85c31144b199b65e24b98b6c755299"><td class="mdescLeft">&#160;</td><td class="mdescRight">Method that returns the correspondence between old and new alignment sequences.  <a href="#adf85c31144b199b65e24b98b6c755299">More...</a><br /></td></tr>
<tr class="separator:adf85c31144b199b65e24b98b6c755299"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad56533a40875ddbe8aa65ee391840464"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#ad56533a40875ddbe8aa65ee391840464">isFileAligned</a> (void)</td></tr>
<tr class="memdesc:ad56533a40875ddbe8aa65ee391840464"><td class="mdescLeft">&#160;</td><td class="mdescRight">isAligned getter.  <a href="#ad56533a40875ddbe8aa65ee391840464">More...</a><br /></td></tr>
<tr class="separator:ad56533a40875ddbe8aa65ee391840464"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3d1ff48743c38351c7e755f94baebf03"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classnew_alignment.html">newAlignment</a> *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a3d1ff48743c38351c7e755f94baebf03">getTranslationCDS</a> (int newResidues, int newSequences, int *columnsToKeep, string *oldSequencesNames, <a class="el" href="classsequences_matrix.html">sequencesMatrix</a> *sequenceMatrix, <a class="el" href="classnew_alignment.html">newAlignment</a> *proteinAlignment)</td></tr>
<tr class="memdesc:a3d1ff48743c38351c7e755f94baebf03"><td class="mdescLeft">&#160;</td><td class="mdescRight">Method to obtain a DNA alignment from a Protein alignment.  <a href="#a3d1ff48743c38351c7e755f94baebf03">More...</a><br /></td></tr>
<tr class="separator:a3d1ff48743c38351c7e755f94baebf03"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae8501233fe4f4ef6bfe30c99c744c7c2"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#ae8501233fe4f4ef6bfe30c99c744c7c2">checkCorrespondence</a> (string *names, int *lenghts, int totalInputSequences, int multiple)</td></tr>
<tr class="memdesc:ae8501233fe4f4ef6bfe30c99c744c7c2"><td class="mdescLeft">&#160;</td><td class="mdescRight">Function to check CDS file.<br />
It checks if sequences of input alignment are all present on the CDS file.<br />
If nucleotide sequence is larger than protein sequence length * 3, warns about it and cuts the nucleotide sequence<br />
If sequence has indetermination symbols, it warns about it.<br />
If nucleotide sequence is smaller than protein sequence * 3, it adds some 'N' at the end of the nucleotide sequence.  <a href="#ae8501233fe4f4ef6bfe30c99c744c7c2">More...</a><br /></td></tr>
<tr class="separator:ae8501233fe4f4ef6bfe30c99c744c7c2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3de39c62c08a265178f1fa7df76c3ab7"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a3de39c62c08a265178f1fa7df76c3ab7">fillNewDataStructure</a> (string *newMatrix, string *newNames)</td></tr>
<tr class="memdesc:a3de39c62c08a265178f1fa7df76c3ab7"><td class="mdescLeft">&#160;</td><td class="mdescRight">Method that copies information of the alignment to the pointers given.  <a href="#a3de39c62c08a265178f1fa7df76c3ab7">More...</a><br /></td></tr>
<tr class="separator:a3de39c62c08a265178f1fa7df76c3ab7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a54b8820f397c14cf4d8d5bc8a89daf24"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a54b8820f397c14cf4d8d5bc8a89daf24">fillNewDataStructure</a> (<a class="el" href="structnew_values.html">newValues</a> *data)</td></tr>
<tr class="memdesc:a54b8820f397c14cf4d8d5bc8a89daf24"><td class="mdescLeft">&#160;</td><td class="mdescRight">Method that copies information of the alignment to the pointers given.  <a href="#a54b8820f397c14cf4d8d5bc8a89daf24">More...</a><br /></td></tr>
<tr class="separator:a54b8820f397c14cf4d8d5bc8a89daf24"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad916722e4ee892ee9f5cf85f0bdb80aa"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#ad916722e4ee892ee9f5cf85f0bdb80aa">calculateColIdentity</a> (float *columnIdentity)</td></tr>
<tr class="memdesc:ad916722e4ee892ee9f5cf85f0bdb80aa"><td class="mdescLeft">&#160;</td><td class="mdescRight">Method that calculates the columns identity value.<br />
This is, the frequency of the most present element in the column, being residue, indetermination and gap allowed.  <a href="#ad916722e4ee892ee9f5cf85f0bdb80aa">More...</a><br /></td></tr>
<tr class="separator:ad916722e4ee892ee9f5cf85f0bdb80aa"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa22c770103d0fce45cfb1840aab85888"><td class="memItemLeft" align="right" valign="top"><a id="aa22c770103d0fce45cfb1840aab85888"></a>
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#aa22c770103d0fce45cfb1840aab85888">printColumnsIdentity_DescriptiveStats</a> (void)</td></tr>
<tr class="memdesc:aa22c770103d0fce45cfb1840aab85888"><td class="mdescLeft">&#160;</td><td class="mdescRight">Method to print Indentity stats: Min, Max, Avg and Std. <br /></td></tr>
<tr class="separator:aa22c770103d0fce45cfb1840aab85888"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a29f7f52115fd4ecf13120e54c71b1a1a"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a29f7f52115fd4ecf13120e54c71b1a1a">setKeepSequencesFlag</a> (bool newFlagValue)</td></tr>
<tr class="memdesc:a29f7f52115fd4ecf13120e54c71b1a1a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Keep Sequences setter.  <a href="#a29f7f52115fd4ecf13120e54c71b1a1a">More...</a><br /></td></tr>
<tr class="separator:a29f7f52115fd4ecf13120e54c71b1a1a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa77c80b10d05f3a069a5f63223e13ca2"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#aa77c80b10d05f3a069a5f63223e13ca2">setKeepSeqsHeaderFlag</a> (bool newFlagValue)</td></tr>
<tr class="memdesc:aa77c80b10d05f3a069a5f63223e13ca2"><td class="mdescLeft">&#160;</td><td class="mdescRight">Keep Header setter.  <a href="#aa77c80b10d05f3a069a5f63223e13ca2">More...</a><br /></td></tr>
<tr class="separator:aa77c80b10d05f3a069a5f63223e13ca2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3b75055480c60265ea57ca557213fe06"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a3b75055480c60265ea57ca557213fe06">printAlignmentInfo</a> (ostream &amp;output)</td></tr>
<tr class="memdesc:a3b75055480c60265ea57ca557213fe06"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print information about sequences number, average sequence length, maximum and minimum sequences length, etc.  <a href="#a3b75055480c60265ea57ca557213fe06">More...</a><br /></td></tr>
<tr class="separator:a3b75055480c60265ea57ca557213fe06"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad50ba126247276cd0bf89119e439c483"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#ad50ba126247276cd0bf89119e439c483">prepareCodingSequence</a> (bool splitByStopCodon, bool ignoreStopCodon, <a class="el" href="classnew_alignment.html">newAlignment</a> *proteinAlignment)</td></tr>
<tr class="memdesc:ad50ba126247276cd0bf89119e439c483"><td class="mdescLeft">&#160;</td><td class="mdescRight">Method to check if the CDS file is correct.<br />
Based on nature of residues: DNA/RNA (Most of the residues)<br />
There is no gaps on the whole dataset.<br />
Each sequence is a multiple of 3.<br />
<br />
It will also remove or split stop codons depending on the flags passed.  <a href="#ad50ba126247276cd0bf89119e439c483">More...</a><br /></td></tr>
<tr class="separator:ad50ba126247276cd0bf89119e439c483"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3377e8d600e1cbb9e65adf2995c877f9"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a3377e8d600e1cbb9e65adf2995c877f9">alignmentSummaryHTML</a> (char *destFile, int residues, int seqs, int *selectedRes, int *selectedSeq, float *consValues)</td></tr>
<tr class="memdesc:a3377e8d600e1cbb9e65adf2995c877f9"><td class="mdescLeft">&#160;</td><td class="mdescRight">Method to report the trimming results in HTML.<br />
It contains information about the original and trimmed alignments, showing which residues and sequences have been kept and which ones have been removed on the output.  <a href="#a3377e8d600e1cbb9e65adf2995c877f9">More...</a><br /></td></tr>
<tr class="separator:a3377e8d600e1cbb9e65adf2995c877f9"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a26d10a23f8f71f95657a05e88a9a8fff"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a26d10a23f8f71f95657a05e88a9a8fff">alignmentColourHTML</a> (ostream &amp;file)</td></tr>
<tr class="memdesc:a26d10a23f8f71f95657a05e88a9a8fff"><td class="mdescLeft">&#160;</td><td class="mdescRight">Method that saves a clustal-color based visualization of the alignment.<br />
It doesn't report any trimming result, meaning this only shows information of an alignment.  <a href="#a26d10a23f8f71f95657a05e88a9a8fff">More...</a><br /></td></tr>
<tr class="separator:a26d10a23f8f71f95657a05e88a9a8fff"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-attribs"></a>
Public Attributes</h2></td></tr>
<tr class="memitem:a2417feb815493866d0137ae5f5cd1028"><td class="memItemLeft" align="right" valign="top"><a id="a2417feb815493866d0137ae5f5cd1028"></a>
<a class="el" href="class_cleaner.html">Cleaner</a> *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a2417feb815493866d0137ae5f5cd1028">Cleaning</a></td></tr>
<tr class="memdesc:a2417feb815493866d0137ae5f5cd1028"><td class="mdescLeft">&#160;</td><td class="mdescRight">Cleaning submodule.<br />
It contains methods and variables related to trimming. <br /></td></tr>
<tr class="separator:a2417feb815493866d0137ae5f5cd1028"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6324267e64b68bb85d993cf602d008eb"><td class="memItemLeft" align="right" valign="top"><a id="a6324267e64b68bb85d993cf602d008eb"></a>
<a class="el" href="class_statistics_manager.html">StatisticsManager</a> *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a6324267e64b68bb85d993cf602d008eb">Statistics</a></td></tr>
<tr class="memdesc:a6324267e64b68bb85d993cf602d008eb"><td class="mdescLeft">&#160;</td><td class="mdescRight">Statistics submodule.<br />
It contains methods and variables related to Statistics calculation and reporting. <br /></td></tr>
<tr class="separator:a6324267e64b68bb85d993cf602d008eb"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8b328dfd8e7d3304cb144f2c2114ab5c"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classsequences_matrix.html">sequencesMatrix</a> *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a8b328dfd8e7d3304cb144f2c2114ab5c">SequencesMatrix</a></td></tr>
<tr class="memdesc:a8b328dfd8e7d3304cb144f2c2114ab5c"><td class="mdescLeft">&#160;</td><td class="mdescRight">SequencesMatrix submodule<br />
.  <a href="#a8b328dfd8e7d3304cb144f2c2114ab5c">More...</a><br /></td></tr>
<tr class="separator:a8b328dfd8e7d3304cb144f2c2114ab5c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a39c8e7f46c620cc32d32a81e1a78c31b"><td class="memItemLeft" align="right" valign="top"><a id="a39c8e7f46c620cc32d32a81e1a78c31b"></a>
<a class="el" href="classstatistics_gaps.html">statisticsGaps</a> *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a39c8e7f46c620cc32d32a81e1a78c31b">sgaps</a></td></tr>
<tr class="memdesc:a39c8e7f46c620cc32d32a81e1a78c31b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Gaps Statistics submodule. <br /></td></tr>
<tr class="separator:a39c8e7f46c620cc32d32a81e1a78c31b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a01d5c3d9550cad228f65b6e9de2495fd"><td class="memItemLeft" align="right" valign="top"><a id="a01d5c3d9550cad228f65b6e9de2495fd"></a>
<a class="el" href="classstatistics_conservation.html">statisticsConservation</a> *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a01d5c3d9550cad228f65b6e9de2495fd">scons</a></td></tr>
<tr class="memdesc:a01d5c3d9550cad228f65b6e9de2495fd"><td class="mdescLeft">&#160;</td><td class="mdescRight">Conservation Statistics submodule. <br /></td></tr>
<tr class="separator:a01d5c3d9550cad228f65b6e9de2495fd"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4a6b47de8bef3e783f30548dfa48d84c"><td class="memItemLeft" align="right" valign="top"><a id="a4a6b47de8bef3e783f30548dfa48d84c"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a4a6b47de8bef3e783f30548dfa48d84c">sequenNumber</a></td></tr>
<tr class="memdesc:a4a6b47de8bef3e783f30548dfa48d84c"><td class="mdescLeft">&#160;</td><td class="mdescRight">Number of sequences present on the alignment. <br /></td></tr>
<tr class="separator:a4a6b47de8bef3e783f30548dfa48d84c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aff4e7761ec80bea74cdf28c0269b27db"><td class="memItemLeft" align="right" valign="top"><a id="aff4e7761ec80bea74cdf28c0269b27db"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#aff4e7761ec80bea74cdf28c0269b27db">residNumber</a></td></tr>
<tr class="memdesc:aff4e7761ec80bea74cdf28c0269b27db"><td class="mdescLeft">&#160;</td><td class="mdescRight">Number of residues present on the alignment if it is aligned. <br /></td></tr>
<tr class="separator:aff4e7761ec80bea74cdf28c0269b27db"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab6b16a670e278cd5e283573665fa450f"><td class="memItemLeft" align="right" valign="top"><a id="ab6b16a670e278cd5e283573665fa450f"></a>
bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#ab6b16a670e278cd5e283573665fa450f">isAligned</a></td></tr>
<tr class="memdesc:ab6b16a670e278cd5e283573665fa450f"><td class="mdescLeft">&#160;</td><td class="mdescRight">Flag that indicates if all sequences on the alignment have the same length. <br /></td></tr>
<tr class="separator:ab6b16a670e278cd5e283573665fa450f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ada2fb94cccdf0760c22dab39d278680c"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#ada2fb94cccdf0760c22dab39d278680c">dataType</a></td></tr>
<tr class="memdesc:ada2fb94cccdf0760c22dab39d278680c"><td class="mdescLeft">&#160;</td><td class="mdescRight">Number that represents the type of the alignment.  <a href="#ada2fb94cccdf0760c22dab39d278680c">More...</a><br /></td></tr>
<tr class="separator:ada2fb94cccdf0760c22dab39d278680c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a23f06ad268f0e720bf3d5699a4cb7225"><td class="memItemLeft" align="right" valign="top"><a id="a23f06ad268f0e720bf3d5699a4cb7225"></a>
string *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a23f06ad268f0e720bf3d5699a4cb7225">sequences</a></td></tr>
<tr class="memdesc:a23f06ad268f0e720bf3d5699a4cb7225"><td class="mdescLeft">&#160;</td><td class="mdescRight">Sequences vector containing the sequences residues. <br /></td></tr>
<tr class="separator:a23f06ad268f0e720bf3d5699a4cb7225"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa4bf60277ddf15e935a15b22becd6ad9"><td class="memItemLeft" align="right" valign="top"><a id="aa4bf60277ddf15e935a15b22becd6ad9"></a>
string *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#aa4bf60277ddf15e935a15b22becd6ad9">seqsName</a></td></tr>
<tr class="memdesc:aa4bf60277ddf15e935a15b22becd6ad9"><td class="mdescLeft">&#160;</td><td class="mdescRight">Sequences names vector containing the sequences names. <br /></td></tr>
<tr class="separator:aa4bf60277ddf15e935a15b22becd6ad9"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad4e91d14819ffa8e6e57ae7c23d35327"><td class="memItemLeft" align="right" valign="top"><a id="ad4e91d14819ffa8e6e57ae7c23d35327"></a>
string *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#ad4e91d14819ffa8e6e57ae7c23d35327">seqsInfo</a></td></tr>
<tr class="memdesc:ad4e91d14819ffa8e6e57ae7c23d35327"><td class="mdescLeft">&#160;</td><td class="mdescRight">Sequences information vector containing extra information about each sequence. <br /></td></tr>
<tr class="separator:ad4e91d14819ffa8e6e57ae7c23d35327"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6c9904f7a01e200bd167f3cf736af4d6"><td class="memItemLeft" align="right" valign="top"><a id="a6c9904f7a01e200bd167f3cf736af4d6"></a>
string&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a6c9904f7a01e200bd167f3cf736af4d6">filename</a></td></tr>
<tr class="memdesc:a6c9904f7a01e200bd167f3cf736af4d6"><td class="mdescLeft">&#160;</td><td class="mdescRight">Original filename where this alignment was loaded from. <br /></td></tr>
<tr class="separator:a6c9904f7a01e200bd167f3cf736af4d6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a34b1ee01d1030ce737034b35f3631e8a"><td class="memItemLeft" align="right" valign="top"><a id="a34b1ee01d1030ce737034b35f3631e8a"></a>
string&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a34b1ee01d1030ce737034b35f3631e8a">aligInfo</a></td></tr>
<tr class="memdesc:a34b1ee01d1030ce737034b35f3631e8a"><td class="mdescLeft">&#160;</td><td class="mdescRight">String containing information of the alignment as a whole. <br /></td></tr>
<tr class="separator:a34b1ee01d1030ce737034b35f3631e8a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a792877eafe085fe6aa5b267901fe5be9"><td class="memItemLeft" align="right" valign="top"><a id="a792877eafe085fe6aa5b267901fe5be9"></a>
float **&#160;</td><td class="memItemRight" valign="bottom"><b>identities</b></td></tr>
<tr class="separator:a792877eafe085fe6aa5b267901fe5be9"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4d5be3aa01fde32cbc0fddbf75c3c01e"><td class="memItemLeft" align="right" valign="top"><a id="a4d5be3aa01fde32cbc0fddbf75c3c01e"></a>
int *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a4d5be3aa01fde32cbc0fddbf75c3c01e">residuesNumber</a></td></tr>
<tr class="memdesc:a4d5be3aa01fde32cbc0fddbf75c3c01e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Number of residues of this alignment, if it is aligned. <br /></td></tr>
<tr class="separator:a4d5be3aa01fde32cbc0fddbf75c3c01e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5239bd4e742631b58efc7f09dd67f165"><td class="memItemLeft" align="right" valign="top"><a id="a5239bd4e742631b58efc7f09dd67f165"></a>
int *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a5239bd4e742631b58efc7f09dd67f165">saveResidues</a></td></tr>
<tr class="memdesc:a5239bd4e742631b58efc7f09dd67f165"><td class="mdescLeft">&#160;</td><td class="mdescRight">Vector containing which residues will be kept/removed.<br />
-1 Indicates a 'removed/toRemove'. Otherwise it will be kept. <br /></td></tr>
<tr class="separator:a5239bd4e742631b58efc7f09dd67f165"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a76d057ddcebc7c88dd7f9e3ff0ce95bf"><td class="memItemLeft" align="right" valign="top"><a id="a76d057ddcebc7c88dd7f9e3ff0ce95bf"></a>
int *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classnew_alignment.html#a76d057ddcebc7c88dd7f9e3ff0ce95bf">saveSequences</a></td></tr>
<tr class="memdesc:a76d057ddcebc7c88dd7f9e3ff0ce95bf"><td class="mdescLeft">&#160;</td><td class="mdescRight">Vector containing which sequences will be kept/removed.<br />
-1 Indicates a 'removed/toRemove'. Otherwise it will be kept. <br /></td></tr>
<tr class="separator:a76d057ddcebc7c88dd7f9e3ff0ce95bf"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p>Class containing an alignment<br />
 This class stores the alignment sequences with it's names, residues and extra information.<br />
 It contains multiple methods regarding the sequences.<br />
 It also contains submodules that provide methods for<b> Calculating statistics</b>, <b> Cleaning the alignment</b> and<b> Printing sequences information</b>. </p>

<p class="definition">Definition at line <a class="el" href="new_alignment_8h_source.html#l00074">74</a> of file <a class="el" href="new_alignment_8h_source.html">newAlignment.h</a>.</p>
</div><h2 class="groupheader">Member Function Documentation</h2>
<a id="a26d10a23f8f71f95657a05e88a9a8fff"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a26d10a23f8f71f95657a05e88a9a8fff">&#9670;&nbsp;</a></span>alignmentColourHTML()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool newAlignment::alignmentColourHTML </td>
          <td>(</td>
          <td class="paramtype">ostream &amp;&#160;</td>
          <td class="paramname"><em>file</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Method that saves a clustal-color based visualization of the alignment.<br />
It doesn't report any trimming result, meaning this only shows information of an alignment. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">file</td><td>Where to save the visualization. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd><b>True</b> if everything went ok. <br />
 <b>False</b> if alignment is not aligned. </dd></dl>

<p class="definition">Definition at line <a class="el" href="new_alignment_8cpp_source.html#l01343">1343</a> of file <a class="el" href="new_alignment_8cpp_source.html">newAlignment.cpp</a>.</p>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classnew_alignment_a26d10a23f8f71f95657a05e88a9a8fff_cgraph.png" border="0" usemap="#classnew_alignment_a26d10a23f8f71f95657a05e88a9a8fff_cgraph" alt=""/></div>
<map name="classnew_alignment_a26d10a23f8f71f95657a05e88a9a8fff_cgraph" id="classnew_alignment_a26d10a23f8f71f95657a05e88a9a8fff_cgraph">
<area shape="rect" id="node2" href="classutils.html#a40640e1942f00797e9fe8284e8fbdb14" title="Maximum of two numbers method. " alt="" coords="328,5,407,32"/>
<area shape="rect" id="node3" href="classutils.html#ad1594a7c6cd34bba8e878fe09837e5d8" title="Checks the color that has to be used on the output report. " alt="" coords="296,56,439,83"/>
<area shape="rect" id="node4" href="classutils.html#af7813b2754509dbf0b9b15f4d85f8401" title="Looks for a pattern. " alt="" coords="487,56,625,83"/>
</map>
</div>

</div>
</div>
<a id="a3377e8d600e1cbb9e65adf2995c877f9"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a3377e8d600e1cbb9e65adf2995c877f9">&#9670;&nbsp;</a></span>alignmentSummaryHTML()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool newAlignment::alignmentSummaryHTML </td>
          <td>(</td>
          <td class="paramtype">char *&#160;</td>
          <td class="paramname"><em>destFile</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>residues</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>seqs</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int *&#160;</td>
          <td class="paramname"><em>selectedRes</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int *&#160;</td>
          <td class="paramname"><em>selectedSeq</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">float *&#160;</td>
          <td class="paramname"><em>consValues</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Method to report the trimming results in HTML.<br />
It contains information about the original and trimmed alignments, showing which residues and sequences have been kept and which ones have been removed on the output. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">destFile</td><td>Filename where to save the results. </td></tr>
    <tr><td class="paramname">residues</td><td>Number of residues present on the trimmed alignment. </td></tr>
    <tr><td class="paramname">seqs</td><td>Number of sequences present on the trimmed alignment. </td></tr>
    <tr><td class="paramname">selectedRes</td><td>Vector containing the indexes of the original alignment residues that are kept on the trimmed alignment. </td></tr>
    <tr><td class="paramname">selectedSeq</td><td>Vector containing the indexes of the original alignment sequences that are kept on the trimmed alignment. </td></tr>
    <tr><td class="paramname">consValues</td><td>Vector containing the consistency values of columns.<br />
 The consistency may be a null pointer, meaning we don't want to report consistency values. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd><b>True</b> if everything went ok. <br />
 <b>False</b> if file couldn't be open or alignment is not aligned. </dd></dl>

<p class="definition">Definition at line <a class="el" href="new_alignment_8cpp_source.html#l01066">1066</a> of file <a class="el" href="new_alignment_8cpp_source.html">newAlignment.cpp</a>.</p>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classnew_alignment_a3377e8d600e1cbb9e65adf2995c877f9_cgraph.png" border="0" usemap="#classnew_alignment_a3377e8d600e1cbb9e65adf2995c877f9_cgraph" alt=""/></div>
<map name="classnew_alignment_a3377e8d600e1cbb9e65adf2995c877f9_cgraph" id="classnew_alignment_a3377e8d600e1cbb9e65adf2995c877f9_cgraph">
<area shape="rect" id="node2" href="classutils.html#a40640e1942f00797e9fe8284e8fbdb14" title="Maximum of two numbers method. " alt="" coords="345,5,424,32"/>
<area shape="rect" id="node3" href="classutils.html#ad1594a7c6cd34bba8e878fe09837e5d8" title="Checks the color that has to be used on the output report. " alt="" coords="313,56,456,83"/>
<area shape="rect" id="node4" href="classutils.html#af7813b2754509dbf0b9b15f4d85f8401" title="Looks for a pattern. " alt="" coords="504,56,643,83"/>
</map>
</div>

</div>
</div>
<a id="ad916722e4ee892ee9f5cf85f0bdb80aa"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ad916722e4ee892ee9f5cf85f0bdb80aa">&#9670;&nbsp;</a></span>calculateColIdentity()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void newAlignment::calculateColIdentity </td>
          <td>(</td>
          <td class="paramtype">float *&#160;</td>
          <td class="paramname"><em>columnIdentity</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Method that calculates the columns identity value.<br />
This is, the frequency of the most present element in the column, being residue, indetermination and gap allowed. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir">[out]</td><td class="paramname">columnIdentity</td><td>Vector to fill with identities for each column. </td></tr>
  </table>
  </dd>
</dl>

<p class="definition">Definition at line <a class="el" href="new_alignment_8cpp_source.html#l00981">981</a> of file <a class="el" href="new_alignment_8cpp_source.html">newAlignment.cpp</a>.</p>

</div>
</div>
<a id="ae8501233fe4f4ef6bfe30c99c744c7c2"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ae8501233fe4f4ef6bfe30c99c744c7c2">&#9670;&nbsp;</a></span>checkCorrespondence()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool newAlignment::checkCorrespondence </td>
          <td>(</td>
          <td class="paramtype">string *&#160;</td>
          <td class="paramname"><em>names</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int *&#160;</td>
          <td class="paramname"><em>lenghts</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>totalInputSequences</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>multiple</em> = <code>1</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Function to check CDS file.<br />
It checks if sequences of input alignment are all present on the CDS file.<br />
If nucleotide sequence is larger than protein sequence length * 3, warns about it and cuts the nucleotide sequence<br />
If sequence has indetermination symbols, it warns about it.<br />
If nucleotide sequence is smaller than protein sequence * 3, it adds some 'N' at the end of the nucleotide sequence. </p>
<dl class="todo"><dt><b><a class="el" href="todo.html#_todo000004">Todo:</a></b></dt><dd>multiple argument is needed? </dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">names</td><td>Vector containing the names to check. </td></tr>
    <tr><td class="paramname">lenghts</td><td>Vector containing the length of each sequence. </td></tr>
    <tr><td class="paramname">totalInputSequences</td><td>Number of sequences present. </td></tr>
    <tr><td class="paramname">multiple</td><td>Multiplies the length of each sequence by this number. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd><b>True</b> if all went right.<br />
 <b>False</b> if there is a sequence in the alignment but not in the names vector. </dd></dl>

<p class="definition">Definition at line <a class="el" href="new_alignment_8cpp_source.html#l00698">698</a> of file <a class="el" href="new_alignment_8cpp_source.html">newAlignment.cpp</a>.</p>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classnew_alignment_ae8501233fe4f4ef6bfe30c99c744c7c2_cgraph.png" border="0" usemap="#classnew_alignment_ae8501233fe4f4ef6bfe30c99c744c7c2_cgraph" alt=""/></div>
<map name="classnew_alignment_ae8501233fe4f4ef6bfe30c99c744c7c2_cgraph" id="classnew_alignment_ae8501233fe4f4ef6bfe30c99c744c7c2_cgraph">
<area shape="rect" id="node2" href="classutils.html#a1dce160519187ef2c6e14beb93a2050d" title="Removes a determined char from the string. " alt="" coords="296,5,451,32"/>
<area shape="rect" id="node3" href="classutils.html#ab04f5770187b0a8c36c3c8ca55f1e67f" title="Minimum of two numbers method. " alt="" coords="336,56,411,83"/>
</map>
</div>

</div>
</div>
<a id="acbeb320f0f70222f81a906ccfc33a016"></a>
<h2 class="memtitle"><span class="permalink"><a href="#acbeb320f0f70222f81a906ccfc33a016">&#9670;&nbsp;</a></span>fillMatrices()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool newAlignment::fillMatrices </td>
          <td>(</td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>aligned</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Method to intialize data that has been modified on the alignment.<br />
The method checks if the sequences have been correctly loaded and are free of errors.<br />
It checks if sequences contain unknown characters, sets the isAligned flag and initializes 'saveResidues' and 'saveSequences', depending on the sizes of the sequences (wether or not they have the same length). </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">aligned</td><td>Flag to make the method check if the alignment is aligned or not. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section note"><dt>Note</dt><dd>Even with the aligned flag set to false, if the alignment is aligned it will initialize the variables 'saveResidues' and 'saveSequences', allowing the alignment to be trimmed. </dd></dl>
<dl class="section return"><dt>Returns</dt><dd><b>True</b> if the alignment information was ok. <br />
 <b>False</b> if there was a problem.<br />
This could happen if the sequences contain any unknown character or if aligned flag is set up to corrent, if the sequences have variable sizes. </dd></dl>

<p class="definition">Definition at line <a class="el" href="new_alignment_8cpp_source.html#l00781">781</a> of file <a class="el" href="new_alignment_8cpp_source.html">newAlignment.cpp</a>.</p>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classnew_alignment_acbeb320f0f70222f81a906ccfc33a016_icgraph.png" border="0" usemap="#classnew_alignment_acbeb320f0f70222f81a906ccfc33a016_icgraph" alt=""/></div>
<map name="classnew_alignment_acbeb320f0f70222f81a906ccfc33a016_icgraph" id="classnew_alignment_acbeb320f0f70222f81a906ccfc33a016_icgraph">
<area shape="rect" id="node2" href="class_cleaner.html#ad075f0ef730c8023cf8a0e35e7e1e53a" title="Method to clean an alignment. &#160;It removes sequences that overpass or equals a certain threshold..." alt="" coords="231,157,407,184"/>
<area shape="rect" id="node9" href="class_cleaner.html#a740bce047dce09be0f6bd1da9f826da3" title="Method to clean an alignment. It carries out strict and strictplus. It removes sequences that overpas..." alt="" coords="251,309,388,336"/>
<area shape="rect" id="node11" href="class_cleaner.html#a74d3049d4997da5e7c6c51cb19b540e8" title="Method to clean sequences based on a minimum overlap threshold. The method selects a combination of p..." alt="" coords="232,360,407,387"/>
<area shape="rect" id="node13" href="class_cleaner.html#a991bf55a7d36b305d67d639ebf028348" title="Method to select the most representative sequence (the longest one) for each cluster from the input a..." alt="" coords="243,411,395,437"/>
<area shape="rect" id="node14" href="class_cleaner.html#a3d1c1309bd1f0908ef8c8ef0415fb4cb" title="Method to remove columns, expressed as ranges. " alt="" coords="234,461,405,488"/>
<area shape="rect" id="node15" href="class_cleaner.html#af27e657bf4874a21b26d9a8217c1e643" title="Method to remove sequences, expressed as ranges. " alt="" coords="228,512,411,539"/>
<area shape="rect" id="node3" href="class_cleaner.html#ac3be9d53018d9146d6dc6cf7ff89f96d" title="Method to clean an alignment based on the gap distribution values. Column blocks that don&#39;t have a mi..." alt="" coords="485,5,623,32"/>
<area shape="rect" id="node4" href="class_cleaner.html#a694d53fdd5d1ac9455858e48537af3d0" title="Method to clean an alignment based on the similarity distribution values. " alt="" coords="463,56,645,83"/>
<area shape="rect" id="node5" href="class_cleaner.html#a1ac5ba504932526d77575365557a4d64" title="Method to clean an alignment based on the similarity and gaps distribution values. . " alt="" coords="501,107,607,133"/>
<area shape="rect" id="node6" href="class_cleaner.html#ae2328fe8c189e2827b46e5dfb8e3deb3" title="Method to clean an alignment based on consistency values obtained from a dataset of alignments..." alt="" coords="464,157,644,184"/>
<area shape="rect" id="node7" href="class_cleaner.html#a1ea9656823727f9ec237c82f9fb529cf" title="Method that carries the gappyout approach. This methods calculates the slope in gaps distribution on ..." alt="" coords="474,208,634,235"/>
<area shape="rect" id="node8" href="class_cleaner.html#a08e6db52c33baccce53a33d844597d7b" title="Method to remove columns composed only by gaps This method is specially useful when we remove missali..." alt="" coords="469,259,639,285"/>
<area shape="rect" id="node10" href="class_cleaner.html#a068b219f913adba5c917a231cff7e437" title="Method to clean an alignment. It carries out strict and strictplus. The method: " alt="" coords="459,309,649,336"/>
<area shape="rect" id="node12" href="class_cleaner.html#a095db8160a11a792a3feade870a24fd7" title="Method to remove sequences missaligned with the rest of sequences in the alignment. For each residue in the sequence, it tests it&#39;s similarity. If the similarity of that residue is higher than overlapColumn value, it counts as a hit for the sequence. After calculating the number of hits for the sequence, it removes the sequence if it has a proportion hits/residues lower tan minimumOverlap. " alt="" coords="464,360,644,387"/>
</map>
</div>

</div>
</div>
<a id="a3de39c62c08a265178f1fa7df76c3ab7"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a3de39c62c08a265178f1fa7df76c3ab7">&#9670;&nbsp;</a></span>fillNewDataStructure() <span class="overload">[1/2]</span></h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void newAlignment::fillNewDataStructure </td>
          <td>(</td>
          <td class="paramtype">string *&#160;</td>
          <td class="paramname"><em>newMatrix</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">string *&#160;</td>
          <td class="paramname"><em>newNames</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Method that copies information of the alignment to the pointers given. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir">[out]</td><td class="paramname">newMatrix</td><td>Sequences vector to fill. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">newNames</td><td>Names vector to fill. </td></tr>
  </table>
  </dd>
</dl>

<p class="definition">Definition at line <a class="el" href="new_alignment_8cpp_source.html#l00452">452</a> of file <a class="el" href="new_alignment_8cpp_source.html">newAlignment.cpp</a>.</p>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classnew_alignment_a3de39c62c08a265178f1fa7df76c3ab7_icgraph.png" border="0" usemap="#classnew_alignment_a3de39c62c08a265178f1fa7df76c3ab7_icgraph" alt=""/></div>
<map name="classnew_alignment_a3de39c62c08a265178f1fa7df76c3ab7_icgraph" id="classnew_alignment_a3de39c62c08a265178f1fa7df76c3ab7_icgraph">
<area shape="rect" id="node2" href="class_cleaner.html#ad075f0ef730c8023cf8a0e35e7e1e53a" title="Method to clean an alignment. &#160;It removes sequences that overpass or equals a certain threshold..." alt="" coords="233,157,409,184"/>
<area shape="rect" id="node9" href="class_cleaner.html#a740bce047dce09be0f6bd1da9f826da3" title="Method to clean an alignment. It carries out strict and strictplus. It removes sequences that overpas..." alt="" coords="253,309,390,336"/>
<area shape="rect" id="node11" href="class_cleaner.html#a74d3049d4997da5e7c6c51cb19b540e8" title="Method to clean sequences based on a minimum overlap threshold. The method selects a combination of p..." alt="" coords="234,360,409,387"/>
<area shape="rect" id="node3" href="class_cleaner.html#ac3be9d53018d9146d6dc6cf7ff89f96d" title="Method to clean an alignment based on the gap distribution values. Column blocks that don&#39;t have a mi..." alt="" coords="484,5,621,32"/>
<area shape="rect" id="node4" href="class_cleaner.html#a694d53fdd5d1ac9455858e48537af3d0" title="Method to clean an alignment based on the similarity distribution values. " alt="" coords="461,56,644,83"/>
<area shape="rect" id="node5" href="class_cleaner.html#a1ac5ba504932526d77575365557a4d64" title="Method to clean an alignment based on the similarity and gaps distribution values. . " alt="" coords="499,107,606,133"/>
<area shape="rect" id="node6" href="class_cleaner.html#ae2328fe8c189e2827b46e5dfb8e3deb3" title="Method to clean an alignment based on consistency values obtained from a dataset of alignments..." alt="" coords="463,157,643,184"/>
<area shape="rect" id="node7" href="class_cleaner.html#a1ea9656823727f9ec237c82f9fb529cf" title="Method that carries the gappyout approach. This methods calculates the slope in gaps distribution on ..." alt="" coords="473,208,633,235"/>
<area shape="rect" id="node8" href="class_cleaner.html#a08e6db52c33baccce53a33d844597d7b" title="Method to remove columns composed only by gaps This method is specially useful when we remove missali..." alt="" coords="468,259,637,285"/>
<area shape="rect" id="node10" href="class_cleaner.html#a068b219f913adba5c917a231cff7e437" title="Method to clean an alignment. It carries out strict and strictplus. The method: " alt="" coords="457,309,648,336"/>
<area shape="rect" id="node12" href="class_cleaner.html#a095db8160a11a792a3feade870a24fd7" title="Method to remove sequences missaligned with the rest of sequences in the alignment. For each residue in the sequence, it tests it&#39;s similarity. If the similarity of that residue is higher than overlapColumn value, it counts as a hit for the sequence. After calculating the number of hits for the sequence, it removes the sequence if it has a proportion hits/residues lower tan minimumOverlap. " alt="" coords="463,360,643,387"/>
</map>
</div>

</div>
</div>
<a id="a54b8820f397c14cf4d8d5bc8a89daf24"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a54b8820f397c14cf4d8d5bc8a89daf24">&#9670;&nbsp;</a></span>fillNewDataStructure() <span class="overload">[2/2]</span></h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void newAlignment::fillNewDataStructure </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structnew_values.html">newValues</a> *&#160;</td>
          <td class="paramname"><em>data</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Method that copies information of the alignment to the pointers given. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir">[out]</td><td class="paramname">data</td><td><a class="el" href="structnew_values.html">newValues</a> struct to fill with sequences and sequences names. </td></tr>
  </table>
  </dd>
</dl>

<p class="definition">Definition at line <a class="el" href="new_alignment_8cpp_source.html#l00470">470</a> of file <a class="el" href="new_alignment_8cpp_source.html">newAlignment.cpp</a>.</p>

</div>
</div>
<a id="ae1ee3ea76e1cb268097f1df0c465d6f5"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ae1ee3ea76e1cb268097f1df0c465d6f5">&#9670;&nbsp;</a></span>getAlignmentType()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int newAlignment::getAlignmentType </td>
          <td>(</td>
          <td class="paramtype">void&#160;</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Alignment type getter. </p>
<dl class="section return"><dt>Returns</dt><dd>Int representing the alignment type.<br />
<ul>
<li>DNAType 1</li>
<li>RNAType 2</li>
<li>AAType 3 </li>
</ul>
</dd></dl>

<p class="definition">Definition at line <a class="el" href="new_alignment_8cpp_source.html#l00434">434</a> of file <a class="el" href="new_alignment_8cpp_source.html">newAlignment.cpp</a>.</p>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classnew_alignment_ae1ee3ea76e1cb268097f1df0c465d6f5_cgraph.png" border="0" usemap="#classnew_alignment_ae1ee3ea76e1cb268097f1df0c465d6f5_cgraph" alt=""/></div>
<map name="classnew_alignment_ae1ee3ea76e1cb268097f1df0c465d6f5_cgraph" id="classnew_alignment_ae1ee3ea76e1cb268097f1df0c465d6f5_cgraph">
<area shape="rect" id="node2" href="classutils.html#a1f1dcbfa7276e54acec72bdbae1b1255" title="Checks an alignment type. " alt="" coords="271,5,447,32"/>
</map>
</div>

</div>
</div>
<a id="a7a471f9393fa1da913bd9a65cc96a9f1"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a7a471f9393fa1da913bd9a65cc96a9f1">&#9670;&nbsp;</a></span>getBlockSize()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int newAlignment::getBlockSize </td>
          <td>(</td>
          <td class="paramtype">void&#160;</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>blockSize getter. </p>
<dl class="section return"><dt>Returns</dt><dd>blockSize. </dd></dl>

<p class="definition">Definition at line <a class="el" href="new_alignment_8cpp_source.html#l00385">385</a> of file <a class="el" href="new_alignment_8cpp_source.html">newAlignment.cpp</a>.</p>

</div>
</div>
<a id="a882c53e08d4435c50328d5bf7b10e2d1"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a882c53e08d4435c50328d5bf7b10e2d1">&#9670;&nbsp;</a></span>getCorrespResidues()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int * newAlignment::getCorrespResidues </td>
          <td>(</td>
          <td class="paramtype">void&#160;</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Method that returns the correspondence between old and new alignment residues. </p>
<dl class="section return"><dt>Returns</dt><dd>Vector that represents which residues are kept (!= -1) and which are rejected (== -1) </dd></dl>

<p class="definition">Definition at line <a class="el" href="new_alignment_8cpp_source.html#l00440">440</a> of file <a class="el" href="new_alignment_8cpp_source.html">newAlignment.cpp</a>.</p>

</div>
</div>
<a id="adf85c31144b199b65e24b98b6c755299"></a>
<h2 class="memtitle"><span class="permalink"><a href="#adf85c31144b199b65e24b98b6c755299">&#9670;&nbsp;</a></span>getCorrespSequences()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int * newAlignment::getCorrespSequences </td>
          <td>(</td>
          <td class="paramtype">void&#160;</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Method that returns the correspondence between old and new alignment sequences. </p>
<dl class="section return"><dt>Returns</dt><dd>Vector that represents which sequences are kept (!= -1) and which are rejected (== -1) </dd></dl>

<p class="definition">Definition at line <a class="el" href="new_alignment_8cpp_source.html#l00444">444</a> of file <a class="el" href="new_alignment_8cpp_source.html">newAlignment.cpp</a>.</p>

</div>
</div>
<a id="a906c64b2a7b38a2b45308d67429ff1f2"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a906c64b2a7b38a2b45308d67429ff1f2">&#9670;&nbsp;</a></span>getNumAminos()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int newAlignment::getNumAminos </td>
          <td>(</td>
          <td class="paramtype">void&#160;</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Residues getter. </p>
<dl class="section return"><dt>Returns</dt><dd>Number of residues in the alignment. </dd></dl>

<p class="definition">Definition at line <a class="el" href="new_alignment_8cpp_source.html#l00364">364</a> of file <a class="el" href="new_alignment_8cpp_source.html">newAlignment.cpp</a>.</p>

</div>
</div>
<a id="a71e1be2c77493d5094a4d7741e5ea26a"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a71e1be2c77493d5094a4d7741e5ea26a">&#9670;&nbsp;</a></span>getNumSpecies()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int newAlignment::getNumSpecies </td>
          <td>(</td>
          <td class="paramtype">void&#160;</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Getter for the number of species. </p>
<dl class="section return"><dt>Returns</dt><dd>Number of sequences in the alignment. </dd></dl>

<p class="definition">Definition at line <a class="el" href="new_alignment_8cpp_source.html#l00357">357</a> of file <a class="el" href="new_alignment_8cpp_source.html">newAlignment.cpp</a>.</p>

</div>
</div>
<a id="ac3d5c9670904bba45707746640aac382"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ac3d5c9670904bba45707746640aac382">&#9670;&nbsp;</a></span>getSequenceNameOrder()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool newAlignment::getSequenceNameOrder </td>
          <td>(</td>
          <td class="paramtype">string *&#160;</td>
          <td class="paramname"><em>names</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int *&#160;</td>
          <td class="paramname"><em>orderVector</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Method to map two sets of sequences, own and external.<br />
The method accepts a vector of names, and test them against the alignment owned sequences.<br />
If the sequence is present, its order on the alignment sequences vector will be stored in orderVector, in the same position it's name was in the names vector.<br />
. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir"></td><td class="paramname">names</td><td>Vector of sequences names to map. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">orderVector</td><td>Vector of orders to fill. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd><b>True</b> if both sets have the same sequences.<br />
 <b>False</b> otherwise. </dd></dl>

<p class="definition">Definition at line <a class="el" href="new_alignment_8cpp_source.html#l00409">409</a> of file <a class="el" href="new_alignment_8cpp_source.html">newAlignment.cpp</a>.</p>

</div>
</div>
<a id="a0db3e76f82a71c7a2fe5b9279f8fa310"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a0db3e76f82a71c7a2fe5b9279f8fa310">&#9670;&nbsp;</a></span>getSequences() <span class="overload">[1/3]</span></h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void newAlignment::getSequences </td>
          <td>(</td>
          <td class="paramtype">string *&#160;</td>
          <td class="paramname"><em>names</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Getter for the sequences names. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir">[out]</td><td class="paramname">names</td><td>Vector of sequences names to fill. </td></tr>
  </table>
  </dd>
</dl>

<p class="definition">Definition at line <a class="el" href="new_alignment_8cpp_source.html#l00389">389</a> of file <a class="el" href="new_alignment_8cpp_source.html">newAlignment.cpp</a>.</p>

</div>
</div>
<a id="af32ab6b3b73c5ad37c9f1c796a4fbc44"></a>
<h2 class="memtitle"><span class="permalink"><a href="#af32ab6b3b73c5ad37c9f1c796a4fbc44">&#9670;&nbsp;</a></span>getSequences() <span class="overload">[2/3]</span></h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void newAlignment::getSequences </td>
          <td>(</td>
          <td class="paramtype">string *&#160;</td>
          <td class="paramname"><em>names</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int *&#160;</td>
          <td class="paramname"><em>lengths</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Getter for the sequences names and its lenghts. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir">[out]</td><td class="paramname">names</td><td>Vector of sequences names to fill. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">lengths</td><td>Vector of lenghts to fill. </td></tr>
  </table>
  </dd>
</dl>

<p class="definition">Definition at line <a class="el" href="new_alignment_8cpp_source.html#l00394">394</a> of file <a class="el" href="new_alignment_8cpp_source.html">newAlignment.cpp</a>.</p>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classnew_alignment_af32ab6b3b73c5ad37c9f1c796a4fbc44_cgraph.png" border="0" usemap="#classnew_alignment_af32ab6b3b73c5ad37c9f1c796a4fbc44_cgraph" alt=""/></div>
<map name="classnew_alignment_af32ab6b3b73c5ad37c9f1c796a4fbc44_cgraph" id="classnew_alignment_af32ab6b3b73c5ad37c9f1c796a4fbc44_cgraph">
<area shape="rect" id="node2" href="classutils.html#a1dce160519187ef2c6e14beb93a2050d" title="Removes a determined char from the string. " alt="" coords="248,5,403,32"/>
</map>
</div>

</div>
</div>
<a id="a2fc0d691c0cce400f007446cc162704e"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a2fc0d691c0cce400f007446cc162704e">&#9670;&nbsp;</a></span>getSequences() <span class="overload">[3/3]</span></h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void newAlignment::getSequences </td>
          <td>(</td>
          <td class="paramtype">string *&#160;</td>
          <td class="paramname"><em>names</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">string *&#160;</td>
          <td class="paramname"><em>sequences</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int *&#160;</td>
          <td class="paramname"><em>lenghts</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Getter for the sequences, its names and lenghts. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir">[out]</td><td class="paramname">names</td><td>Vector of sequences names to fill. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">sequences</td><td>Vector of sequences to fill. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">lenghts</td><td>Vector of lenghts to fill. </td></tr>
  </table>
  </dd>
</dl>

<p class="definition">Definition at line <a class="el" href="new_alignment_8cpp_source.html#l00401">401</a> of file <a class="el" href="new_alignment_8cpp_source.html">newAlignment.cpp</a>.</p>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classnew_alignment_a2fc0d691c0cce400f007446cc162704e_cgraph.png" border="0" usemap="#classnew_alignment_a2fc0d691c0cce400f007446cc162704e_cgraph" alt=""/></div>
<map name="classnew_alignment_a2fc0d691c0cce400f007446cc162704e_cgraph" id="classnew_alignment_a2fc0d691c0cce400f007446cc162704e_cgraph">
<area shape="rect" id="node2" href="classutils.html#a1dce160519187ef2c6e14beb93a2050d" title="Removes a determined char from the string. " alt="" coords="248,5,403,32"/>
</map>
</div>

</div>
</div>
<a id="a3d1ff48743c38351c7e755f94baebf03"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a3d1ff48743c38351c7e755f94baebf03">&#9670;&nbsp;</a></span>getTranslationCDS()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classnew_alignment.html">newAlignment</a> * newAlignment::getTranslationCDS </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>newResidues</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>newSequences</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int *&#160;</td>
          <td class="paramname"><em>columnsToKeep</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">string *&#160;</td>
          <td class="paramname"><em>oldSequencesNames</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="classsequences_matrix.html">sequencesMatrix</a> *&#160;</td>
          <td class="paramname"><em>sequenceMatrix</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="classnew_alignment.html">newAlignment</a> *&#160;</td>
          <td class="paramname"><em>proteinAlignment</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Method to obtain a DNA alignment from a Protein alignment. </p>
<dl class="todo"><dt><b><a class="el" href="todo.html#_todo000003">Todo:</a></b></dt><dd>Check the need for so many arguments. </dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">newResidues</td><td>Number of residues of the new alignment, it should be 3 x old.residNumber </td></tr>
    <tr><td class="paramname">newSequences</td><td>Number of residues of the new alignment, it should be the same as old.sequenNumber </td></tr>
    <tr><td class="paramname">columnsToKeep</td><td>Columns to keep ? IDK really. </td></tr>
    <tr><td class="paramname">oldSequencesNames</td><td>Sequences names. </td></tr>
    <tr><td class="paramname">sequenceMatrix</td><td>sequence Matrix. </td></tr>
    <tr><td class="paramname">proteinAlignment</td><td>original alignment. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>Pointer to new alignment. </dd></dl>

<p class="definition">Definition at line <a class="el" href="new_alignment_8cpp_source.html#l00244">244</a> of file <a class="el" href="new_alignment_8cpp_source.html">newAlignment.cpp</a>.</p>

</div>
</div>
<a id="ad56533a40875ddbe8aa65ee391840464"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ad56533a40875ddbe8aa65ee391840464">&#9670;&nbsp;</a></span>isFileAligned()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool newAlignment::isFileAligned </td>
          <td>(</td>
          <td class="paramtype">void&#160;</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>isAligned getter. </p>
<dl class="section return"><dt>Returns</dt><dd>Wheter the alignment is aligned or not. </dd></dl>

<p class="definition">Definition at line <a class="el" href="new_alignment_8cpp_source.html#l00448">448</a> of file <a class="el" href="new_alignment_8cpp_source.html">newAlignment.cpp</a>.</p>

</div>
</div>
<a id="ad50ba126247276cd0bf89119e439c483"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ad50ba126247276cd0bf89119e439c483">&#9670;&nbsp;</a></span>prepareCodingSequence()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool newAlignment::prepareCodingSequence </td>
          <td>(</td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>splitByStopCodon</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>ignoreStopCodon</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="classnew_alignment.html">newAlignment</a> *&#160;</td>
          <td class="paramname"><em>proteinAlignment</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Method to check if the CDS file is correct.<br />
Based on nature of residues: DNA/RNA (Most of the residues)<br />
There is no gaps on the whole dataset.<br />
Each sequence is a multiple of 3.<br />
<br />
It will also remove or split stop codons depending on the flags passed. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">splitByStopCodon</td><td>Flag that informs the method to split sequences if it finds any stop codon. </td></tr>
    <tr><td class="paramname">ignoreStopCodon</td><td>Flag that informs the method to stop reading sequence if it finds any stop codon. </td></tr>
    <tr><td class="paramname">proteinAlignment</td><td>Alignment containing protein sequences which names contains all names in the current alignment. </td></tr>
  </table>
  </dd>
</dl>

<p class="definition">Definition at line <a class="el" href="new_alignment_8cpp_source.html#l00489">489</a> of file <a class="el" href="new_alignment_8cpp_source.html">newAlignment.cpp</a>.</p>

</div>
</div>
<a id="a3b75055480c60265ea57ca557213fe06"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a3b75055480c60265ea57ca557213fe06">&#9670;&nbsp;</a></span>printAlignmentInfo()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void newAlignment::printAlignmentInfo </td>
          <td>(</td>
          <td class="paramtype">ostream &amp;&#160;</td>
          <td class="paramname"><em>output</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Print information about sequences number, average sequence length, maximum and minimum sequences length, etc. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">output</td><td>Output stream. </td></tr>
  </table>
  </dd>
</dl>

<p class="definition">Definition at line <a class="el" href="new_alignment_8cpp_source.html#l00855">855</a> of file <a class="el" href="new_alignment_8cpp_source.html">newAlignment.cpp</a>.</p>

</div>
</div>
<a id="a42894635dca6466308d9e21cec14d365"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a42894635dca6466308d9e21cec14d365">&#9670;&nbsp;</a></span>setBlockSize()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void newAlignment::setBlockSize </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>blockSize</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>blockSize Setter. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">blockSize</td><td>New value. </td></tr>
  </table>
  </dd>
</dl>

<p class="definition">Definition at line <a class="el" href="new_alignment_8cpp_source.html#l00377">377</a> of file <a class="el" href="new_alignment_8cpp_source.html">newAlignment.cpp</a>.</p>

</div>
</div>
<a id="aa77c80b10d05f3a069a5f63223e13ca2"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aa77c80b10d05f3a069a5f63223e13ca2">&#9670;&nbsp;</a></span>setKeepSeqsHeaderFlag()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">newAlignment::setKeepSeqsHeaderFlag </td>
          <td>(</td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>newFlagValue</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Keep Header setter. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">newFlagValue</td><td>New value </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a id="a29f7f52115fd4ecf13120e54c71b1a1a"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a29f7f52115fd4ecf13120e54c71b1a1a">&#9670;&nbsp;</a></span>setKeepSequencesFlag()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void newAlignment::setKeepSequencesFlag </td>
          <td>(</td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>newFlagValue</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Keep Sequences setter. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">newFlagValue</td><td>New value </td></tr>
  </table>
  </dd>
</dl>

<p class="definition">Definition at line <a class="el" href="new_alignment_8cpp_source.html#l00381">381</a> of file <a class="el" href="new_alignment_8cpp_source.html">newAlignment.cpp</a>.</p>

</div>
</div>
<a id="a8dd7e7438161fcaca8c8f35e63c5079e"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a8dd7e7438161fcaca8c8f35e63c5079e">&#9670;&nbsp;</a></span>setWindowsSize()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void newAlignment::setWindowsSize </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>ghWindow</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>shWindow</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Windows setter. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">ghWindow</td><td>Half the Gap Window. </td></tr>
    <tr><td class="paramname">shWindow</td><td>Half the similarity Window. </td></tr>
  </table>
  </dd>
</dl>

<p class="definition">Definition at line <a class="el" href="new_alignment_8cpp_source.html#l00372">372</a> of file <a class="el" href="new_alignment_8cpp_source.html">newAlignment.cpp</a>.</p>

</div>
</div>
<h2 class="groupheader">Member Data Documentation</h2>
<a id="ada2fb94cccdf0760c22dab39d278680c"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ada2fb94cccdf0760c22dab39d278680c">&#9670;&nbsp;</a></span>dataType</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">newAlignment::dataType</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Number that represents the type of the alignment. </p>
<dl class="todo"><dt><b><a class="el" href="todo.html#_todo000002">Todo:</a></b></dt><dd>We should change this from a define to a enum. </dd></dl>

<p class="definition">Definition at line <a class="el" href="new_alignment_8h_source.html#l00095">95</a> of file <a class="el" href="new_alignment_8h_source.html">newAlignment.h</a>.</p>

</div>
</div>
<a id="a8b328dfd8e7d3304cb144f2c2114ab5c"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a8b328dfd8e7d3304cb144f2c2114ab5c">&#9670;&nbsp;</a></span>SequencesMatrix</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">newAlignment::SequencesMatrix</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>SequencesMatrix submodule<br />
. </p>
<dl class="todo"><dt><b><a class="el" href="todo.html#_todo000001">Todo:</a></b></dt><dd>Give a good explanation of the module. </dd></dl>

<p class="definition">Definition at line <a class="el" href="new_alignment_8h_source.html#l00083">83</a> of file <a class="el" href="new_alignment_8h_source.html">newAlignment.h</a>.</p>

</div>
</div>
<hr/>The documentation for this class was generated from the following files:<ul>
<li>include/<a class="el" href="new_alignment_8h_source.html">newAlignment.h</a></li>
<li>include/newAlignment.dox</li>
<li>source/<a class="el" href="new_alignment_8cpp_source.html">newAlignment.cpp</a></li>
</ul>
</div><!-- contents -->
</div><!-- doc-content -->
<!-- start footer part -->
<div id="nav-path" class="navpath"><!-- id is needed for treeview function! -->
  <ul>
    <li class="navelem"><a class="el" href="classnew_alignment.html">newAlignment</a></li>
    <li class="footer">Generated by
    <a href="http://www.doxygen.org/index.html">
    <img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.8.13 </li>
  </ul>
</div>
</body>
</html>
