<!-- HTML header for doxygen 1.8.8-->
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
    <meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
    <meta http-equiv="X-UA-Compatible" content="IE=9"/>
    <meta name="generator" content="Doxygen 1.8.15"/>
    <title>trimAl: trimAl Main Page</title>
    <link href="tabs.css" rel="stylesheet" type="text/css"/>
    <script type="text/javascript" src="jquery.js"></script>
    <script type="text/javascript" src="dynsections.js"></script>
    <link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtreedata.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
  $(document).ready(initResizable);
/* @license-end */</script>
    <link href="doxygen.css" rel="stylesheet" type="text/css" />
    <link href="CustomExtra.css" rel="stylesheet" type="text/css"/>
    <script type="text/javascript">
        function myFunction(ID) {
            var divs = document.getElementsByName(ID);
            for (var x of divs) {
                if (x.style.display === "none") {
                    x.style.display = "block";
                } else {
                    x.style.display = "none";
                }
            }
        } 
    </script>
</head>
<body>
    <div id="top"><!-- do not remove this div, it is closed by doxygen! -->
        <div id="titlearea">
            <table cellspacing="0" cellpadding="0" width=100%>
                <tbody>
                    <tr style="height: 12px;">
                        <td id="projectlogo"><img alt="Logo" src="Logo.png"/></td>
                    </tr>
                </tbody>
            </table>
            <div class="header">
                <br>
            </div>
        </div>
        <!-- end header part -->
    <!-- Generated by Doxygen 1.8.15 -->
</div><!-- top -->
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
      <div id="nav-sync" class="sync"></div>
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(document).ready(function(){initNavTree('index.html','');});
/* @license-end */
</script>
<div id="doc-content">
<div class="PageDoc"><div class="header">
  <div class="headertitle">
<div class="title">trimAl Main Page </div>  </div>
</div><!--header-->
<div class="contents">
<div class="textblock"><p> 
<a class="badge-align" href="https://www.codacy.com/app/vicfernrod/trimal?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=Vicfero/trimal&amp;utm_campaign=Badge_Grade">
    <img src="https://api.codacy.com/project/badge/Grade/ba6b9c6a765b4383988cb14a847d9127"/>
</a>

<a class="badge-align" href="https://github.com/scapella/trimal">
    <img src="https://img.shields.io/badge/ -Github-lightgrey.svg?logo=github&logoColor=white"/>
</a>
</p>
<h1><a class="anchor" id="introduction_sec"></a>
Introduction</h1>
<p>This is <b>trimAl's</b> information page. You can also find information related to <a class="el" href="read_al.html">readAl </a>, a MSA format conversor.<br />
<b>trimAl</b> is a tool for the automated removal of spurious sequences or poorly aligned regions from a multiple sequence alignment.<br />
<br />
 <b>trimAl</b> can consider several parameters, alone or in multiple combinations, in order to select the most-reliable positions in the alignment. <br />
These include the <em> proportion of sequences with a gap</em>, the level of <em> residue similarity</em> and, if several alignments for the same set of sequences are provided, the <em> consistency level of columns among alignments</em>. <br />
Moreover, <b>trimAl</b> allows to <em> manually select a set of columns and sequences </em> to be removed from the alignment.<br />
<br />
 <b>trimAl</b> implements a series of automated algorithms that trim the alignment searching for optimum thresholds based on inherent characteristics of the input alignment, to be used so that the signal-to-noise ratio after alignment trimming phase is increased.<br />
<br />
 Among <b>trimAl's</b> additional features, <b>trimAl</b> allows getting the <em>complementary alignment</em> (columns that were trimmed), to <em> compute statistics </em> from the alignment, to <em> select the output file format </em>, to <em> get a summary </em> of <b>trimAl's</b> trimming in HTML and SVG formats, and many other options. <br />
 </p>
<h1><a class="anchor" id="installation_sec"></a>
Installation</h1>
<p>The simplest way to compile this package is: </p><pre class="fragment"> 
    1.- Move to the project folder
    
    2.- Configure the project. 
        This will also create the file "include/ReadWriteMS/formats_header.h" needed to handle formats.
            &gt; cmake .
    
    3.- Compile the project. 
            &gt; make
        
        It is possible to increase the speed of compilation by compiling in multiple threads.
        Specify the number of threads providing the -jN argument, where N is the number of threads.
            &gt; make -j2 
            &gt; make -j3 
            &gt; ...
 
    4.- [Optional] Move or copy binaries from folder './bin/' to '/usr/local/bin' or '/usr/bin'
        This will allow you to use directly trimAl without specifiying path to binary.
  
</pre><h1><a class="anchor" id="publication_sec"></a>
Publications</h1>
<p><a href="http://trimal.cgenomics.org/_media/trimal.2009.pdf">- trimAl: a tool for automated alignment trimming in large-scale phylogenetic analyses. </a> <br />
<a href="http://trimal.cgenomics.org/_media/supplmaterial.b.pdf">- Supplementary material </a> <br />
<a href="http://trimal.cgenomics.org/_media/otherbenchs.b.pdf">- Other benchmarks </a> <br />
</p>
<h1><a class="anchor" id="usage_sec"></a>
Usage</h1>
<pre class="fragment">trimAl 1.4.rev22 build[2015-05-21].
2009-2015. Salvador Capella-Gutierrez and Toni Gabaldón.

    trimAl webpage: http://trimal.cgenomics.org

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, the last available version.

Please cite:
    trimAl: a tool for automated Alignment  trimming in large-scale phylogenetic analyses.
    Salvador Capella-Gutierrez; Jose M. Silla-Martinez; Toni Gabaldon.
    Bioinformatics 2009, 25:1972-1973.

Basic usage
    trimal -in &lt;inputfile&gt; -out &lt;outputfile&gt; -(other options).

Common options (for a complete list please see the User Guide or visit http://trimal.cgenomics.org):

    -h --help                Print this information and show some examples.
    --version                Print the trimAl version.

    -in &lt;inputfile&gt;          Input file in several formats.
                                 Available input formats: [iformat]
    -out &lt;outputfile&gt;        Output alignment to file. (default stdout). (default input format)
                                 It will replace optional the tags [in]        -&gt; Original filename without extension.
                                                                   [format]    -&gt; Output's format name
                                                                   [extension] -&gt; Output's extension


    -compareset &lt;inputfile&gt;  Input list of paths for the files containing the alignments to compare.
    -forceselect &lt;inputfile&gt; Force selection of the given input file in the files comparison method.

    -backtrans &lt;inputfile&gt;   Use a Coding Sequences file to get a backtranslation for a given AA alignment
    -ignorestopcodon         Ignore stop codons in the input coding sequences
    -splitbystopcodon        Split input coding sequences up to first stop codon appearance

    -matrix &lt;inpufile&gt;       Input file for user-defined similarity matrix (default is Blosum62).

    -htmlout &lt;outputfile&gt;    Get a summary of trimal's work in an HTML file.
    -svgout &lt;outputfile&gt;     Get a summary of trimal's work in an SVG file.
    -sgvstats &lt;outputfile&gt;   Get a graphic in SVG showing stats of the indicators used in the cleanse of the alignment.

    -keepheader              Keep original sequence header including non-alphanumeric characters.
                                Only available for input FASTA format files. (future versions will extend this feature)

    -nbrf                    Output file in NBRF/PIR format
    -mega                    Output file in MEGA format
    -nexus                   Output file in NEXUS format
    -clustal                 Output file in CLUSTAL format

    -fasta                   Output file in FASTA format
    -fasta_m10               Output file in FASTA format. Sequences name length up to 10 characters.

    -phylip                  Output file in PHYLIP/PHYLIP4 format
    -phylip_m10              Output file in PHYLIP/PHYLIP4 format. Sequences name length up to 10 characters.
    -phylip_paml             Output file in PHYLIP format compatible with PAML
    -phylip_paml_m10         Output file in PHYLIP format compatible with PAML. Sequences name length up to 10 characters.
    -phylip3.2               Output file in PHYLIP3.2 format
    -phylip3.2_m10           Output file in PHYLIP3.2 format. Sequences name length up to 10 characters.

    -complementary           Get the complementary alignment.
    -colnumbering            Get the relationship between the columns in the old and new alignment.

    -selectcols { n,l,m-k }  Selection of columns to be removed from the alignment. Range: [0 - (Number of Columns - 1)]. (see User Guide).
    -selectseqs { n,l,m-k }  Selection of sequences to be removed from the alignment. Range: [0 - (Number of Sequences - 1)]. (see User Guide).

    -gt -gapthreshold &lt;n&gt;    1 - (fraction of sequences with a gap allowed). Range: [0 - 1]
    -st -simthreshold &lt;n&gt;    Minimum average similarity allowed. Range: [0 - 1]
    -ct -conthreshold &lt;n&gt;    Minimum consistency value allowed.Range: [0 - 1]
    -cons &lt;n&gt;                Minimum percentage of the positions in the original Alignment  to conserve. Range: [0 - 100]

    -nogaps                  Remove all positions with gaps in the alignment.
    -noallgaps               Remove columns composed only by gaps.
    -keepseqs                Keep sequences even if they are composed only by gaps.

    -gappyout                Use automated selection on "gappyout" mode. This method only uses information based on gaps' distribution. (see User Guide).
    -strict                  Use automated selection on "strict" mode. (see User Guide).
    -strictplus              Use automated selection on "strictplus" mode. (see User Guide).
                                 (Optimized for Neighbour Joining phylogenetic tree reconstruction).

    -automated1              Use a heuristic selection of the automatic method based on similarity statistics. (see User Guide).
                                (Optimized for Maximum Likelihood phylogenetic tree reconstruction).

    -terminalonly            Only columns out of internal boundaries (first and last column without gaps) are
                             candidates to be trimmed depending on the applied method
    -block &lt;n&gt;               Minimum column block size to be kept in the trimmed alignment. Available with manual and automatic (gappyout) methods

    -resoverlap &lt;n&gt;          Minimum overlap of a positions with other positions in the column to be considered a "good position". Range: [0 - 1]. (see User Guide).
    -seqoverlap &lt;n&gt;          Minimum percentage of "good positions" that a sequence must have in order to be conserved. Range: [0 - 100](see User Guide).

    -clusters &lt;n&gt;            Get the most Nth representatives sequences from a given alignment. Range: [1 - (Number of sequences)]
    -maxidentity &lt;n&gt;         Get the representatives sequences for a given identity threshold. Range: [0 - 1].

    -w &lt;n&gt;                   (half) Window size, score of position i is the average of the window (i - n) to (i + n).
    -gw &lt;n&gt;                  (half) Window size only applies to statistics/methods based on Gaps.
    -sw &lt;n&gt;                  (half) Window size only applies to statistics/methods based on Similarity.
    -cw &lt;n&gt;                  (half) Window size only applies to statistics/methods based on Consistency.

    -sgc                     Print gap scores for each column in the input alignment.
    -sgt                     Print accumulated gap scores for the input alignment.
    -ssc                     Print similarity scores for each column in the input alignment.
    -sst                     Print accumulated similarity scores for the input alignment.
    -sfc                     Print sum-of-pairs scores for each column from the selected alignment
    -sft                     Print accumulated sum-of-pairs scores for the selected alignment
    -sident                  Print identity scores for all sequences in the input alignment. (see User Guide).
    -soverlap                Print overlap scores matrix for all sequences in the input alignment. (see User Guide).


    -lf / --listformats     List available formats to load from and save to.

    -v &lt;n&gt; / --verbose &lt;n&gt;  Specify the verbose level of the program.
                                Available options: error / 3, warning / 2, info / 1, none / 0
                                Default value: error / 3

    -vcf &lt;inputfile, ...&gt;   Specify one or more VCF files to produce MSAs using the input file (-in &lt;n&gt;) as reference genome

    -formats &lt;format1, ...&gt; Specify one or more formats to save the resulting MSA.
                                Combinations with specific format arguments (-fasta, -nexus, etc) is allowed.
                                Available output formats: [oformat]

    --alternative_matrix degenerated_nt_identity
                            Specify the degenerated nt identity matrix as the similarity matrix to use.
                                If a matrix is not specified, the best suited among a set will be selected.
</pre> <h1><a class="anchor" id="citation_sec"></a>
Citing this tool</h1>
<pre class="fragment">    trimAl: a tool for automated alignment trimming in large-scale phylogenetic analyses.
    Salvador Capella-Gutierrez; Jose M. Silla-Martinez; Toni Gabaldon.
    Bioinformatics 2009, 25:1972-1973.
</pre><h1><a class="anchor" id="trimAl_code_sec"></a>
Understanding the code</h1>
<p>You can start by taking a look at: <br />
<a class="el" href="classtrim_al_manager.html">- trimAl Manager </a> <br />
<a class="el" href="class_alignment.html">- Alignment </a> <br />
</p>
<h1><a class="anchor" id="Acknowledgments"></a>
Acknowledgments</h1>
<p><a href="http://www.alanzucconi.com/2015/07/26/enum-flags-and-bitwise-operators/">Alan Zucconi </a>: As his post explaining bitwise operators is used as reference explaining <a class="el" href="defines_8h.html#a69d5409d6e7370d6bd3b4f931702e936">SequenceTypes</a><br />
 </p>
</div></div><!-- PageDoc -->
</div><!-- contents -->
</div><!-- doc-content -->
<!-- start footer part -->
<div id="nav-path" class="navpath"><!-- id is needed for treeview function! -->
  <ul>
    <li class="footer">Generated by
    <a href="http://www.doxygen.org/index.html">
    <img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.8.15 </li>
  </ul>
</div>
</body>
</html>
