<!-- HTML header for doxygen 1.8.8-->
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
    <head>
        <meta http-equiv="X-UA-Compatible" content="IE=edge">
        <!-- For Mobile Devices -->
        <meta name="viewport" content="width=device-width, initial-scale=1">

        <meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
        <meta name="generator" content="Doxygen 1.7.1"/>
        
        <script type="text/javascript" src="https://code.jquery.com/jquery-2.1.1.min.js"></script>

        <!--BEGIN PROJECT_NAME--><title>trimAl: trimAl: trimAl Main Page</title><!--END PROJECT_NAME-->
        <!--BEGIN !PROJECT_NAME--><title>trimAl: trimAl Main Page</title><!--END !PROJECT_NAME-->
        <!--<link href="$relpath^tabs.css" rel="stylesheet" type="text/css"/>-->
        <script type="text/javascript" src="$relpath^dynsections.js"></script>
        $treeview
        $search
        $mathjax
        <link href="$relpath^$stylesheet" rel="stylesheet" type="text/css" />
        $extrastylesheet
        <link href='https://fonts.googleapis.com/css?family=Roboto+Slab' rel='stylesheet' type='text/css'>
        <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.1/css/bootstrap.min.css">
        <script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.1/js/bootstrap.min.js"></script>
        <script type="text/javascript" src="$relpath^doxy-boot.js"></script>
    </head>
    <body>
        <nav class="navbar navbar-default" role="navigation">
            <div class="container">
                <div class="navbar-header">
                    <a class="navbar-brand">trimAl </a>
                </div>
            </div>
        </nav>
        <div id="top"><!-- do not remove this div, it is closed by doxygen! -->
            <div class="content" id="content">
                <div class="container">
                    <div class="row">
                        <div class="col-sm-12 panel " style="padding-bottom: 15px;">
                            <div style="margin-bottom: 15px;">
<!-- end header part -->
<!-- Generated by Doxygen 1.7.1 -->
<script type="text/javascript">
function hasClass(ele,cls) {
  return ele.className.match(new RegExp('(\\s|^)'+cls+'(\\s|$)'));
}

function addClass(ele,cls) {
  if (!this.hasClass(ele,cls)) ele.className += " "+cls;
}

function removeClass(ele,cls) {
  if (hasClass(ele,cls)) {
    var reg = new RegExp('(\\s|^)'+cls+'(\\s|$)');
    ele.className=ele.className.replace(reg,' ');
  }
}

function toggleVisibility(linkObj) {
 var base = linkObj.getAttribute('id');
 var summary = document.getElementById(base + '-summary');
 var content = document.getElementById(base + '-content');
 var trigger = document.getElementById(base + '-trigger');
 if ( hasClass(linkObj,'closed') ) {
   summary.style.display = 'none';
   content.style.display = 'block';
   trigger.src = 'open.png';
   removeClass(linkObj,'closed');
   addClass(linkObj,'opened');
 } else if ( hasClass(linkObj,'opened') ) {
   summary.style.display = 'block';
   content.style.display = 'none';
   trigger.src = 'closed.png';
   removeClass(linkObj,'opened');
   addClass(linkObj,'closed');
 }
 return false;
}
</script>
<div class="navigation" id="top">
  <div class="tabs">
    <ul class="tablist">
      <li class="current"><a href="main.html"><span>Main&nbsp;Page</span></a></li>
      <li><a href="pages.html"><span>Related&nbsp;Pages</span></a></li>
      <li><a href="modules.html"><span>Modules</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li><a href="files.html"><span>Files</span></a></li>
    </ul>
  </div>
</div>
<div class="header">
  <div class="headertitle">
<h1>trimAl Main Page </h1>  </div>
</div>
<div class="contents">
<h2><a class="anchor" id="introduction_sec"></a>
Introduction</h2>
<p>This is <b>trimAl's</b> information page. Also, you can find information related to the <a class="el" href="readal_pag.html">readAl program.</a> <br/>
 <b>trimAl</b> is a tool for the automated removal of spurious sequences or poorly aligned regions from a multiple sequence alignment.<br/>
<br/>
</p>
<p>trimAl can consider several parameters, alone or in multiple combinations, in order to select the most-reliable positions in the alignment. <br/>
 These include the <em> proportion of sequences with a gap</em>, the level of <em> residue similarity</em> and, if several alignments for the same set of sequences are provided, the <em> consistency level of columns among alignments</em>. <br/>
 Moreover, <b>trimAl</b> is able to <em> manually select a set of columns and sequences </em> to be removed from the alignment.<br/>
<br/>
</p>
<p><b>trimAl</b> implements a series of automated algorithms that apply different thresholds, based on the characteristics of each alignment, to be used so that the signal-to-noise ratio after alignment trimming phase is optimized.<br/>
</p>
<p>The user can remove <em> spurious sequences </em> from the alignment before applying any method to improve the alignment's quality.<br/>
</p>
<p>Among <b>trimAl's</b> additional features, <b>trimAl</b> allows getting the <em>complementary alignment</em> (columns that were trimmed), to <em> compute statistics </em> from the alignment, to <em> select the output file format </em>, to <em> get a summary </em> of <b>trimAl's</b> trimming in HTML format, and many other options. <br/>
</p>
<h2><a class="anchor" id="installation_sec"></a>
Installation</h2>
<p>The simplest way to compile this package is:</p>
<p>1.- Move to the project folder</p>
<p>2.- Configure the project: &gt; cmake .</p>
<p>3.- Compile the project: &gt; make</p>
<p>3b.- [Optional] Run tests: &gt; ctest</p>
<p>4.- Move or copy binaries folder './bin/' to 'usr/local/bin' or 'usr/bin'</p>
<h2><a class="anchor" id="publication_sec"></a>
Publications</h2>
<p><a href="http://trimal.cgenomics.org/_media/supplmaterial.b.pdf">- trimAl: a tool for automated alignment trimming in large-scale phylogenetic analyses. </a> <br/>
 <a href="http://trimal.cgenomics.org/_media/supplmaterial.b.pdf">- Supplementary material </a> <br/>
 <a href="http://trimal.cgenomics.org/_media/otherbenchs.b.pdf">- Other benchmarks </a> <br/>
</p>
<h2><a class="anchor" id="usage_sec"></a>
Usage</h2>
<div class="fragment"><pre class="fragment">
Basic usage
	new_trimal -in &lt;inputfile&gt; -out &lt;outputfile&gt; -(other options).

Common options (for a complete list please see the User Guide or visit http://trimal.cgenomics.org):

    -h                       Print this information and show some examples.
    --version                Print the trimAl version.

    -in &lt;inputfile&gt;          Input file in several formats (clustal, fasta, NBRF/PIR, nexus, phylip3.2, phylip).

    -compareset &lt;inputfile&gt;  Input list of paths for the files containing the alignments to compare.
    -forceselect &lt;inputfile&gt; Force selection of the given input file in the files comparison method.

    -backtrans &lt;inputfile&gt;   Use a Coding Sequences file to get a backtranslation for a given AA alignment
    -ignorestopcodon         Ignore stop codons in the input coding sequences
    -splitbystopcodon        Split input coding sequences up to first stop codon appearance

    -matrix  &lt;inpufile&gt;       Input file for user-defined similarity matrix (default is Blosum62).

    -out  &lt;outputfile&gt;        Output alignment in the same input format (default stdout). (default input format)
    -htmlout &lt;outputfile&gt;    Get a summary of trimal's work in an HTML file.

    -keepheader              Keep original sequence header including non-alphanumeric characters.
                             Only available for input FASTA format files. (future versions will extend this feature)

    -nbrf                    Output file in NBRF/PIR format
    -mega                    Output file in MEGA format
    -nexus                   Output file in NEXUS format
    -clustal                 Output file in CLUSTAL format

    -fasta                   Output file in FASTA format
    -fasta_m10               Output file in FASTA format. Sequences name length up to 10 characters.

    -phylip                  Output file in PHYLIP/PHYLIP4 format
    -phylip_m10              Output file in PHYLIP/PHYLIP4 format. Sequences name length up to 10 characters.
    -phylip_paml             Output file in PHYLIP format compatible with PAML
    -phylip_paml_m10         Output file in PHYLIP format compatible with PAML. Sequences name length up to 10 characters.
    -phylip3.2               Output file in PHYLIP3.2 format
    -phylip3.2_m10           Output file in PHYLIP3.2 format. Sequences name length up to 10 characters.

    -complementary           Get the complementary alignment.
    -colnumbering            Get the relationship between the columns in the old and new alignment.

    -selectcols { n,l,m-k }  Selection of columns to be removed from the alignment. Range: [0 - (Number of Columns - 1)]. (see User Guide).
    -selectseqs { n,l,m-k }  Selection of sequences to be removed from the alignment. Range: [0 - (Number of Sequences - 1)]. (see User Guide).

    -gt -gapthreshold \c&lt;n&gt;    1 - (fraction of sequences with a gap allowed). Range: [0 - 1]
    -st -simthreshold &lt;n&gt;    Minimum average similarity allowed. Range: [0 - 1]
    -ct -conthreshold &lt;n&gt;    Minimum consistency value allowed.Range: [0 - 1]
    -cons &lt;n&gt;                Minimum percentage of the positions in the original alignment to conserve. Range: [0 - 100]

    -nogaps                  Remove all positions with gaps in the alignment.
    -noallgaps               Remove columns composed only by gaps.
    -keepseqs                Keep sequences even if they are composed only by gaps.

    -gappyout                Use automated selection on "gappyout" mode. This method only uses information based on gaps' distribution. (see User Guide).
    -strict                  Use automated selection on "strict" mode. (see User Guide).
    -strictplus              Use automated selection on "strictplus" mode. (see User Guide).
                             (Optimized for Neighbour Joining phylogenetic tree reconstruction).

    -automated1              Use a heuristic selection of the automatic method based on similarity statistics. (see User Guide). 
                             Optimized for Maximum Likelihood phylogenetic tree reconstruction.

    -terminalonly            Only columns out of internal boundaries (first and last column without gaps) are 
                             candidated to be trimmed depending on the applied method
    -block &lt;n&gt;               Minimum column block size to be kept in the trimmed alignment. Available with manual and automatic (gappyout) methods

    -resoverlap              Minimum overlap of a positions with other positions in the column to be considered a "good position". Range: [0 - 1]. (see User Guide).
    -seqoverlap              Minimum percentage of "good positions" that a sequence must have in order to be conserved. Range: [0 - 100](see User Guide).

    -clusters &lt;n&gt;            Get the most Nth representatives sequences from a given alignment. Range: [1 - (Number of sequences)]
    -maxidentity &lt;n&gt;         Get the representatives sequences for a given identity threshold. Range: [0 - 1].

    -w &lt;n&gt;                   (half) Window size, score of position i is the average of the window (i - n) to (i + n).
    -gw &lt;n&gt;                  (half) Window size only applies to statistics/methods based on Gaps.
    -sw &lt;n&gt;                  (half) Window size only applies to statistics/methods based on Similarity.
    -cw &lt;n&gt;                  (half) Window size only applies to statistics/methods based on Consistency.

    -sgc                     Print gap scores for each column in the input alignment.
    -sgt                     Print accumulated gap scores for the input alignment.
    -ssc                     Print similarity scores for each column in the input alignment.
    -sst                     Print accumulated similarity scores for the input alignment.
    -sfc                     Print sum-of-pairs scores for each column from the selected alignment
    -sft                     Print accumulated sum-of-pairs scores for the selected alignment
    -sident                  Print identity scores for all sequences in the input alignment. (see User Guide).
</pre></div> <h2><a class="anchor" id="citation_sec"></a>
Citing this tool</h2>
<p>trimAl: a tool for automated alignment trimming in large-scale phylogenetic analyses. Salvador Capella-Gutierrez; Jose M. Silla-Martinez; Toni Gabaldon. Bioinformatics 2009, 25:1972-1973.</p>
<h2><a class="anchor" id="trimAl_code_sec"></a>
Understanding the code</h2>
<p>You can start by taking a look at: <br/>
 <a class="el" href="classtrim_al_manager.html">- trimAl Manager </a> <br/>
 <a class="el" href="classnew_alignment.html">- Alignment </a> <br/>
 </p>
</div>
<!-- HTML footer for doxygen 1.8.8-->
<!-- start footer part -->
<!--BEGIN GENERATE_TREEVIEW-->
<div id="nav-path" class="navpath"><!-- id is needed for treeview function! -->
  <ul>
    $navpath
    <li class="footer">$generatedby
    <a href="http://www.doxygen.org/index.html">
    <img class="footer" src="$relpath^doxygen.png" alt="doxygen"/></a> 1.7.1 </li>
  </ul>
</div>
<!--END GENERATE_TREEVIEW-->
</div>
</div>
</div>
</div>
</div>
<!--BEGIN !GENERATE_TREEVIEW-->
<hr class="footer"/><address class="footer"><small>
$generatedby &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="$relpath^doxygen.png" alt="doxygen"/>
</a> 1.7.1
</small></address>
<!--END !GENERATE_TREEVIEW-->
</body>
</html>
