<!-- HTML header for doxygen 1.8.8-->
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.13"/>
<title>Trimal: readAl Main Page</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtreedata.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<script type="text/javascript">
  $(document).ready(initResizable);
</script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0" width=100%>
 <tbody>
 <tr style="height: 12px;">
  <td style="padding-left: 0.5em;">
   <div id="projectname">Trimal
   </div>
   <div id="projectbrief">A Multiple Alignment Sequence framework for Cleaning and Reformating.</div>
  </td>
<!--  <td style="float:right;margin-top:18px;vertical-align:bottom">
     <a href="http://assimilationsystems.com/download/">
       <img src="http://assimilationsystems.com/wp-content/uploads/2015/03/Download-Doxygen.png" style="float:right;" height="20" width="73">
     </a>
  </td>-->
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.13 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
</script>
<div id="main-nav"></div>
</div><!-- top -->
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
      <div id="nav-sync" class="sync"></div>
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
$(document).ready(function(){initNavTree('readal_pag.html','');});
</script>
<div id="doc-content">
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div class="header">
  <div class="headertitle">
<div class="title">readAl Main Page </div>  </div>
</div><!--header-->
<div class="contents">
<div class="textblock"><h1><a class="anchor" id="introduction_sec"></a>
Introduction</h1>
<p>One of the main problems that the researchers have to face during their studies is the <em>conversion among the various formats</em> required by different programs. <br />
Nowadays there around <em>20 different Multiple Sequence Alignment (MSAs) formats</em>. <br />
<br />
 Here we present <b>readAl</b>, a tool for format alignment conversion among the most representative formats. <br />
<b>readAl</b> has been designed to be used in several situations, including, alignment trimming, phylogenetic tree reconstruction, profile reconstruction, and so on.<br />
<br />
 <b>readAl</b> has been implemented in <em>C++ programming language</em>. <br />
This program is part of the <b>trimAl</b> package and is used internally by <b>trimAl</b> to convert among different formats.<br />
 </p>
<h1><a class="anchor" id="installation_sec"></a>
Installation</h1>
<p>The simplest way to compile this program is: </p><pre class="fragment">1.- Move to the project folder

2.- Configure the project: 
    &gt; cmake .

3.- Compile the project: 
    &gt; make
    &gt; make readalMS (if you only want readalMS to be compiled)

4.- Move or copy binaries folder './bin/' to 'usr/local/bin' or 'usr/bin'
</pre><h1><a class="anchor" id="usage_sec"></a>
Usage</h1>
<p>This program is free software: you can redistribute it and/or modify it under the terms of the <b>GNU General Public License</b> as published by the Free Software Foundation, the last available version.</p>
<h2><a class="anchor" id="basic_usage"></a>
BASIC USAGE</h2>
<pre class="fragment">readalMS -in &lt;inputfiles&gt; -out &lt;pattern&gt; -format [formats] [options].

-h                    Show this information.
-in &lt;inputfiles&gt;      Input files in several formats. Separated by spaces.
-out &lt;pattern&gt;        Output file name (default STDOUT).
                      It will replace the tags [in]        -&gt; Original filename without extension.
                                               [format]    -&gt; Output's format name
                                               [extension] -&gt; Output's extension

-formats             Formats you want the output to be converted to.
                     Available formats are CLUSTAL, FASTA, PIR, PHYLIP32, PHYLIP40, PHYLIP_PAML, NEXUS, MEGAI, MEGAS, HTML. 
                     Being the HTML format not a format itself, but a colored report of the alignment files.

-format              Print information about input file format and if sequences are aligned or not.
-type                Print information about biological sequences datatype (e.g. nucleotides:dna, nucleotides:rna, aminoacids, etc)
-info                Print information about sequences number, average sequence length, max &amp; min sequence length
-reverse             Output the reverse of sequences in input file.

-shortNames          Shortens the names so they fit on certain formats
-keepHeaders         Keeps the headers of the original format if it had any
</pre><h2><a class="anchor" id="legacy_opt"></a>
LEGACY OPTIONS</h2>
<p>Take in mind that this arguments may be discontinued any time. </p><pre class="fragment">-onlyseqs            Generate output with only residues from input file

-html                Output residues colored according their physicochemical properties. HTML file.

-nbrf                Output file in NBRF/PIR format
-mega                Output file in MEGA format
-nexus               Output file in NEXUS format
-clustal             Output file in CLUSTAL format

-fasta               Output file in FASTA format
-fasta_m10           Output file in FASTA format. Sequences name up to 10 characters.

-phylip              Output file in PHYLIP/PHYLIP4 format
-phylip_m10          Output file in PHYLIP/PHYLIP4 format. Sequences name up to 10 characters.
-phylip_paml         Output file in PHYLIP format compatible with PAML
-phylip_paml_m10     Output file in PHYLIP format compatible with PAML. Sequences name up to 10 characters.
-phylip3.2           Output file in PHYLIP3.2 format
-phylip3.2_m10       Output file in PHYLIP3.2 format. Sequences name up to 10 characters.
</pre><h2><a class="anchor" id="examples"></a>
EXAMPLES OF USE</h2>
<pre class="fragment">readalMS -in ./dataset/AA1.fas -out ./dataset/[in].output.[extension] -formats clustal
  -&gt; Will produce ./dataset/AA1.output.clw

readalMS -in ./dataset/example1.clw -out ./dataset/[in].[format].[extension] -formats fasta phylip32 phylip40
  -&gt; Will produce ./dataset/example1.FASTA.fasta ./dataset/example1.PHYLIP32.phy ./dataset/example1.PHYLIP40.phy

readalMS -in ./dataset/example1.clw -out ./dataset/[in]/[format].[extension] -formats fasta phylip32 phylip40
  -&gt; Will produce ./dataset/example1/FASTA.fasta ./dataset/example1/PHYLIP32.phy ./dataset/example1/PHYLIP40.phy
     ONLY if ./dataset/example1/ already exists.

readalMS -in ./dataset/AA1.fas ./dataset/AA2.fas -out ./dataset/[in].output.[extension] -formats clustal pir
  -&gt; Will produce ./dataset/AA1.output.clw ./dataset/AA2.output.clw ./dataset/AA1.output.pir ./dataset/AA2.output.pir

readalMS -in ./dataset/AA1.fas -format -type -info
  -&gt; Will produce terminal output giving information about AA1.fas alignment file

readalMS -in ./dataset/AA1.fas ./dataset/AA2.fas -out ./dataset/[in].output.[extension] -formats html
  -&gt; Will produce ./dataset/AA1.output.html ./dataset/AA2.output.html
     Those files are not indeed reformats of the original alignments, but an HTML colored report of the alignment file.</pre> </div></div><!-- contents -->
</div><!-- doc-content -->
<!-- start footer part -->
<div id="nav-path" class="navpath"><!-- id is needed for treeview function! -->
  <ul>
    <li class="footer">Generated by
    <a href="http://www.doxygen.org/index.html">
    <img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.8.13 </li>
  </ul>
</div>
</body>
</html>
