<!-- HTML header for doxygen 1.8.8-->
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
    <meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
    <meta http-equiv="X-UA-Compatible" content="IE=9"/>
    <meta name="generator" content="Doxygen 1.8.15"/>
    <title>trimAl: What can I do with trimAl?</title>
    <link href="tabs.css" rel="stylesheet" type="text/css"/>
    <script type="text/javascript" src="jquery.js"></script>
    <script type="text/javascript" src="dynsections.js"></script>
    <link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtreedata.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
  $(document).ready(initResizable);
/* @license-end */</script>
    <link href="doxygen.css" rel="stylesheet" type="text/css" />
    <link href="CustomExtra.css" rel="stylesheet" type="text/css"/>
    <script type="text/javascript">
        function myFunction(ID) {
            var divs = document.getElementsByName(ID);
            for (var x of divs) {
                if (x.style.display === "none") {
                    x.style.display = "block";
                } else {
                    x.style.display = "none";
                }
            }
        } 
    </script>
</head>
<body>
    <div id="top"><!-- do not remove this div, it is closed by doxygen! -->
        <div id="titlearea">
            <table cellspacing="0" cellpadding="0" width=100%>
                <tbody>
                    <tr style="height: 12px;">
                        <td id="projectlogo"><img alt="Logo" src="Logo.png"/></td>
                    </tr>
                </tbody>
            </table>
            <div class="header">
                <br>
            </div>
        </div>
        <!-- end header part -->
    <!-- Generated by Doxygen 1.8.15 -->
</div><!-- top -->
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
      <div id="nav-sync" class="sync"></div>
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(document).ready(function(){initNavTree('whatcanido.html','');});
/* @license-end */
</script>
<div id="doc-content">
<div class="PageDoc"><div class="header">
  <div class="headertitle">
<div class="title">What can I do with trimAl? </div>  </div>
</div><!--header-->
<div class="contents">
<div class="textblock"><h1><a class="anchor" id="TrimmingMethods"></a>
Trimming Methods</h1>
<h2><a class="anchor" id="MetaAutomated"></a>
Meta Automated Methods</h2>
<p>Meta Automated methods analyze your alignment to decide between automated statistical methods.<br />
 Currently, only Automated1 is available, and it will analyze your MSA and select between Strict and Gappyout<br />
 </p><table class="markdownTable">
<tr class="markdownTableHead">
<th class="markdownTableHeadNone">Method  </th><th class="markdownTableHeadNone">Description  </th><th class="markdownTableHeadNone">Example of use   </th></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyNone">Automated1  </td><td class="markdownTableBodyNone">Chooses between Strict and Gappyout <br />
 <em> Optimized for Maximum Likelihood phylogenetic tree reconstruction </em>  </td><td class="markdownTableBodyNone">bin/trimal -in dataset/example.007.AA.fasta -automated1   </td></tr>
</table>
<h2><a class="anchor" id="AutomatedS"></a>
Automated Statistical Methods</h2>
<p>Automated Statistical methods will search for optimum statistic thresholds to clean your alignment using statistics distribution.<br />
</p>
<table class="markdownTable">
<tr class="markdownTableHead">
<th class="markdownTableHeadNone">Method  </th><th class="markdownTableHeadNone">Description  </th><th class="markdownTableHeadNone">Example of use   </th></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyNone">Strict  </td><td class="markdownTableBodyNone">Uses Gaps and Similarity statistics to clean the alignment.  </td><td class="markdownTableBodyNone">bin/trimal -in dataset/example.007.AA.fasta -strict   </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyNone">Strictplus  </td><td class="markdownTableBodyNone">Uses Gaps and Similarity statistics to clean the alignment.<br />
 <em> Optimized for Neighbour Joining phylogenetic tree reconstruction </em>  </td><td class="markdownTableBodyNone">bin/trimal -in dataset/example.007.AA.fasta -strictplus   </td></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyNone">Gappyout  </td><td class="markdownTableBodyNone">Uses Gaps statistic to clean the alignment.  </td><td class="markdownTableBodyNone">bin/trimal -in dataset/example.007.AA.fasta -gappyout   </td></tr>
</table>
<h2><a class="anchor" id="ManualThresholds"></a>
Manual Threshold Methods</h2>
<p>Manual Threshold methods allow the user to provide thresholds and windows for each statistic available.<br />
It also allows the user to obtain the raw values for each statistic, per column and accumulative.<br />
 </p><table class="markdownTable">
<tr class="markdownTableHead">
<th class="markdownTableHeadNone">Statistic <br />
 Name  </th><th class="markdownTableHeadNone">Threshold <br />
 Argument  </th><th class="markdownTableHeadNone">Window <br />
 Argument  </th><th class="markdownTableHeadNone">Statistic <br />
 Per Colum <br />
 Argument  </th><th class="markdownTableHeadNone">Statistic <br />
 Accumulative <br />
 Argument  </th><th class="markdownTableHeadNone">Example <br />
 of use   </th></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyNone">Gaps  </td><td class="markdownTableBodyNone">-gt &lt;n&gt; [0 - 1]  </td><td class="markdownTableBodyNone">-gw &lt;n&gt; [0 - 1/4*N]  </td><td class="markdownTableBodyNone">-sgc  </td><td class="markdownTableBodyNone">-sgt  </td><td class="markdownTableBodyNone">bin/trimal -in dataset/example.007.AA.fasta -gt 0.5 -gw 2   </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyNone">Similarity  </td><td class="markdownTableBodyNone">-st &lt;n&gt; [0 - 1]  </td><td class="markdownTableBodyNone">-sw &lt;n&gt; [0 - 1/4*N]  </td><td class="markdownTableBodyNone">-ssc  </td><td class="markdownTableBodyNone">-sst  </td><td class="markdownTableBodyNone">bin/trimal -in dataset/example.007.AA.fasta -st 0.5 -sw 2   </td></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyNone">Consitency  </td><td class="markdownTableBodyNone">-ct &lt;n&gt; [0 - 1]  </td><td class="markdownTableBodyNone">-cw &lt;n&gt; [0 - 1/4*N]  </td><td class="markdownTableBodyNone">-sfc  </td><td class="markdownTableBodyNone">-sft  </td><td class="markdownTableBodyNone">bin/trimal -compareset data -ct 0.5 -cw 2   </td></tr>
</table>
<p>N = Number of residues in the input alignment</p>
<h2><a class="anchor" id="Overlap"></a>
Manual Overlap Threshold Methods</h2>
<p>Manual Overlap method allow the user to provide a minimum threshold for similarity to keep a sequence or residue/column. <br />
Both arguments, sequence and residue overlap, are mutually dependent, so, if any is set, the other one must be set, too.</p>
<table class="markdownTable">
<tr class="markdownTableHead">
<th class="markdownTableHeadNone">Overlap Level  </th><th class="markdownTableHeadNone">Command line argument   </th></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyNone">Sequence  </td><td class="markdownTableBodyNone">-seqoverlap &lt;n&gt; [0-100]   </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyNone">Residue  </td><td class="markdownTableBodyNone">-resoverlap &lt;n&gt; [0-1]   </td></tr>
</table>
<p><b>Example of use</b> <br />
As sequence and residue arguments are mutually dependent, the example is provided combined: </p><pre class="fragment"> 
 bin/trimal -in dataset/example.007.AA.fasta -resoverlap 10 -seqoverlap 0.5 
 
</pre><h2><a class="anchor" id="Manual"></a>
Manual Methods</h2>
<p>Manual methods allow the user to provide a pattern to remove residues or sequences.<br />
 </p><table class="markdownTable">
<tr class="markdownTableHead">
<th class="markdownTableHeadNone">Method  </th><th class="markdownTableHeadNone">Command line argument  </th><th class="markdownTableHeadNone">Example of use   </th></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyNone">Sequences  </td><td class="markdownTableBodyNone">-selectseqs { pattern }  </td><td class="markdownTableBodyNone">bin/trimal -in dataset/example.007.AA.fasta -selectseqs { 0,4 }   </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyNone">Residues  </td><td class="markdownTableBodyNone">-selectcols { pattern }  </td><td class="markdownTableBodyNone">bin/trimal -in dataset/example.007.AA.fasta -selectcols { 1-4 }   </td></tr>
</table>
<p> <button onclick="myFunction(' manual-t')" class="btn btn-danger"><b> Example of use</b></button> </p><div name=" manual-t " style="display:none; width: -moz-fit-content"></div><div name=" manual-t " style="display:none; width: -moz-fit-content">The user must provide a pattern after the argument.<br />
 This pattern is a set of comma separated values: <pre class="fragment"> 
 bin/trimal -in dataset/example.007.AA.fasta -selectseqs { 0,1,2,3,4 } 
 
</pre> <pre class="fragment">This would delete sequences 0, 1, 2, 3 and 4
</pre></div><div name=" manual-t " style="display:none; width: -moz-fit-content">Instead of providing raw position IDs, the user can also provide ranges, by separating start and end with a hyphen: <pre class="fragment"> 
 bin/trimal -in dataset/example.007.AA.fasta -selectseqs { 0-4 } 
 
</pre> <pre class="fragment">This would delete sequences 0, 1, 2, 3 and 4
</pre></div><div name=" manual-t " style="display:none; width: -moz-fit-content">Combination of both separators are allowed: <pre class="fragment"> 
 bin/trimal -in dataset/example.007.AA.fasta -selectseqs { 0,4-6,12,16-18 } 
 
</pre> <pre class="fragment">This would delete sequences 0, 4, 5, 6, 12, 16, 17 and 18
</pre></div><div name=" manual-t " style="display:none; width: -moz-fit-content">Meaning the call above is synonymous to the following: <pre class="fragment"> 
 bin/trimal -in dataset/example.007.AA.fasta -selectseqs { 0,4,5,6,12,16,17,18 } 
 
</pre> <pre class="fragment">This would delete sequences 0, 4, 5, 6, 12, 16, 17 and 18
</pre></div><div name=" manual-t " style="display:none; width: -moz-fit-content"></div><h2><a class="anchor" id="Automated-gap"></a>
Automated Methods</h2>
<p>Automated methods will remove columns based on raw gaps values.<br />
 </p><table class="markdownTable">
<tr class="markdownTableHead">
<th class="markdownTableHeadNone">Method  </th><th class="markdownTableHeadNone">Description  </th><th class="markdownTableHeadNone">Example of use   </th></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyNone">NoGaps  </td><td class="markdownTableBodyNone">Removes any column containing at least, one gap  </td><td class="markdownTableBodyNone">bin/trimal -in dataset/example.007.AA.fasta -nogaps   </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyNone">NoAllGaps  </td><td class="markdownTableBodyNone">Removes columns containing only gaps  </td><td class="markdownTableBodyNone">bin/trimal -in dataset/example.007.AA.fasta -noallgaps   </td></tr>
</table>
<p> <button onclick="myFunction(' automated-gap')" class="btn btn-danger"><b>  Example of use</b></button> </p><div name=" automated-gap" style="display:none; width: -moz-fit-content"></div><div name=" automated-gap" style="display:none; width: -moz-fit-content"><b> Alignment </b> <pre class="fragment"> 6 43
Sp6          ---ASGAILT-L-GIYLFTLCAVISVSWYLAWLGLEINMMAII
Sp17         FAYTAPDLL-LI-GFLLKTVA-TFGDTWFQLWQGLDLNKMPVF
Sp10         -----DPAVL-F-VIMLGTIT-KFSSEWFFAWLGLEINMMVII
Sp8          ---GLGKVIV-Y-GIVLGTKSDQFSVVWLFPWNGLQIHMMGII
Sp26         AAAAAAALLTYL-GLFLGTDYENFAAAAANAWLGLEINMMAQI
Sp33         -----PTILNIA-GLHMETDI-NFSLAWFQAWGGLEINKQAIL
</pre></div><div name=" automated-gap" style="display:none; width: -moz-fit-content"><b> Alignment </b> + "-noallgaps" <pre class="fragment"> 6 42
Sp6          ---ASGAILT-LGIYLFTLCAVISVSWYLAWLGLEINMMAII
Sp17         FAYTAPDLL-LIGFLLKTVA-TFGDTWFQLWQGLDLNKMPVF
Sp10         -----DPAVL-FVIMLGTIT-KFSSEWFFAWLGLEINMMVII
Sp8          ---GLGKVIV-YGIVLGTKSDQFSVVWLFPWNGLQIHMMGII
Sp26         AAAAAAALLTYLGLFLGTDYENFAAAAANAWLGLEINMMAQI
Sp33         -----PTILNIAGLHMETDI-NFSLAWFQAWGGLEINKQAIL
</pre></div><div name=" automated-gap" style="display:none; width: -moz-fit-content"><b> Alignment </b> + "-nogaps" <pre class="fragment"> 6 34
Sp6          GAILLGIYLFTLCVISVSWYLAWLGLEINMMAII
Sp17         PDLLIGFLLKTVATFGDTWFQLWQGLDLNKMPVF
Sp10         DPAVFVIMLGTITKFSSEWFFAWLGLEINMMVII
Sp8          GKVIYGIVLGTKSQFSVVWLFPWNGLQIHMMGII
Sp26         AALLLGLFLGTDYNFAAAAANAWLGLEINMMAQI
Sp33         PTILAGLHMETDINFSLAWFQAWGGLEINKQAIL
</pre> </div><h1><a class="anchor" id="TrimmingTweaks"></a>
Trimming Tweaks</h1>
<h2><a class="anchor" id="CustomMatrix"></a>
Custom Matrix</h2>
<p>Similarity values are obtained using a reference matrix<br />
While trimAl contains matrices for each sequence type, the user can provide their own matrices.<br />
<br />
 Examples of these matrices can be found on the dataset folder:<br />
</p>
<p> <button onclick="myFunction(' blossum62')" class="btn btn-danger"><b>  Blossum62</b></button> </p><div name=" blossum62" style="display:none; width: -moz-fit-content"> <pre class="fragment">a R N D C Q E G H I L k M F P S T W Y V
A +4 -1 -2 -2 +0 -1 -1 +0 -2 -1 -1 -1 -1 -2 -1 +1 +0 -3 -2 +0
R -1 +5 +0 -2 -3 +1 +0 -2 +0 -3 -2 +2 -1 -3 -2 -1 -1 -3 -2 -3
N -2 +0 +6 +1 -3 +0 +0 +0 +1 -3 -3 +0 -2 -3 -2 +1 +0 -4 -2 -3
D -2 -2 +1 +6 -3 +0 +2 -1 -1 -3 -4 -1 -3 -3 -1 +0 -1 -4 -3 -3
C +0 -3 -3 -3 +9 -3 -4 -3 -3 -1 -1 -3 -1 -2 -3 -1 -1 -2 -2 -1
Q -1 +1 +0 +0 -3 +5 +2 -2 +0 -3 -2 +1 +0 -3 -1 +0 -1 -2 -1 -2
E -1 +0 +0 +2 -4 +2 +5 -2 +0 -3 -3 +1 -2 -3 -1 +0 -1 -3 -2 -2
G +0 -2 +0 -1 -3 -2 -2 +6 -2 -4 -4 -2 -3 -3 -2 +0 -2 -2 -3 -3
H -2 +0 +1 -1 -3 +0 +0 -2 +8 -3 -3 -1 -2 -1 -2 -1 -2 -2 +2 -3
I -1 -3 -3 -3 -1 -3 -3 -4 -3 +4 +2 -3 +1 +0 -3 -2 -1 -3 -1 +3
L -1 -2 -3 -4 -1 -2 -3 -4 -3 +2 +4 -2 +2 +0 -3 -2 -1 -2 -1 +1
K -1 +2 +0 -1 -3 +1 +1 -2 -1 -3 -2 +5 -1 -3 -1 +0 -1 -3 -2 -2
M -1 -1 -2 -3 -1 +0 -2 -3 -2 +1 +2 -1 +5 +0 -2 -1 -1 -1 -1 +1
F -2 -3 -3 -3 -2 -3 -3 -3 -1 +0 +0 -3 +0 +6 -4 -2 -2 +1 +3 -1
P -1 -2 -2 -1 -3 -1 -1 -2 -2 -3 -3 -1 -2 -4 +7 -1 -1 -4 -3 -2
S +1 -1 +1 +0 -1 +0 +0 +0 -1 -2 -2 +0 -1 -2 -1 +4 +1 -3 -2 -2
T +0 -1 +0 -1 -1 -1 -1 -2 -2 -1 -1 -1 -1 -2 -1 +1 +5 -2 -2 +0
W -3 -3 -4 -4 -2 -2 -3 -2 -2 -3 -2 -3 -1 +1 -4 -3 -2 +1 +2 -3
Y -2 -2 -2 -3 -2 -1 -2 -3 +2 -1 -1 -2 -1 +3 -3 -2 -2 +2 +7 -1
V +0 -3 -3 -3 -1 -2 -2 -3 -3 +3 +1 -2 +1 -1 -2 -2 +0 -3 -1 +4
</pre> </div><p> <button onclick="myFunction(' deg-dna')" class="btn btn-danger"><b>  Degenerated_DNA</b></button> </p><div name=" deg-dna" style="display:none; width: -moz-fit-content"> <pre class="fragment">A C D G K M S R T W Y
A 1 0 0 0 0 0 0 0 0 0 0
C 0 1 0 0 0 0 0 0 0 0 0
D 0 0 1 0 0 0 0 0 0 0 0
G 0 0 0 1 0 0 0 0 0 0 0
K 0 0 0 0 1 0 0 0 0 0 0
M 0 0 0 0 0 1 0 0 0 0 0
S 0 0 0 0 0 0 1 0 0 0 0
R 0 0 0 0 0 0 0 1 0 0 0
T 0 0 0 0 0 0 0 0 1 0 0
W 0 0 0 0 0 0 0 0 0 1 0
Y 0 0 0 0 0 0 0 0 0 0 1
</pre> </div><h2><a class="anchor" id="Complementary"></a>
Complementary</h2>
<p>To extend the capability of trimAl, it is possible to obtain the complementary alignment.<br />
This means, flipping the values of residues to keep/reject. <br />
<br />
  <button onclick="myFunction(' complementary')" class="btn btn-danger"><b>  Example of use</b></button> </p><div name=" complementary" style="display:none; width: -moz-fit-content"></div><div name=" complementary" style="display:none; width: -moz-fit-content"><b> Alignment </b> <pre class="fragment">&gt;Sp8
FPWNGLQIHMMGIII---------------------------------------------------------------------------
&gt;Sp17
FPWNGLQIHMMGIIIIIIIIIIIIIIIIII------------------------------------------------------------
&gt;Sp10
FPWNGLQIHMMGIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII---------------------------------------------
&gt;Sp26
FPWNGLQIHMMGIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII------------------------------
&gt;Sp33
FPWNGLQIHMMGIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII---------------
&gt;Sp6
FPWNGLQIHMMGIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII
</pre></div><div name=" complementary" style="display:none; width: -moz-fit-content"><b> Alignment </b> + "-gt 0.8" <pre class="fragment">&gt;Sp8
FPWNGLQIHMMGIII---------------
&gt;Sp17
FPWNGLQIHMMGIIIIIIIIIIIIIIIIII
&gt;Sp10
FPWNGLQIHMMGIIIIIIIIIIIIIIIIII
&gt;Sp26
FPWNGLQIHMMGIIIIIIIIIIIIIIIIII
&gt;Sp33
FPWNGLQIHMMGIIIIIIIIIIIIIIIIII
&gt;Sp6
FPWNGLQIHMMGIIIIIIIIIIIIIIIIII
</pre></div><div name=" complementary" style="display:none; width: -moz-fit-content"><b> Alignment </b> + "-gt 0.8 -complementary" <pre class="fragment">&gt;Sp8
------------------------------------------------------------
&gt;Sp17
------------------------------------------------------------
&gt;Sp10
IIIIIIIIIIIIIII---------------------------------------------
&gt;Sp26
IIIIIIIIIIIIIIIIIIIIIIIIIIIIII------------------------------
&gt;Sp33
IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII---------------
&gt;Sp6
IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII
</pre> </div><h2><a class="anchor" id="TerminalOnly"></a>
Terminal Only</h2>
<p>Terminal Only option allows you to only trim the alignment outside the central block.<br />
 Central Block is defined as the block which contains <br />
 the center of the alignments and doesn't contain any gap inside.<br />
 Considering the Central Block as the best alignment section of an alignment, <br />
 due to be the one being built with no gaps, <br />
 we can trim the alignment keeping this Central Block untouched.<br />
<br />
  <button onclick="myFunction(' terminalonly')" class="btn btn-danger"><b>  Example of use</b></button> </p><div name=" terminalonly" style="display:none; width: -moz-fit-content"> <pre class="fragment">&gt;Csa004271
---------------------------------MYMAMGHFFDRDDVALKNISEYFKECS
EEEREHANKMIEFHNKRGGTTTYFPIKAPGSFDPANFNTIKAMNCALALEVNVNKSLLAL
HE--TANGDPEFQDFIEANFLHEQVDAIKKLKDYITNLKLVG---TGLGEFLFDKHFKSS
-----
&gt;Xtr21234
----MISQVRQNYSHDCEAAVNRMVNLEMYASYTYLSMSHYFDRDDVALHHVAEFFKEQS
KEERECAEKLMKCQNKRGGRIVLQDIKKPERDEWG--STLDAMQTALDLEKHVNQALLDL
HNLATERKDPHICDFLESEHLDEQVKHMKKFGDHITNLKRLGVPQNGMGEYLFDKHSLS-
-----
&gt;LcaH
----MSSQVRQNFHQDCEAAINRQINLELYASYVYLSMAYYFDRDDQALHNFAKFFRHQS
HEEREHAEKLMKLQNQRGGRIFLQDVRKPDRDEWG--SGVEALECALQLEKSVNQSLLDL
HKLCSDHNDPHLCDFIETHYLDEQVKSIKELADWVTNLRRMGAPQNGMAEYLFDKHTLGK
ES--S
&gt;Hsa167996
MTTASTSQVRQNYHQDSEAAINRQINLELYASYVYLSMSYYFDRDDVALKNFAKYFLHQS
HEEREHAEKLMKLQNQRGGRIFLQDIKKPDCDDWE--SGLNAMECALHLEKNVNQSLLEL
HKLATDKNDPHLCDFIETHYLNEQVKAIKELGDHVTNLRKMGAPESGLAEYLFDKHTLGD
SDNES
&gt;Mmu024661
MTTASPSQVRQNYHQDAEAAINRQINLELYASYVYLSMSCYFDRDDVALKNFAKYFLHQS
HEEREHAEKLMKLQNQRGGRIFLQDIKKPDRDDWE--SGLNAMECALHLEKSVNQSLLEL
HKLATDKNDPHLCDFIETYYLSEQVKSIKELGDHVTNLRKMGAPEAGMAEYLFDKHTLGH
GD-ES
&gt;Dre37936
---METSQIRQNYVRDCEAAINKMINLELYAGYTYTSMAHYFKRDDVALPGFAKFFKKNS
EEEREHAEKFMEFQNKRGGRIVLQDIKKPDRDVWG--NGLIAMQCALQLEKNVNQALLDL
HKLATEMGDPHLCDFLETHYLNEQVEAIKKLGDHITNLSKMDAGNNRMAEYLFDKHTLDS
-----
&gt;LcaM
----MESQVRQNYHRDCEAAVNRMVNMEMFASYTYTSMAFYFSRDDVALPGFSHFFKENS
DEEREHAEKLLSFQNKRGGHIFLQDIKKPERDEWG--SGLEAMQCALQLKKNVNQALLDL
HKLASDHGDPHLCDFLETHYLNEQVEAIKKLGDYISNLSRMDAQKNKMAEYLFDKHSLGG
KS---
&gt;Tru14292
----MESQVRQNYHRDCEAAINKMINMELYASYTYTSMAFFFSRDDVALPGFAHFFKENS
DEEREHAEKLLSFQNKRGGRIFLQDIKKPERDEWG--SGLEAMQCALQLEKKVNQALLDL
HKLASDHVDPHLCDFLESHYLNEQVEAIKKLGDYITNLSRMDAQNNKMAEYLFDKHTLGS
KS---
&gt;Ola20972
----MESQVRQNYHRDCEAAINRMVNMELFASYTYTSMAFYFDRDDVALPGFSHFFKENS
HEEKEHADKLLSFQNKRGGRIFLQDVKKPERDEWG--SGLEAMQCALQLEKNVNQALLDL
HKVASDHKDPHMCDFLETHYLNEQVESIKKIGDHITNLTRMDAHTNKMAEYLFDKHTLGS
KS---
</pre></div><div name=" terminalonly" style="display:none; width: -moz-fit-content"><b> Alignment </b> + "-gt 0.9" <pre class="fragment">&gt;Csa004271
-----------------YMFRDDALFSEEEAKNRGGPNFAALLVNLLLHDPDFELEQVKD
NLELFDKH---
&gt;Xtr21234
---SQRQNDEAANNEAYYMFRDDALFSEEEAKNRGGP--AALLVNLLLHDPDFELEQVKD
NLELFDKH---
&gt;LcaH
---SQRQNDEAANNEAYYMFRDDALFSEEEAKNRGGP--AALLVNLLLHDPDFELEQVKD
NLELFDKH--S
&gt;Hsa167996
MTTSQRQNDEAANNEAYYMFRDDALFSEEEAKNRGGP--AALLVNLLLHDPDFELEQVKD
NLELFDKHNES
&gt;Mmu024661
MTTSQRQNDEAANNEAYYMFRDDALFSEEEAKNRGGP--AALLVNLLLHDPDFELEQVKD
NLELFDKH-ES
&gt;Dre37936
---SQRQNDEAANNEAYYMFRDDALFSEEEAKNRGGP--AALLVNLLLHDPDFELEQVKD
NLELFDKH---
&gt;LcaM
---SQRQNDEAANNEAYYMFRDDALFSEEEAKNRGGP--AALLVNLLLHDPDFELEQVKD
NLELFDKH---
&gt;Tru14292
---SQRQNDEAANNEAYYMFRDDALFSEEEAKNRGGP--AALLVNLLLHDPDFELEQVKD
NLELFDKH---
&gt;Ola20972
---SQRQNDEAANNEAYYMFRDDALFSEEEAKNRGGP--AALLVNLLLHDPDFELEQVKD
NLELFDKH---
</pre></div><div name=" terminalonly" style="display:none; width: -moz-fit-content"><b> Alignment </b> + "-gt 0.9 -terminalony" <pre class="fragment">&gt;Csa004271
-----------------MYMAMGHFFDRDDVALKNISEYFKECSEEEREHANKMIEFHNK
RGGTTTYFPIKAPGSFDPANFNTIKAMNCALALEVNVNKSLLALHE--TANGDPEFQDFI
EANFLHEQVDAIKKLKDYITNLKLVG---TGLGEFLFDKHFKS---
&gt;Xtr21234
---SQRQNDEAANNEAYTYLSMSHYFDRDDVALHHVAEFFKEQSKEERECAEKLMKCQNK
RGGRIVLQDIKKPERDEWG--STLDAMQTALDLEKHVNQALLDLHNLATERKDPHICDFL
ESEHLDEQVKHMKKFGDHITNLKRLGVPQNGMGEYLFDKHSLS---
&gt;LcaH
---SQRQNDEAANNEAYVYLSMAYYFDRDDQALHNFAKFFRHQSHEEREHAEKLMKLQNQ
RGGRIFLQDVRKPDRDEWG--SGVEALECALQLEKSVNQSLLDLHKLCSDHNDPHLCDFI
ETHYLDEQVKSIKELADWVTNLRRMGAPQNGMAEYLFDKHTLG--S
&gt;Hsa167996
MTTSQRQNDEAANNEAYVYLSMSYYFDRDDVALKNFAKYFLHQSHEEREHAEKLMKLQNQ
RGGRIFLQDIKKPDCDDWE--SGLNAMECALHLEKNVNQSLLELHKLATDKNDPHLCDFI
ETHYLNEQVKAIKELGDHVTNLRKMGAPESGLAEYLFDKHTLGNES
&gt;Mmu024661
MTTSQRQNDEAANNEAYVYLSMSCYFDRDDVALKNFAKYFLHQSHEEREHAEKLMKLQNQ
RGGRIFLQDIKKPDRDDWE--SGLNAMECALHLEKSVNQSLLELHKLATDKNDPHLCDFI
ETYYLSEQVKSIKELGDHVTNLRKMGAPEAGMAEYLFDKHTLG-ES
&gt;Dre37936
---SQRQNDEAANNEAYTYTSMAHYFKRDDVALPGFAKFFKKNSEEEREHAEKFMEFQNK
RGGRIVLQDIKKPDRDVWG--NGLIAMQCALQLEKNVNQALLDLHKLATEMGDPHLCDFL
ETHYLNEQVEAIKKLGDHITNLSKMDAGNNRMAEYLFDKHTLD---
&gt;LcaM
---SQRQNDEAANNEAYTYTSMAFYFSRDDVALPGFSHFFKENSDEEREHAEKLLSFQNK
RGGHIFLQDIKKPERDEWG--SGLEAMQCALQLKKNVNQALLDLHKLASDHGDPHLCDFL
ETHYLNEQVEAIKKLGDYISNLSRMDAQKNKMAEYLFDKHSLG---
&gt;Tru14292
---SQRQNDEAANNEAYTYTSMAFFFSRDDVALPGFAHFFKENSDEEREHAEKLLSFQNK
RGGRIFLQDIKKPERDEWG--SGLEAMQCALQLEKKVNQALLDLHKLASDHVDPHLCDFL
ESHYLNEQVEAIKKLGDYITNLSRMDAQNNKMAEYLFDKHTLG---
&gt;Ola20972
---SQRQNDEAANNEAYTYTSMAFYFDRDDVALPGFSHFFKENSHEEKEHADKLLSFQNK
RGGRIFLQDVKKPERDEWG--SGLEAMQCALQLEKNVNQALLDLHKVASDHKDPHMCDFL
ETHYLNEQVESIKKIGDHITNLTRMDAHTNKMAEYLFDKHTLG---
</pre> </div><h2><a class="anchor" id="Block"></a>
Block</h2>
<p>Block argument allows to clean the alignment using different methods, <br />
and later keep only residues that fall inside a keep-residues block of <br />
the specified number of residues.<br />
 This allows us to remove columns that obtained good scores by random chance.</p>
<p> <button onclick="myFunction(' block')" class="btn btn-danger"><b>  Example of use</b></button> </p><div name=" block" style="display:none; width: -moz-fit-content"></div><div name=" block" style="display:none; width: -moz-fit-content"><b> Alignment </b> <pre class="fragment">&gt;Sp6
---ASGAILT-LGIYLFTLCAVISVSWYLAWLGLEINMMAII
&gt;Sp17
FAYTAPDLL-LIGFLLKTVA-TFGDTWFQLWQGLDLNKMPVF
&gt;Sp10
-----DPAVL-FVIMLGTIT-KFSSEWFFAWLGLEINMMVII
&gt;Sp8
---GLGKVIV-YGIVLGTKSDQFSVVWLFPWNGLQIHMMGII
&gt;Sp26
AAAAAAALLTYLGLFLGTDYENFAAAAANAWLGLEINMMAQI
&gt;Sp33
-----PTILNIAGLHMETDI-NFSLAWFQAWGGLEINKQAIL
</pre></div><div name=" block" style="display:none; width: -moz-fit-content"><b> Alignment </b> + "-strict -block 2" <pre class="fragment">&gt;Sp6
ILGIYLISVSWYLAWLGLEINMMAII
&gt;Sp17
LLGFLLFGDTWFQLWQGLDLNKMPVF
&gt;Sp10
AVVIMLFSSEWFFAWLGLEINMMVII
&gt;Sp8
VIGIVLFSVVWLFPWNGLQIHMMGII
&gt;Sp26
LLGLFLFAAAAANAWLGLEINMMAQI
&gt;Sp33
ILGLHMFSLAWFQAWGGLEINKQAIL
</pre></div><div name=" block" style="display:none; width: -moz-fit-content"><b> Alignment </b> + "-strict -block 6" <pre class="fragment">&gt;Sp6
ISVSWYLAWLGLEINMMAII
&gt;Sp17
FGDTWFQLWQGLDLNKMPVF
&gt;Sp10
FSSEWFFAWLGLEINMMVII
&gt;Sp8
FSVVWLFPWNGLQIHMMGII
&gt;Sp26
FAAAAANAWLGLEINMMAQI
&gt;Sp33
FSLAWFQAWGGLEINKQAIL
</pre></div><div name=" block" style="display:none; width: -moz-fit-content"> </div><h2><a class="anchor" id="Clusters"></a>
Clusters</h2>
<p>It is possible to apply a clusterization of the sequences, <br />
to keep only the most representative sequences of the variability <br />
of the whole alignment. <br />
</p>
<p>It will output a alignment with the most representative sequence<br />
 for each cluster. <br />
</p>
<p>The user must provide the number of clusters to create, which is the <br />
 same as the number of sequences present on the final alignment;</p>
<p> <button onclick="myFunction(' clusters')" class="btn btn-danger"><b>  Example of use</b></button> </p><div name=" clusters" style="display:none; width: -moz-fit-content"></div><div name=" clusters" style="display:none; width: -moz-fit-content"><b> Alignment </b> <pre class="fragment">&gt;Sp6
---ASGAILT-LGIYLFTLCAVISVSWYLAWLGLEINMMAII
&gt;Sp17
FAYTAPDLL-LIGFLLKTVA-TFGDTWFQLWQGLDLNKMPVF
&gt;Sp10
-----DPAVL-FVIMLGTIT-KFSSEWFFAWLGLEINMMVII
&gt;Sp8
---GLGKVIV-YGIVLGTKSDQFSVVWLFPWNGLQIHMMGII
&gt;Sp26
AAAAAAALLTYLGLFLGTDYENFAAAAANAWLGLEINMMAQI
&gt;Sp33
-----PTILNIAGLHMETDI-NFSLAWFQAWGGLEINKQAIL
</pre></div><div name=" clusters" style="display:none; width: -moz-fit-content"><b> Alignment </b> + "-clusters 1" <pre class="fragment">&gt;Sp26
AAAAAAALLTYLGLFLGTDYENFAAAAANAWLGLEINMMAQI
</pre></div><div name=" clusters" style="display:none; width: -moz-fit-content"><b> Alignment </b> + "-clusters 4" <pre class="fragment">&gt;Sp17
FAYTAPDLL-LIGFLLKTVA-TFGDTWFQLWQGLDLNKMPVF
&gt;Sp8
---GLGKVIV-YGIVLGTKSDQFSVVWLFPWNGLQIHMMGII
&gt;Sp26
AAAAAAALLTYLGLFLGTDYENFAAAAANAWLGLEINMMAQI
&gt;Sp33
-----PTILNIAGLHMETDI-NFSLAWFQAWGGLEINKQAIL
</pre></div><div name=" clusters" style="display:none; width: -moz-fit-content"> </div><h2><a class="anchor" id="MaxIdentity"></a>
Max Identity</h2>
<p>Max Identity allows us to remove sequences with an identity<br />
 bigger than a user-provided threshold. <br />
</p>
<p>This removes the sequences that are too similar.</p>
<p> <button onclick="myFunction(' max-identity')" class="btn btn-danger"><b>  Example of use</b></button> </p><div name=" max-identity" style="display:none; width: -moz-fit-content"></div><div name=" max-identity" style="display:none; width: -moz-fit-content"><b> Alignment </b> <pre class="fragment">&gt;Sp6
---ASGAILT-LGIYLFTLCAVISVSWYLAWLGLEINMMAII
&gt;Sp17
FAYTAPDLL-LIGFLLKTVA-TFGDTWFQLWQGLDLNKMPVF
&gt;Sp10
-----DPAVL-FVIMLGTIT-KFSSEWFFAWLGLEINMMVII
&gt;Sp8
---GLGKVIV-YGIVLGTKSDQFSVVWLFPWNGLQIHMMGII
&gt;Sp26
AAAAAAALLTYLGLFLGTDYENFAAAAANAWLGLEINMMAQI
&gt;Sp33
-----PTILNIAGLHMETDI-NFSLAWFQAWGGLEINKQAIL
</pre></div><div name=" max-identity" style="display:none; width: -moz-fit-content"><b> Alignment </b> + "-maxidentity 0.3" <pre class="fragment">&gt;Sp8
---GLGKVIV-YGIVLGTKSDQFSVVWLFPWNGLQIHMMGII
&gt;Sp26
AAAAAAALLTYLGLFLGTDYENFAAAAANAWLGLEINMMAQI
</pre></div><div name=" max-identity" style="display:none; width: -moz-fit-content"><b> Alignment </b> + "-maxidentity 0.4" <pre class="fragment">&gt;Sp17
FAYTAPDLL-LIGFLLKTVA-TFGDTWFQLWQGLDLNKMPVF
&gt;Sp8
---GLGKVIV-YGIVLGTKSDQFSVVWLFPWNGLQIHMMGII
&gt;Sp26
AAAAAAALLTYLGLFLGTDYENFAAAAANAWLGLEINMMAQI
&gt;Sp33
-----PTILNIAGLHMETDI-NFSLAWFQAWGGLEINKQAIL
</pre></div><div name=" max-identity" style="display:none; width: -moz-fit-content"> </div><h2><a class="anchor" id="Compareset"></a>
Compareset</h2>
<p>Compareset allows the user to provide multipe MSA and to obtain Consistency scores. <br />
 These scores can be used to select the most consistent alignment in the set provided <br />
 or to use them as a trimming statistics to remove columns that are not consistent enough <br />
 among the set. <br />
</p>
<p>While Compareset will select the most consistent alignment, it is possible to force it to <br />
 select an specific alignment by using the -forceselect argument.</p>
<p> <button onclick="myFunction(' compareset')" class="btn btn-danger"><b>  Example of use</b></button> </p><div name=" compareset" style="display:none; width: -moz-fit-content"></div><div name=" compareset" style="display:none; width: -moz-fit-content"><b> alignments_comparison </b> <pre class="fragment">example.001.AA.phy
example.002.AA.phy
example.003.AA.phy
</pre></div><div name=" compareset" style="display:none; width: -moz-fit-content"><b> Alignment example.001.AA.phy </b> <pre class="fragment"> 6 60
Sp8          ----------GLGKV---IVY-GIVLGTKS-DQFSNWVVWL-----FPWNGLQIHMMGII
Sp17         --------FAYTAPD---LLLIGFLLKTVA-T-FG--DTWF-----QLWQGLDLNKMPVF
Sp10         ----------DPAVL----FV--IMLGTIT-K-FS--SEWF-----FAWLGLEINMMVII
Sp26         AAAAAAAA----ALL---TYL-GLFLGTDY-----EN---FAAAAANAWLGLEINMMAQI
Sp33         -----------PTIL---NIA-GLHMETDI-N-FS--LAWF-----QAWGGLEINKQAIL
Sp6          ----------ASGAI---LTL-GIYLFTLC-AVIS--VSWY-----LAWLGLEINMMAII
</pre></div><div name=" compareset" style="display:none; width: -moz-fit-content"><b> Alignment example.002.AA.phy </b> <pre class="fragment"> 6 46
Sp8          -----GLGKVIV-YGIVLGTKSDQFSNWVVWLFPWNGLQIHMMGII
Sp17         --FAYTAPDLLL-IGFLLKTV-ATFG--DTWFQLWQGLDLNKMPVF
Sp10         ------DPAVL--FVIMLGTI-TKFS--SEWFFAWLGLEINMMVII
Sp26         AAAAAAAAALLTYLGLFLGTDYENFA--AAAANAWLGLEINMMAQI
Sp33         -------PTILNIAGLHMETD-INFS--LAWFQAWGGLEINKQAIL
Sp6          -----ASGAILT-LGIYLFTLCAVIS--VSWYLAWLGLEINMMAII
</pre></div><div name=" compareset" style="display:none; width: -moz-fit-content"><b> Alignment example.003.AA.phy </b> <pre class="fragment"> 6 47
Sp8          -----GLGKVIV-YGIV-LGTKSDQFSNWVVWLFPWNGLQIHMMGII
Sp10         -------DPAVL-FVIM-LGTIT-KFS--SEWFFAWLGLEINMMVII
Sp26         AAAAAAAAALLTYLGL-FLGTDYENFA--AAAANAWLGLEINMMAQI
Sp6          -----ASGAILT-LGIY-LFTLCAVIS--VSWYLAWLGLEINMMAII
Sp17         --FAYTAPDLL-LIGFL-LKTVA-TFG--DTWFQLWQGLDLNKMPVF
Sp33         -------PTILNIAGLHME-TDI-NFS--LAWFQAWGGLEINKQAIL
</pre></div><div name=" compareset" style="display:none; width: -moz-fit-content"><b> alignments_comparison </b> + "-compareset" = <b> Alignment example.002.AA.phy </b> <pre class="fragment"> 6 46
Sp8          -----GLGKVIV-YGIVLGTKSDQFSNWVVWLFPWNGLQIHMMGII
Sp17         --FAYTAPDLLL-IGFLLKTV-ATFG--DTWFQLWQGLDLNKMPVF
Sp10         ------DPAVL--FVIMLGTI-TKFS--SEWFFAWLGLEINMMVII
Sp26         AAAAAAAAALLTYLGLFLGTDYENFA--AAAANAWLGLEINMMAQI
Sp33         -------PTILNIAGLHMETD-INFS--LAWFQAWGGLEINKQAIL
Sp6          -----ASGAILT-LGIYLFTLCAVIS--VSWYLAWLGLEINMMAII
</pre></div><div name=" compareset" style="display:none; width: -moz-fit-content"><b> alignments_comparison </b> + "-compareset -forceselect example.003.AA.phy" = <b> Alignment example.003.AA.phy </b> <pre class="fragment"> 6 47
Sp8          -----GLGKVIV-YGIV-LGTKSDQFSNWVVWLFPWNGLQIHMMGII
Sp10         -------DPAVL-FVIM-LGTIT-KFS--SEWFFAWLGLEINMMVII
Sp26         AAAAAAAAALLTYLGL-FLGTDYENFA--AAAANAWLGLEINMMAQI
Sp6          -----ASGAILT-LGIY-LFTLCAVIS--VSWYLAWLGLEINMMAII
Sp17         --FAYTAPDLL-LIGFL-LKTVA-TFG--DTWFQLWQGLDLNKMPVF
Sp33         -------PTILNIAGLHME-TDI-NFS--LAWFQAWGGLEINKQAIL
</pre></div><div name=" compareset" style="display:none; width: -moz-fit-content"><b> alignments_comparison </b> + "-compareset -forceselect example.003.AA.phy -ct 0.8" <pre class="fragment"> 6 27
Sp8          GIVLGTKQFSVVWFPWNGLQIHMMGII
Sp17         GFLLKTV-FGDTWQLWQGLDLNKMPVF
Sp10         -IMLGTI-FSSEWFAWLGLEINMMVII
Sp26         GLFLGTD------NAWLGLEINMMAQI
Sp33         GLHMETD-FSLAWQAWGGLEINKQAIL
Sp6          GIYLFTLVISVSWLAWLGLEINMMAII
</pre> </div><h2><a class="anchor" id="Conserve"></a>
Minimum Alignment Conservation Threshold</h2>
<p>It is possible, depending on the MSA provided and the trimming methods selected <br />
 to trim the MSA in a excessively aggresive way. <br />
</p>
<p>This may lead to a MSA with few or no residues.</p>
<p>To solve this issue, one of the solutions proposed is to conserve <br />
 a minimum percentage of the alignment.</p>
<p>To do so, the alignment will start on the center of the alignment,<br />
 which is supposed to be better aligned than the terminal ends, and <br />
 starts recovering residues in both directions until the desired <br />
 Minimum Alignment Conservation Threshold has been achieved.</p>
<p> <button onclick="myFunction(' min-cons')" class="btn btn-danger"><b>  Example of use</b></button> </p><div name=" min-cons" style="display:none; width: -moz-fit-content"></div><div name=" min-cons" style="display:none; width: -moz-fit-content"><b> Alignment </b> <pre class="fragment">&gt;CRAB_ANAPL
MDITIHNPLIRRPLFSWLAPSRIFDQIFGEHLQESELLPASPSLSPFLMRSPIFRMPSWL
ETGLSEMRLEKDKFSVNLDVKHFSPEELKVKVLGDMVEIHGKHEERQDEHGFIAREFNRK
YRIPADVDPLTITSSLSLDGVLTVSAPRKQSDVPERSIPITREEKPAIAGAQR
&gt;CRAB_BOVIN
MDIAIHHPWIRRPFFPFHSPSRLFDQFFGEHLLESDLFPASTSLSPFYLRPPSFLRAPSW
IDTGLSEMRLEKDRFSVNLDVKHFSPEELKVKVLGDVIEVHGKHEERQDEHGFISREFHR
KYRIPADVDPLAITSSLSSDGVLTVNGPRKQASGPERTIPITREEKPAVTAAP
&gt;CRAB_CHICK
MDITIHNPLVRRPLFSWLTPSRIFDQIFGEHLQESELLPTSPSLSPFLMRSPFFRMPSWL
ETGLSEMRLEKDKFSVNLDVKHFSPEELKVKVLGDMIEIHGKHEERQDEHGFIAREFSRK
YRIPADVDPLTITSSLSLDGVLTVSAPRKQSDVPERSIPITREEKPAIAGSQ-
&gt;CRAB_HUMAN
MDIAIHHPWIRRPFFPFHSPSRLFDQFFGEHLLESDLFPTSTSLSPFYLRPPSFLRAPSW
FDTGLSEMRLEKDRFSVNLDVKHFSPEELKVKVLGDVIEVHGKHEERQDEHGFISREFHR
KYRIPADVDPLTITSSLSSDGVLTVNGPRKQVSGPERTIPITREEKPAVTAAP
&gt;CRAB_MESAU
MDIAIHHPWIRRPFFPFHSPSRLFDQFFGEHLLESDLFSTATSLSPFYLRPPSFLRAPSW
IDTGLSEMRMEKDRFSVNLDVKHFSPEELKVKVLGDVVEVHGKHEERQDEHGFISREFHR
KYRIPADVDPLTITSSLSSDGVLTVNGPRKQASGPERTIPITREEKPAVTAAP
&gt;CRAB_MOUSE
MDIAIHHPWIRRPFFPFHSPSRLFDQFFGEHLLESDLFSTATSLSPFYLRPPSFLRAPSW
IDTGLSEMRLEKDRFSVNLDVKHFSPEELKVKVLGDVIEVHGKHEERQDEHGFISREFHR
KYRIPADVDPLAITSSLSSDGVLTVNGPRKQVSGPERTIPITREEKPAVAAAP
&gt;CRAB_RABIT
MDIAIHHPWIRRPFFPFHSPSRLFDQFFGEHLLESDLFPTSTSLSPFYLRPPSFLRAPSW
IDTGLSEMRLEKDRFSVNLDVKHFSPEELKVKVLGDVIEVHGKHEERQDEHGFISREFHR
KYRIPADVDPLTITSSLSSDGVLTVNGPRKQAPGPERTIPITREEKPAVTAAP
&gt;CRAB_RAT
MDIAIHHPWIRRPFFPFHSPSRLFDQFFGEHLLESDLFSTATSLSPFYLRPPSFLRAPSW
IDTGLSEMRMEKDRFSVNLDVKHFSPEELKVKVLGDVIEVHGKHEERQDEHGFISREFHR
KYRIPADVDPLTITSSLSSDGVLTVNGPRKQASGPERTIPITREEKPAVTAAP
&gt;CRAB_SQUAC
MDIAIQHPWLRRPLFPSSIFPSRIFDQNFGEHFDPDLFPSFSSMLSPFYWRMGAPMARMP
SWAQTGLSELRLDKDKFAIHLDVKHFTPEELRVKILGDFIEVQAQHEERQDEHGYVSREF
HRKYKVPAGVDPLVITCSLSADGVLTITGPRKVADVPERSVPISRDEKPAVAG
</pre></div><div name=" min-cons" style="display:none; width: -moz-fit-content"><b> Alignment </b> + "-st 0.8" <pre class="fragment">&gt;CRAB_ANAPL
MDIIPRRPFLS--
&gt;CRAB_BOVIN
MDIIPRRPFLS-K
&gt;CRAB_CHICK
MDIIPRRPFLS-K
&gt;CRAB_HUMAN
MDIIPRRPFLSK-
&gt;CRAB_MESAU
MDIIPRRPFLSKK
&gt;CRAB_MOUSE
MDIIPRRPFLS--
&gt;CRAB_RABIT
MDIIPRRPFLS--
&gt;CRAB_RAT
MDIIPRRPFLS-K
&gt;CRAB_SQUAC
MDIIPRRPFLSKK
</pre></div><div name=" min-cons" style="display:none; width: -moz-fit-content"><b> Alignment </b> + "-st 0.8 -cons 10" <pre class="fragment">&gt;CRAB_ANAPL
MDIIPIRRPFELSLE---
&gt;CRAB_BOVIN
MDIIPIRRPFELSLE--K
&gt;CRAB_CHICK
MDIIPVRRPFELSLE-RK
&gt;CRAB_HUMAN
MDIIPIRRPFELSLEK--
&gt;CRAB_MESAU
MDIIPIRRPFELSLEKKK
&gt;CRAB_MOUSE
MDIIPIRRPFELSLE---
&gt;CRAB_RABIT
MDIIPIRRPFELSLE---
&gt;CRAB_RAT
MDIIPIRRPFELSLE-KK
&gt;CRAB_SQUAC
MDIIPLRRPFDLSMQKKK
</pre></div><div name=" min-cons" style="display:none; width: -moz-fit-content"><b> Alignment </b> + "-st 0.8 -cons 40" <pre class="fragment">&gt;CRAB_ANAPL
MDITIHNPIRRPLFSPSRIFDQFGEHLESELLPASSLSPFMRPFPKKELVEQDETSSVKQ
STEKAQK-----
&gt;CRAB_BOVIN
MDIAIHHPIRRPFFPPSRLFDQFGEHLESDLFPASSLSPFLRPFAEDEVVERQDITSTRK
QIEEAAK----K
&gt;CRAB_CHICK
MDITIHNPVRRPLFSPSRIFDQFGEHLESELLPTSSLSPFMRPFPKKELIEQDETSSVKQ
STEKSQ----RK
&gt;CRAB_HUMAN
MDIAIHHPIRRPFFPPSRLFDQFGEHLESDLFPTSSLSPFLRPFAEDEVVERQDITSTRK
QIEEAAK--K--
&gt;CRAB_MESAU
MDIAIHHPIRRPFFPPSRLFDQFGEHLESDLFSTASLSPFLRPFAEDEVVERQDITSTRK
QIEEAAKKKKKK
&gt;CRAB_MOUSE
MDIAIHHPIRRPFFPPSRLFDQFGEHLESDLFSTASLSPFLRPFAEDEVVERQDITSTRK
QIEEAAKK----
&gt;CRAB_RABIT
MDIAIHHPIRRPFFPPSRLFDQFGEHLESDLFPTSSLSPFLRPFAEDEVVERQDITSTRK
QIEEAAKK----
&gt;CRAB_RAT
MDIAIHHPIRRPFFPPSRLFDQFGEHLESDLFSTASLSPFLRPFAEDEVVERQDITSTRK
QIEEAA----KK
&gt;CRAB_SQUAC
MDIAIQHPLRRPLFPFPSRIFDNFGEHDPDLFPSFSMLSPYWMAARDTVGQEERLVIVGP
RVSRVAPQQKKK
</pre> </div><h1><a class="anchor" id="otherTweaks"></a>
Other Tweaks</h1>
<h2><a class="anchor" id="ColNumbering"></a>
Column Numbering</h2>
<p>Outputs the relationship between residues in the output and input alignment.<br />
 It will output a -1 if the residue was removed, and the original residue position otherwise.</p>
<p> <button onclick="myFunction(' col-numbering')" class="btn btn-danger"><b> Example of use</b></button> </p><div name=" col-numbering " style="display:none; width: -moz-fit-content"></div><div name=" col-numbering " style="display:none; width: -moz-fit-content"><b> Alignment </b> + "-gt 0 -colnumbering" <pre class="fragment"> 6 42
Sp6          ---ASGAILT-LGIYLFTLCAVISVSWYLAWLGLEINMMAII
Sp17         FAYTAPDLL-LIGFLLKTVA-TFGDTWFQLWQGLDLNKMPVF
Sp10         -----DPAVL-FVIMLGTIT-KFSSEWFFAWLGLEINMMVII
Sp8          ---GLGKVIV-YGIVLGTKSDQFSVVWLFPWNGLQIHMMGII
Sp26         AAAAAAALLTYLGLFLGTDYENFAAAAANAWLGLEINMMAQI
Sp33         -----PTILNIAGLHMETDI-NFSLAWFQAWGGLEINKQAIL
</pre> <pre class="fragment">#ColumnsMap     0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41
</pre></div><div name=" col-numbering " style="display:none; width: -moz-fit-content"><b> Alignment </b> + "-gt 1 -colnumbering" <pre class="fragment"> 6 34
Sp6          GAILLGIYLFTLCVISVSWYLAWLGLEINMMAII
Sp17         PDLLIGFLLKTVATFGDTWFQLWQGLDLNKMPVF
Sp10         DPAVFVIMLGTITKFSSEWFFAWLGLEINMMVII
Sp8          GKVIYGIVLGTKSQFSVVWLFPWNGLQIHMMGII
Sp26         AALLLGLFLGTDYNFAAAAANAWLGLEINMMAQI
Sp33         PTILAGLHMETDINFSLAWFQAWGGLEINKQAIL
</pre> <pre class="fragment">#ColumnsMap     -1, -1, -1, -1, -1, 5, 6, 7, 8, -1, -1, 11, 12, 13, 14, 15, 16, 17, 18, 19, -1, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41
</pre></div><div name=" col-numbering " style="display:none; width: -moz-fit-content"></div><h2><a class="anchor" id="TrimmingReport"></a>
TrimmingReport</h2>
<script>
    $(document).ready(function() {
        $('[name="svgout"]' ).css({
            'max-width': ($('[name="htmlout"]').width() + 'px')
        });
    });
</script><p>trimAl is able to output a trimming report in two formats: HTML and SVG. <br />
There is no recommended format, but SVG is scalable and takes 10% of time to load on a browser than HTML. <br />
</p>
<p>You can check this functionality if you resize the page: <br />
 HTML will stay with a fixed size, while SVG will shrink.</p>
<p>The SVG created by trimAl are interactable, so you can go over a stat name or line in the graph <br />
 and it will be highlighted.</p>
<p>The SVG contains a legend of the background on the residues, <br />
and also includes the path of the input alignment.</p>
<p> <button onclick="myFunction(' html-out')" class="btn btn-danger"><b>  HTML Output Report</b></button> </p><div name=" html-out" style="display:none; width: -moz-fit-content"> <!DOCTYPE html>
<html><head>
    <meta http-equiv="Content-Type" content="text/html;charset=ISO-8859-1" />
    <title>trimAl v1.4 Summary</title>
    <style type="text/css" media="all">
    #b  { background-color: #3366ff; }
    #r  { background-color: #cc0000; }
    #g  { background-color: #33cc00; }
    #p  { background-color: #ff6666; }
    #m  { background-color: #cc33cc; }
    #o  { background-color: #ff9900; }
    #c  { background-color: #46C7C7; }
    #y  { background-color: #FFFF00; }
    .sel  { background-color: #B9B9B9; }
    .nsel { background-color: #E9E9E9; }
    .c1   { background-color: #FFFBF2; }
    .c2   { background-color: #FFF8CC; }
    .c3   { background-color: #FAF0BE; }
    .c4   { background-color: #F0EAD6; }
    .c5   { background-color: #F3E5AB; }
    .c6   { background-color: #F4C430; }
    .c7   { background-color: #C2B280; color: white; }
    .c8   { background-color: #DAA520; color: white; }
    .c9   { background-color: #B8860B; color: white; }
    .c10  { background-color: #918151; color: white; }
    .c11  { background-color: #967117; color: white; }
    .c12  { background-color: #6E5411; color: white; }
    </style>
  </head>

  <body>
  <pre>
    <span class=sel>Selected Sequences:     6 /Selected Residues:      17</span>
    <span class=nsel>Deleted Sequences:      0 /Deleted Residues:       29</span>

    Gaps Scores:         <span  class=c1>  =0=  </span><span  class=c2> <.001 </span><span  class=c3> <.050 </span><span  class=c4> <.100 </span><span  class=c5> <.150 </span><span  class=c6> <.200 </span><span  class=c7> <.250 </span><span  class=c8> <.350 </span><span  class=c9> <.500 </span><span class=c10> <.750 </span><span class=c11> <1.00 </span><span class=c12>  =1=  </span>
    Similarity Scores:   <span  class=c1>  =0=  </span><span  class=c2> <1e-6 </span><span  class=c3> <1e-5 </span><span  class=c4> <1e-4 </span><span  class=c5> <.001 </span><span  class=c6> <.010 </span><span  class=c7> <.100 </span><span  class=c8> <.250 </span><span  class=c9> <.500 </span><span class=c10> <.750 </span><span class=c11> <1.00 </span><span class=c12>  =1=  </span>
    Consistency Scores:  <span  class=c1>  =0=  </span><span  class=c2> <.001 </span><span  class=c3> <.050 </span><span  class=c4> <.100 </span><span  class=c5> <.150 </span><span  class=c6> <.200 </span><span  class=c7> <.250 </span><span  class=c8> <.350 </span><span  class=c9> <.500 </span><span class=c10> <.750 </span><span class=c11> <1.00 </span><span class=c12>  =1=  </span>

                                 10        20        30        40
                         =========+=========+=========+=========+======
    <span class=sel>Sp8</span>                  -----<span id=o>G</span>L<span id=o>G</span>K<span id=b>V</span><span id=b>I</span>V-<span id=c>Y</span><span id=o>G</span><span id=b>I</span><span id=b>V</span><span id=b>L</span><span id=o>G</span><span id=g>T</span>KSDQ<span id=b>F</span><span id=g>S</span>NW<span id=b>V</span>V<span id=b>W</span><span id=b>L</span>F<span id=y>P</span><span id=b>W</span>N<span id=o>G</span><span id=b>L</span><span id=g>Q</span><span id=b>I</span>H<span id=b>M</span><span id=b>M</span><span id=o>G</span><span id=b>I</span><span id=b>I</span>
    <span class=sel>Sp17</span>                 --FAYTA<span id=y>P</span>D<span id=b>L</span><span id=b>L</span>L-<span id=b>I</span><span id=o>G</span><span id=b>F</span><span id=b>L</span><span id=b>L</span>K<span id=g>T</span>V-AT<span id=b>F</span><span id=o>G</span>--DT<span id=b>W</span><span id=b>F</span>Q<span id=b>L</span><span id=b>W</span>Q<span id=o>G</span><span id=b>L</span><span id=m>D</span><span id=b>L</span><span id=g>N</span>K<span id=b>M</span><span id=y>P</span><span id=b>V</span><span id=b>F</span>
    <span class=sel>Sp10</span>                 ------D<span id=y>P</span>A<span id=b>V</span><span id=b>L</span>--<span id=b>F</span>V<span id=b>I</span><span id=b>M</span><span id=b>L</span><span id=o>G</span><span id=g>T</span>I-TK<span id=b>F</span><span id=g>S</span>--SE<span id=b>W</span><span id=b>F</span>F<span id=b>A</span><span id=b>W</span>L<span id=o>G</span><span id=b>L</span><span id=m>E</span><span id=b>I</span><span id=g>N</span><span id=b>M</span><span id=b>M</span><span id=b>V</span><span id=b>I</span><span id=b>I</span>
    <span class=sel>Sp26</span>                 AAAAAAA<span id=b>A</span>A<span id=b>L</span><span id=b>L</span>TY<span id=b>L</span><span id=o>G</span><span id=b>L</span><span id=b>F</span><span id=b>L</span><span id=o>G</span><span id=g>T</span>DYEN<span id=b>F</span>A--<span id=b>A</span>A<span id=b>A</span><span id=b>A</span>N<span id=b>A</span><span id=b>W</span>L<span id=o>G</span><span id=b>L</span><span id=m>E</span><span id=b>I</span><span id=g>N</span><span id=b>M</span><span id=b>M</span><span id=b>A</span>Q<span id=b>I</span>
    <span class=sel>Sp33</span>                 -------<span id=y>P</span>T<span id=b>I</span><span id=b>L</span>NI<span id=b>A</span><span id=o>G</span><span id=b>L</span><span id=c>H</span><span id=b>M</span>E<span id=g>T</span>D-IN<span id=b>F</span><span id=g>S</span>--<span id=b>L</span>A<span id=b>W</span><span id=b>F</span>Q<span id=b>A</span><span id=b>W</span><span id=o>G</span><span id=o>G</span><span id=b>L</span><span id=m>E</span><span id=b>I</span><span id=g>N</span>KQ<span id=b>A</span><span id=b>I</span><span id=b>L</span>
    <span class=sel>Sp6</span>                  -----AS<span id=o>G</span>A<span id=b>I</span><span id=b>L</span>T-<span id=b>L</span><span id=o>G</span><span id=b>I</span><span id=c>Y</span><span id=b>L</span>F<span id=g>T</span>LCAV<span id=b>I</span><span id=g>S</span>--<span id=b>V</span>S<span id=b>W</span><span id=c>Y</span>L<span id=b>A</span><span id=b>W</span>L<span id=o>G</span><span id=b>L</span><span id=m>E</span><span id=b>I</span><span id=g>N</span><span id=b>M</span><span id=b>M</span><span id=b>A</span><span id=b>I</span><span id=b>I</span>

    Selected Cols:       <span class=nsel> </span><span class=nsel> </span><span class=nsel> </span><span class=nsel> </span><span class=nsel> </span><span class=nsel> </span><span class=nsel> </span><span class=nsel> </span><span class=nsel> </span><span class=nsel> </span><span class=nsel> </span><span class=nsel> </span><span class=nsel> </span><span class=nsel> </span><span class=nsel> </span><span class=nsel> </span><span class=nsel> </span><span class=nsel> </span><span class=nsel> </span><span class=nsel> </span><span class=nsel> </span><span class=nsel> </span><span class=nsel> </span><span class=nsel> </span><span class=nsel> </span><span class=nsel> </span><span class=nsel> </span><span class=nsel> </span><span class=nsel> </span><span class=sel> </span><span class=sel> </span><span class=sel> </span><span class=sel> </span><span class=sel> </span><span class=sel> </span><span class=sel> </span><span class=sel> </span><span class=sel> </span><span class=sel> </span><span class=sel> </span><span class=sel> </span><span class=sel> </span><span class=sel> </span><span class=sel> </span><span class=sel> </span><span class=sel> </span>

    Gaps Scores:         <span  class=c6> </span><span  class=c6> </span><span  class=c8> </span><span  class=c8> </span><span  class=c8> </span><span class=c10> </span><span class=c11> </span><span class=c12> </span><span class=c12> </span><span class=c12> </span><span class=c12> </span><span class=c11> </span><span  class=c8> </span><span class=c12> </span><span class=c12> </span><span class=c12> </span><span class=c12> </span><span class=c12> </span><span class=c12> </span><span class=c12> </span><span class=c12> </span><span class=c10> </span><span class=c12> </span><span class=c12> </span><span class=c12> </span><span class=c12> </span><span  class=c6> </span><span  class=c6> </span><span class=c12> </span><span class=c12> </span><span class=c12> </span><span class=c12> </span><span class=c12> </span><span class=c12> </span><span class=c12> </span><span class=c12> </span><span class=c12> </span><span class=c12> </span><span class=c12> </span><span class=c12> </span><span class=c12> </span><span class=c12> </span><span class=c12> </span><span class=c12> </span><span class=c12> </span><span class=c12> </span>
    Similarity Scores:   <span class=c1> </span><span class=c1> </span><span  class=c2> </span><span class=c12> </span><span  class=c2> </span><span  class=c5> </span><span  class=c4> </span><span  class=c5> </span><span  class=c5> </span><span  class=c7> </span><span  class=c8> </span><span  class=c4> </span><span  class=c3> </span><span  class=c4> </span><span  class=c6> </span><span  class=c6> </span><span  class=c3> </span><span  class=c8> </span><span  class=c4> </span><span class=c12> </span><span  class=c3> </span><span  class=c2> </span><span  class=c4> </span><span  class=c4> </span><span  class=c7> </span><span  class=c6> </span><span class=c1> </span><span class=c1> </span><span  class=c4> </span><span  class=c5> </span><span  class=c6> </span><span  class=c5> </span><span  class=c3> </span><span  class=c5> </span><span class=c12> </span><span  class=c3> </span><span class=c12> </span><span class=c12> </span><span  class=c7> </span><span  class=c8> </span><span  class=c7> </span><span  class=c5> </span><span  class=c7> </span><span  class=c4> </span><span  class=c6> </span><span  class=c6> </span>
    Consistency Scores:  <span class=c1> </span><span class=c1> </span><span class=c10> </span><span class=c10> </span><span class=c10> </span><span class=c10> </span><span  class=c8> </span><span  class=c9> </span><span  class=c9> </span><span  class=c9> </span><span  class=c9> </span><span  class=c9> </span><span class=c12> </span><span class=c10> </span><span class=c11> </span><span class=c12> </span><span class=c11> </span><span class=c11> </span><span class=c11> </span><span class=c12> </span><span class=c12> </span><span class=c12> </span><span  class=c8> </span><span class=c10> </span><span class=c11> </span><span class=c11> </span><span class=c1> </span><span class=c1> </span><span class=c11> </span><span class=c11> </span><span class=c11> </span><span class=c11> </span><span class=c12> </span><span class=c12> </span><span class=c12> </span><span class=c12> </span><span class=c12> </span><span class=c12> </span><span class=c12> </span><span class=c12> </span><span class=c12> </span><span class=c12> </span><span class=c12> </span><span class=c12> </span><span class=c12> </span><span class=c12> </span>
    </pre>
  </body>
</html>
 </div><p> <button onclick="myFunction(' complementary')" class="btn btn-danger"><b>  SVG Output Report</b></button> </p><div name=" complementary" style="display:none; width: -moz-fit-content"> <svg xmlns="http://www.w3.org/2000/svg"             xmlns:xlink="http://www.w3.org/1999/xlink"             version="1.1"             viewbox="0 0 917.5 495">
<script type="text/ecmascript"> 
            // <![CDATA[[
                function onMouseOverLabel(evt) { 
                    var Line = document.getElementById(evt.target.id.replace("Label", "Line")); 
                    Line.setAttribute("style", "fill:none;stroke:black;stroke-width:3"); 
                    evt.target.setAttribute("font-size", 17); 
                } 
                function onMouseOutLabel(evt) { 
                    var Line = document.getElementById(evt.target.id.replace("Label", "Line")); 
                    Line.setAttribute("style", "fill:none;stroke:black;stroke-width:1"); 
                    evt.target.setAttribute("font-size", 15); 
                } 
                function onMouseOverLine(evt) { 
                    var Label = document.getElementById(evt.target.id.replace("Line", "Label")); 
                    evt.target.setAttribute("style", "fill:none;stroke:black;stroke-width:3"); 
                    Label.setAttribute("font-size", 17); 
                } 
                function onMouseOutLine(evt) { 
                    var Label = document.getElementById(evt.target.id.replace("Line", "Label")); 
                    evt.target.setAttribute("style", "fill:none;stroke:black;stroke-width:1"); 
                    Label.setAttribute("font-size", 15); 
                } 
            // ]]>
 </script>
<pattern id="selected_pattern" patternUnits="userSpaceOnUse" width="10" height="10"> 
<rect width="10" height="10" fill="green" opacity="0.5"/><path d="M-1,1 l2,-2 M0,10 l10,-10 M9,11 l2,-2" stroke="white" stroke-width="1"/></pattern> 
<pattern id="rejected" patternUnits="userSpaceOnUse" width="10" height="10"> 
<rect width="10" height="10" fill="red" opacity="0.5"/><path d="M2,-2 l-1,1 M10,-10 l0,10 M2,-2 l9,11" stroke="white" stroke-width="1"/></pattern> 
<rect                 style="fill:none;stroke:lightgrey"                 height="97.5"                 width="457.5px"                 x ="12.5px"                y ="12.5"/>
<text                 font-family = "monospace"                 font-size="15px"                 dy=".35em"                 x ="20"                 y="27"                 kerning="0"                 text-anchor="start"                 lengthAdjust="spacingAndGlyphs"                textLength="337.5"                 style="font-weight:100"                 xml:space="preserve" >../dataset/example.001.AA.msl;</text>
<text                 font-family = "monospace"                 font-size="15px"                 dy=".35em"                 x ="20"                 y="57"                 kerning="0"                 text-anchor="start"                 lengthAdjust="spacingAndGlyphs"                textLength="243.75"                 style="font-weight:100"                 xml:space="preserve" >Selected Sequences 	6 / 6</text>
<text                 font-family = "monospace"                 font-size="15px"                 dy=".35em"                 x ="20"                 y="87"                 kerning="0"                 text-anchor="start"                 lengthAdjust="spacingAndGlyphs"                textLength="263.25"                 style="font-weight:100"                 xml:space="preserve" >Selected Residues 		17 / 46</text>
<rect                 style="fill:none;stroke:lightgrey"                 height="97.5"                 width="457.5px"                 x ="12.5px"                y ="117.5"/>
<rect                     style="fill:royalblue;stroke:black;stroke-width:1px"                     height="15"                     width="15px"                     x ="20px"                    y ="125"/>
<text                 font-family = "monospace"                 font-size="11.25px"                 dy=".35em"                 x ="50"                 y ="132.5"                 kerning="0"                 text-anchor="start"                 lengthAdjust="spacingAndGlyphs"                textLength="82.5"                 style="font-weight:100"                 xml:space="preserve" >Hidrophobic</text>
<rect                     style="fill:aqua;stroke:black;stroke-width:1px"                     height="15"                     width="15px"                     x ="170px"                    y ="125"/>
<text                 font-family = "monospace"                 font-size="11.25px"                 dy=".35em"                 x ="200"                 y ="132.5"                 kerning="0"                 text-anchor="start"                 lengthAdjust="spacingAndGlyphs"                textLength="60"                 style="font-weight:100"                 xml:space="preserve" >Aromatic</text>
<rect                     style="fill:lime;stroke:black;stroke-width:1px"                     height="15"                     width="15px"                     x ="320px"                    y ="125"/>
<text                 font-family = "monospace"                 font-size="11.25px"                 dy=".35em"                 x ="350"                 y ="132.5"                 kerning="0"                 text-anchor="start"                 lengthAdjust="spacingAndGlyphs"                textLength="37.5"                 style="font-weight:100"                 xml:space="preserve" >Polar</text>
<rect                     style="fill:magenta;stroke:black;stroke-width:1px"                     height="15"                     width="15px"                     x ="20px"                    y ="158"/>
<text                 font-family = "monospace"                 font-size="11.25px"                 dy=".35em"                 x ="50"                 y ="165.5"                 kerning="0"                 text-anchor="start"                 lengthAdjust="spacingAndGlyphs"                textLength="112.5"                 style="font-weight:100"                 xml:space="preserve" >Negative Charge</text>
<rect                     style="fill:orange;stroke:black;stroke-width:1px"                     height="15"                     width="15px"                     x ="170px"                    y ="158"/>
<text                 font-family = "monospace"                 font-size="11.25px"                 dy=".35em"                 x ="200"                 y ="165.5"                 kerning="0"                 text-anchor="start"                 lengthAdjust="spacingAndGlyphs"                textLength="60"                 style="font-weight:100"                 xml:space="preserve" >Glycines</text>
<rect                     style="fill:darkviolet;stroke:black;stroke-width:1px"                     height="15"                     width="15px"                     x ="320px"                    y ="158"/>
<text                 font-family = "monospace"                 font-size="11.25px"                 dy=".35em"                 x ="350"                 y ="165.5"                 kerning="0"                 text-anchor="start"                 lengthAdjust="spacingAndGlyphs"                textLength="60"                 style="font-weight:100"                 xml:space="preserve" >Prolines</text>
<rect                     style="fill:red;stroke:black;stroke-width:1px"                     height="15"                     width="15px"                     x ="20px"                    y ="191"/>
<text                 font-family = "monospace"                 font-size="11.25px"                 dy=".35em"                 x ="50"                 y ="198.5"                 kerning="0"                 text-anchor="start"                 lengthAdjust="spacingAndGlyphs"                textLength="112.5"                 style="font-weight:100"                 xml:space="preserve" >Positive Charge</text>
<rect                     style="fill:lightgrey;stroke:black;stroke-width:1px"                     height="15"                     width="15px"                     x ="170px"                    y ="191"/>
<text                 font-family = "monospace"                 font-size="11.25px"                 dy=".35em"                 x ="200"                 y ="198.5"                 kerning="0"                 text-anchor="start"                 lengthAdjust="spacingAndGlyphs"                textLength="82.5"                 style="font-weight:100"                 xml:space="preserve" >Unconserved</text>
<rect                     style="fill:yellow;stroke:black;stroke-width:1px"                     height="15"                     width="15px"                     x ="320px"                    y ="191"/>
<text                 font-family = "monospace"                 font-size="11.25px"                 dy=".35em"                 x ="350"                 y ="198.5"                 kerning="0"                 text-anchor="start"                 lengthAdjust="spacingAndGlyphs"                textLength="60"                 style="font-weight:100"                 xml:space="preserve" >Prolines</text>
<text                 font-family = "monospace"                 font-size="15px"                 dy=".35em"                 x ="395"                 y="237"                 kerning="0"                 text-anchor="start"                 lengthAdjust="spacingAndGlyphs"                textLength="517.5"                 style="font-weight:100"                 xml:space="preserve" >0         10        20        30        40        </text>
<text                 font-family = "monospace"                 font-size="15px"                 dy=".35em"                 x ="395"                 y="252"                 kerning="0"                 text-anchor="start"                 lengthAdjust="spacingAndGlyphs"                textLength="517.5"                 style="font-weight:100"                 xml:space="preserve" >|         |         |         |         |         </text>
<line x1="395" x2="912.5" y1="252" y2="252" style="stroke:rgb(0,0,0);stroke-width:2"/>
<rect  style="fill:orange"  height="15"  width="11.25px"  x ="451.25"  y ="275" />
<rect  style="fill:orange"  height="15"  width="11.25px"  x ="473.75"  y ="275" />
<rect  style="fill:yellow"  height="30"  width="11.25px"  x ="473.75"  y ="290" />
<rect  style="fill:royalblue"  height="15"  width="11.25px"  x ="473.75"  y ="320" />
<rect  style="fill:yellow"  height="15"  width="11.25px"  x ="473.75"  y ="335" />
<rect  style="fill:orange"  height="15"  width="11.25px"  x ="473.75"  y ="350" />
<rect  style="fill:royalblue"  height="90"  width="11.25px"  x ="496.25"  y ="275" />
<rect  style="fill:royalblue"  height="90"  width="11.25px"  x ="507.5"  y ="275" />
<rect  style="fill:aqua"  height="15"  width="11.25px"  x ="541.25"  y ="275" />
<rect  style="fill:royalblue"  height="75"  width="11.25px"  x ="541.25"  y ="290" />
<rect  style="fill:orange"  height="30"  width="11.25px"  x ="552.5"  y ="275" />
<rect  style="fill:orange"  height="45"  width="11.25px"  x ="552.5"  y ="320" />
<rect  style="fill:royalblue"  height="90"  width="11.25px"  x ="563.75"  y ="275" />
<rect  style="fill:royalblue"  height="60"  width="11.25px"  x ="575"  y ="275" />
<rect  style="fill:aqua"  height="30"  width="11.25px"  x ="575"  y ="335" />
<rect  style="fill:royalblue"  height="90"  width="11.25px"  x ="586.25"  y ="275" />
<rect  style="fill:orange"  height="15"  width="11.25px"  x ="597.5"  y ="275" />
<rect  style="fill:orange"  height="30"  width="11.25px"  x ="597.5"  y ="305" />
<rect  style="fill:lime"  height="90"  width="11.25px"  x ="608.75"  y ="275" />
<rect  style="fill:royalblue"  height="90"  width="11.25px"  x ="665"  y ="275" />
<rect  style="fill:lime"  height="15"  width="11.25px"  x ="676.25"  y ="275" />
<rect  style="fill:orange"  height="15"  width="11.25px"  x ="676.25"  y ="290" />
<rect  style="fill:lime"  height="15"  width="11.25px"  x ="676.25"  y ="305" />
<rect  style="fill:lime"  height="30"  width="11.25px"  x ="676.25"  y ="335" />
<rect  style="fill:royalblue"  height="15"  width="11.25px"  x ="710"  y ="275" />
<rect  style="fill:royalblue"  height="45"  width="11.25px"  x ="710"  y ="320" />
<rect  style="fill:royalblue"  height="90"  width="11.25px"  x ="732.5"  y ="275" />
<rect  style="fill:royalblue"  height="75"  width="11.25px"  x ="743.75"  y ="275" />
<rect  style="fill:aqua"  height="15"  width="11.25px"  x ="743.75"  y ="350" />
<rect  style="fill:yellow"  height="15"  width="11.25px"  x ="766.25"  y ="275" />
<rect  style="fill:royalblue"  height="75"  width="11.25px"  x ="766.25"  y ="290" />
<rect  style="fill:royalblue"  height="90"  width="11.25px"  x ="777.5"  y ="275" />
<rect  style="fill:orange"  height="15"  width="11.25px"  x ="788.75"  y ="335" />
<rect  style="fill:orange"  height="90"  width="11.25px"  x ="800"  y ="275" />
<rect  style="fill:royalblue"  height="90"  width="11.25px"  x ="811.25"  y ="275" />
<rect  style="fill:lime"  height="15"  width="11.25px"  x ="822.5"  y ="275" />
<rect  style="fill:magenta"  height="75"  width="11.25px"  x ="822.5"  y ="290" />
<rect  style="fill:royalblue"  height="90"  width="11.25px"  x ="833.75"  y ="275" />
<rect  style="fill:lime"  height="75"  width="11.25px"  x ="845"  y ="290" />
<rect  style="fill:royalblue"  height="15"  width="11.25px"  x ="856.25"  y ="275" />
<rect  style="fill:royalblue"  height="30"  width="11.25px"  x ="856.25"  y ="305" />
<rect  style="fill:royalblue"  height="15"  width="11.25px"  x ="856.25"  y ="350" />
<rect  style="fill:royalblue"  height="60"  width="11.25px"  x ="867.5"  y ="275" />
<rect  style="fill:royalblue"  height="15"  width="11.25px"  x ="867.5"  y ="350" />
<rect  style="fill:orange"  height="15"  width="11.25px"  x ="878.75"  y ="275" />
<rect  style="fill:yellow"  height="15"  width="11.25px"  x ="878.75"  y ="290" />
<rect  style="fill:royalblue"  height="60"  width="11.25px"  x ="878.75"  y ="305" />
<rect  style="fill:royalblue"  height="45"  width="11.25px"  x ="890"  y ="275" />
<rect  style="fill:royalblue"  height="30"  width="11.25px"  x ="890"  y ="335" />
<rect  style="fill:royalblue"  height="90"  width="11.25px"  x ="901.25"  y ="275" />
<text                 font-family = "monospace"                 font-size="15px"                 dy=".35em"                 x ="5"                 y="282"                 kerning="0"                 text-anchor="start"                 textLength="33.75"                 style="font-weight:100">Sp8</text>
<text                 font-family = "monospace"                 font-size="15px"                 dy=".35em"                 x ="395"                 y="282"                 kerning="0"                 text-anchor="start"                 lengthAdjust="spacingAndGlyphs"                 textLength="517.5"                 style="font-weight:100">-----GLGKVIV-YGIVLGTKSDQFSNWVVWLFPWNGLQIHMMGII</text>
<text                 font-family = "monospace"                 font-size="15px"                 dy=".35em"                 x ="5"                 y="297"                 kerning="0"                 text-anchor="start"                 textLength="45"                 style="font-weight:100">Sp17</text>
<text                 font-family = "monospace"                 font-size="15px"                 dy=".35em"                 x ="395"                 y="297"                 kerning="0"                 text-anchor="start"                 lengthAdjust="spacingAndGlyphs"                 textLength="517.5"                 style="font-weight:100">--FAYTAPDLLL-IGFLLKTV-ATFG--DTWFQLWQGLDLNKMPVF</text>
<text                 font-family = "monospace"                 font-size="15px"                 dy=".35em"                 x ="5"                 y="312"                 kerning="0"                 text-anchor="start"                 textLength="45"                 style="font-weight:100">Sp10</text>
<text                 font-family = "monospace"                 font-size="15px"                 dy=".35em"                 x ="395"                 y="312"                 kerning="0"                 text-anchor="start"                 lengthAdjust="spacingAndGlyphs"                 textLength="517.5"                 style="font-weight:100">------DPAVL--FVIMLGTI-TKFS--SEWFFAWLGLEINMMVII</text>
<text                 font-family = "monospace"                 font-size="15px"                 dy=".35em"                 x ="5"                 y="327"                 kerning="0"                 text-anchor="start"                 textLength="45"                 style="font-weight:100">Sp26</text>
<text                 font-family = "monospace"                 font-size="15px"                 dy=".35em"                 x ="395"                 y="327"                 kerning="0"                 text-anchor="start"                 lengthAdjust="spacingAndGlyphs"                 textLength="517.5"                 style="font-weight:100">AAAAAAAAALLTYLGLFLGTDYENFA--AAAANAWLGLEINMMAQI</text>
<text                 font-family = "monospace"                 font-size="15px"                 dy=".35em"                 x ="5"                 y="342"                 kerning="0"                 text-anchor="start"                 textLength="45"                 style="font-weight:100">Sp33</text>
<text                 font-family = "monospace"                 font-size="15px"                 dy=".35em"                 x ="395"                 y="342"                 kerning="0"                 text-anchor="start"                 lengthAdjust="spacingAndGlyphs"                 textLength="517.5"                 style="font-weight:100">-------PTILNIAGLHMETD-INFS--LAWFQAWGGLEINKQAIL</text>
<text                 font-family = "monospace"                 font-size="15px"                 dy=".35em"                 x ="5"                 y="357"                 kerning="0"                 text-anchor="start"                 textLength="33.75"                 style="font-weight:100">Sp6</text>
<text                 font-family = "monospace"                 font-size="15px"                 dy=".35em"                 x ="395"                 y="357"                 kerning="0"                 text-anchor="start"                 lengthAdjust="spacingAndGlyphs"                 textLength="517.5"                 style="font-weight:100">-----ASGAILT-LGIYLFTLCAVIS--VSWYLAWLGLEINMMAII</text>
<rect  style="fill:none;stroke:black" height="75"  width="517.5px"  x ="395"  y ="380" />
<rect  style="fill:url(#rejected);stroke:black" height="75"  width="326.25px"  opacity="0.5" x ="395"  y ="380" />
<rect  style="fill:url(#selected_pattern);stroke:black" height="75"  width="191.25px"  opacity="0.5" x ="721.25"  y ="380" />
<text                         font-family = "monospace"                         font-size="15px"                         dy=".35em"                         x ="380"                         y="380"                         kerning="0"                         text-anchor="end"                         lengthAdjust="spacingAndGlyphs"                        style="font-weight:100"                         xml:space="preserve" >1</text>
<text                         font-family = "monospace"                         font-size="15px"                         dy=".35em"                         x ="380"                         y="455"                         kerning="0"                         text-anchor="end"                         lengthAdjust="spacingAndGlyphs"                        style="font-weight:100"                         xml:space="preserve" >0</text>
<polyline                         style="fill:none;stroke:black;stroke-width:1"                         stroke-dasharray="10,4"                         id ="GapsLine0"                         onmouseover="onMouseOverLine(evt)"                         onmouseout="onMouseOutLine(evt)"                         points="400.625,400 411.875,400 423.125,410 434.375,410 445.625,410 456.875,430 468.125,440 479.375,450 490.625,450 501.875,450 513.125,450 524.375,440 535.625,410 546.875,450 558.125,450 569.375,450 580.625,450 591.875,450 603.125,450 614.375,450 625.625,450 636.875,420 648.125,450 659.375,450 670.625,450 681.875,450 693.125,400 704.375,400 715.625,450 726.875,450 738.125,450 749.375,450 760.625,450 771.875,450 783.125,450 794.375,450 805.625,450 816.875,450 828.125,450 839.375,450 850.625,450 861.875,450 873.125,450 884.375,450 895.625,450 906.875,450 "/>
<text                         font-family = "monospace"                         font-size="15px"                         dy=".35em"                         x ="5"                         y="387"                         kerning="0"                         text-anchor="start"                         id ="GapsLabel0"                         onmouseover="onMouseOverLabel(evt)"                         onmouseout="onMouseOutLabel(evt)"                         textLength="123.75"                         style="font-weight:bold">Gaps Values</text>
<line x1="5" x2="128.75" y1="398.75" y2="398.75" style="stroke:rgb(0,0,0);stroke-width:1" stroke-dasharray="10,4" id ="GapsSubLine0"/>
<polyline                 style="fill:none;stroke:black;stroke-width:1"                 stroke-dasharray="5,1"                 id ="SimLine0"                 onmouseover="onMouseOverLine(evt)"                 onmouseout="onMouseOutLine(evt)"                 points="400.625,450 411.875,450 423.125,450 434.375,390 445.625,450 456.875,449.986 468.125,449.997 479.375,449.991 490.625,449.99 501.875,448.226 513.125,436.619 524.375,449.998 535.625,450 546.875,449.997 558.125,449.543 569.375,449.763 580.625,449.999 591.875,439.847 603.125,449.999 614.375,390 625.625,450 636.875,450 648.125,449.999 659.375,449.998 670.625,447.709 681.875,449.731 693.125,450 704.375,450 715.625,449.999 726.875,449.992 738.125,449.868 749.375,449.99 760.625,450 771.875,449.967 783.125,390 794.375,450 805.625,390 816.875,390 828.125,449.188 839.375,437.456 850.625,448.648 861.875,449.955 873.125,449.004 884.375,449.995 895.625,449.885 906.875,449.606 "/><text                         font-family = "monospace"                         font-size="15px"                         dy=".35em"                         x ="5"                         y="417.5"                         kerning="0"                         text-anchor="start"                         id ="SimLabel0"                         onmouseover="onMouseOverLabel(evt)"                         onmouseout="onMouseOutLabel(evt)"                         textLength="112.5"                         style="font-weight:bold">Sim Values</text>
<line x1="5" x2="117.5" y1="428.75" y2="428.75" style="stroke:rgb(0,0,0);stroke-width:1" stroke-dasharray="5,1" id ="SimSubLine0"/>
<polyline                 style="fill:none;stroke:black;stroke-width:1"                 stroke-dasharray="2,2"                 id ="ConsLine0"                 onmouseover="onMouseOverLine(evt)"                 onmouseout="onMouseOutLine(evt)"                 points="400.625,450 411.875,450 423.125,420 434.375,420 445.625,420 456.875,415 468.125,429 479.375,426 490.625,422 501.875,422 513.125,422 524.375,426 535.625,390 546.875,408 558.125,400 569.375,390 580.625,400 591.875,400 603.125,400 614.375,390 625.625,390 636.875,390 648.125,430 659.375,412 670.625,400 681.875,400 693.125,450 704.375,450 715.625,400 726.875,400 738.125,400 749.375,400 760.625,390 771.875,390 783.125,390 794.375,390 805.625,390 816.875,390 828.125,390 839.375,390 850.625,390 861.875,390 873.125,390 884.375,390 895.625,390 906.875,390 "/><text                         font-family = "monospace"                         font-size="15px"                         dy=".35em"                         x ="5"                         y="447.5"                         kerning="0"                         text-anchor="start"                         id ="ConsLabel0"                         onmouseover="onMouseOverLabel(evt)"                         onmouseout="onMouseOutLabel(evt)"                         textLength="112.5"                         style="font-weight:bold">Cons Values</text>
<line x1="5" x2="117.5" y1="458.75" y2="458.75" style="stroke:rgb(0,0,0);stroke-width:1" stroke-dasharray="2,2" id ="ConsSubLine0"/>
</svg> </div><h2><a class="anchor" id="Backstrans"></a>
Backstranslation</h2>
</div></div><!-- PageDoc -->
</div><!-- contents -->
</div><!-- doc-content -->
<!-- start footer part -->
<div id="nav-path" class="navpath"><!-- id is needed for treeview function! -->
  <ul>
    <li class="footer">Generated by
    <a href="http://www.doxygen.org/index.html">
    <img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.8.15 </li>
  </ul>
</div>
</body>
</html>
